---
title: "ISOFRAC_SimilarityLoop"
author: "Henrik Zauber"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_depth: 2
    toc_float: yes
    number_sections: yes
    code_folding: hide
  pdf_document:
    toc: yes
    toc_depth: '2'
editor_options:
  chunk_output_type: console
---

# PacBIO Combined Gels Enrichment:
## Fixing DB

```{r}
library(data.table)
library(ggplot2)
setwd(dirname(rstudioapi::documentPath()))
Results_CombinedGels_PacBio <- list()
if(0){
  load("../IsoFrac_20230217_SimilaritiesTesting_2/ORFpacBioProteinID_20230328.rda")
  load("../IsoFrac_20230217_SimilaritiesTesting_2/ORFpacBioProteinID_20230629.rda")
  ORFpacBioProteinID$PacBIO <- T
  ORFpacBioProteinID$cP <- F
  ORFpacBioProteinID$nP <- F
  ORFpacBioProteinID$AS <- F
  ORFpacBioProteinID$tr_canonical  <- F
  ORFpacBioProteinID$sp_canonical  <- F
  ORFpacBioProteinID$Canonical_in_gff  <- F
  ORFpacBioProteinID$SEQUENCE <- ORFpacBioProteinID$ORF_seq
  ORFpacBioProteinID$ORF_seq <- NULL
  table(duplicated(ORFpacBioProteinID$SEQUENCE))
  sel <- ORFpacBioProteinID$ORF_len>2
  sel[is.na(sel)]<- T
  ORFpacBioProteinID <- ORFpacBioProteinID[sel,]
  
  # needs to be manually activated in case Protein Sequences are missing.
  alldt_gn <- lapply(TableListNormEnsHdbScanMW_Peaks,function(x){x$dt_gn})
  alldt_gn <- rbindlist(alldt_gn)
  alldt_gn$seq <- NULL
  alldt_gn <- unique(alldt_gn)
  ORFpacBioProteinID$ENSG[is.na(ORFpacBioProteinID$ENSG)] <- "UNKNOWN"
  # ProteinSequencesCategories <<- ProteinSequencesCategories
  if(!exists("ProteinSequencesCategories")){
      load("ProteinSequencesCategories2.rda")
  }
  table(ORFpacBioProteinID$ENSG=="UNKNOWN")
  PSsingle <- strsplit(ProteinSequencesCategories$ENSG,";")
  ORFpacBioProteinID$OriginalENSG <- ORFpacBioProteinID$ENSG
  hu <- alldt_gn[,{
    gr <<- .BY
    ensi <<- unlist(strsplit(gr$ens,";"))
    if(all(is.na(match(ensi,unlist(PSsingle))))){
      # first check for mapping with UniAccession
      if(any(ProteinIDTag <- ProteinSequencesCategories$ID==gr$UniAccession)){
        ProteinSequencesCategories[ProteinIDTag]$ENSG <<- gr$ens
        cat("\r",.GRP,"Correction")
        
      }else{
        stop()
      }
      
    }else{
      cat("\r",.GRP,"All FINE")
      
    }
    .SD
  },.(ens,hgnc_symbol,UniAccession   )]
  
  table(ORFpacBioProteinID$ENSG!=ORFpacBioProteinID$OriginalENSG)
  save(ORFpacBioProteinID,ENSG_Count_level2,ENSG_Count_level1,ProteinSequences,file="ORFpacBioProteinID_20230328_mod2.rda")
  table(duplicated(ORFpacBioProteinID$SEQUENCE))
}else{

  ORFpacBio <- fread("./Wei_Chen_Lab/comprehensiveORF_PacBio_GENCODE_v2.txt",sep = "\t")
  library(Peptides)
  ORFpacBioProteinID <- ORFpacBio[,{
  cat("\r",.GRP)
  temp <- .SD
  ensg <- .BY$GENCODE_gene_ID
  temp2 <- temp[,.(ProteinID=paste(ensg,.GRP,sep="#"),
                   MW=mw(.BY$ORF_seq),
                  ENSG=ensg,
                  PacBioNames=paste(unique(paste(GENCODE_transcript_ID,sep="|")),collapse=";")
                   ),.(ORF_seq)]
  
  temp2
},.(GENCODE_gene_ID)]
  save(ORFpacBioProteinID,file="ORFpacBioProteinID.rda")
  ORFpacBioProteinIDNew <- ORFpacBioProteinID[,{
      temp <- .SD
    temp$ProteinID <- NULL
    temp$paralogues <- NULL
    temp$overlaped <- NULL
    temp$ORF_len <- NULL
    temp <- unique(temp)

  if(dim(temp)[1]>1){

    dim(unique(temp))
    stop()
  }
    temp
temp
},SEQUENCE]


}
# Results_CombinedGels_PacBio <- list()
if(!exists("Results_CombinedGels_PacBio")){
  Results_CombinedGels_PacBio <- list()
}

PacBIO <- ORFpacBioProteinIDNew

PacBIO[,PacBioCode:={
  pb <<- .BY$Other
  paste(unique(grep("^PB\\.",unlist(strsplit(pb,";|\\|")),value=T)),collapse=";")
},Other]


PacBIO$ENST2 <- paste(PacBIO$ENST,PacBIO$PacBioCode,sep="_")
PacBIO$ENST2[PacBIO$PacBioCode==""] <- PacBIO$ENST[PacBIO$PacBioCode==""]
PacBIO$ENST2[10]
PacBIO$ENST[10]
PacBIO$ENST <- PacBIO$ENST2
```
## Extra Functions

```{r}

library(RSQLite)
IF_DB_Peptidemapping <- function(x,IsoFrac_Domain_SequenceAnalysis,ProteinSequencesCategories,symbol=19,invertOrder=F,geli=1,allgels=T,facetPlot=T,gelTraces=T){
  
  cat("\r",x)
  subsetTable <<- IsoFrac_Domain_SequenceAnalysis[ENS==x]
  if(allgels){
    subsetTable$HDBscanCluster <- paste(subsetTable$HDBscanCluster,gsub("_",".",subsetTable$gels),sep="#")
    
  }else{
    subsetTable <- subsetTable[grep(c("5per","8per","10per")[geli],gels)]
  }
  # Domain Plot:
  # p <- (plotDomains(x=x,subset_dt = IsoFrac_Domain_SequenceAnalysis[ENS==x]))
  # MSA Alignment
  SEt <- ProteinSequencesCategories[ENSG==x,{paste(paste(unique(ProteinID),collapse=";"),round(mw(.BY$SEQUENCE)),sep=" MW:")},.(SEQUENCE)]
  SE <- SEt$SEQUENCE
  names(SE) <-gsub("_HUMAN","",SEt$V1)
  dfAlign <- msa::msa(SE,type = "protein",verbose=F,method = "ClustalOmega")
  
  dff <- as.data.frame(dfAlign@unmasked)
  M <- matrix(NA,ncol=nchar(dff$x[1]),nrow=dim(dff)[1])
  
  MatchFun <- lapply(strsplit((dff$x),""),function(x){
    x <- x
    sapply(1:length(x),function(y){
      if(x[y]!="-"){
        s <- sum(x[1:y]!="-")
      }else{
        s <- 0
      }
      
    })
  })
  if(length(names(SE))>0){
    names(MatchFun) <- names(dfAlign@unmasked)
  }
  
  MatchFun <- lapply(MatchFun,function(x){x[x==0] <- NA;x})
  hu <- as.data.table(MatchFun)
  hu$Position <- 1:dim(hu)[1]
  
  hu <- melt(hu,id.vars = "Position")
  hu[,MW:={as.numeric(gsub(".*.MW:","",variable))},value]
  hu$variable  <- factor(as.character(hu$variable),unique(hu$variable[order(hu$MW)]))
  
  
  DomainInfo <- PfamDomainMappingsDomainSequence[ensembl_gene_id==x]
  hu$Domain <- NULL
  hu$Domain <- ""
  print("Mapping Domains and Peptide to Protein Sequence")
  for(i in unique(hu$variable)){
    # Extracting Sequence
    SEi <- which(names(SE)==i)
    SEsel <- SE[SEi]
    # Adding Domain Information
    if(length(SEsel)>0){
      DomainInfoi <- DomainInfo[SEQUENCE==SEsel]
      if(dim(DomainInfoi)[1]>0){
        Dfun <- DomainInfoi[,{
          sel <- (!is.na(match(hu$value,.BY$pfam_start:.BY$pfam_end))&hu$variable==i)
          # print(any(sel))
          hu$Domain[sel] <<- paste(hu$Domain[sel],.BY$Pfam_Description)
          NULL
        },.(pfam_start,pfam_end,Pfam_Description)]
        
      }
      
      
    }
    ## Mapping Peptide Sequences
    hgnc <- SemiTrypticInfo[ENSG==x]$hgnc
    selectedP <- subsetTable[ENS==x,]
    SEsel_IL <- gsub("I","L",SEsel)
    PeptidesMappings <- selectedP[,{
      gr <- aregexec(.BY$sequence,SEsel_IL)
      gr <- gr[gr>0]
      if(length(gr)>0){
        sapply(gr,function(x){
          x:(x+nchar(.BY$sequence)-1)
        })
      }else{
        # print(.BY$sequence)
        as.integer(NA)
      }
      
      
    },.(sequence,DomainCategory,ClusterType,Category,IsoFrac_MW,HDBscanCluster)]
    PeptidesMappings$sequence <- paste(PeptidesMappings$sequence,PeptidesMappings$HDBscanCluster,sep="_")
    # PeptidesMappings[Category!="tryptic"]
    PeptidesMappingsCollapsed <- PeptidesMappings[,.(Category=paste(unique(Category[Category!="tryptic"]),collapse=";"),Mapped = length(Category),IsoFrac=paste(sort(unique(IsoFrac_MW)),collapse=";"),Sequences=paste(unique(sequence),collapse=";")),V1]
    # Matching to bigger table
    Posis <- hu$value
    # Posis[hu$variable!=i] <- NA # removing no protein specific entries
    PeptidesMappingsCollapsedMatched <- PeptidesMappingsCollapsed[match(Posis,V1)]
    hu$Category[hu$variable==i] <- PeptidesMappingsCollapsedMatched$Category[hu$variable==i]
    hu$Identified_by_Peptide[hu$variable==i] <- PeptidesMappingsCollapsedMatched$V1[hu$variable==i]
    hu$IsoFracMW[hu$variable==i] <- PeptidesMappingsCollapsedMatched$IsoFrac[hu$variable==i]
    hu$Identified <- !is.na(hu$Identified_by_Peptide)
    hu$Sequences[hu$variable==i] <- PeptidesMappingsCollapsedMatched$Sequences[hu$variable==i]
    
  }
  
  hu[,Domain_all :={
    di <- unique(.SD$Domain)
    di <- paste(unique(di),collapse=";")
  },Position]
  #### Extracting nterminal specific data 
  selectedP_terminal <- hu[Category=="new_nterm"|Category=="new_cterm",.(value=unique(value)),.(IsoFracMW,variable,Category,Position)]
  hu$Domain_all <- gsub("^; ","",hu$Domain_all)
  
  MSAPlot <- ggplot(hu[!is.na(value)][order(MW)],aes(Position,variable,color=Domain_all,size=Identified))+geom_point(pch=symbol,alpha=0.5)+ggplot2::scale_color_viridis_d()+ylab("DataBase Entries")
  
  SpeciPeps <- hu[!is.na(value)&Identified==TRUE,.(IsoFracMW_single={unlist(strsplit(IsoFracMW,";"))}),.(Position,variable,value,Domain,Category,Identified_by_Peptide,Identified,Domain_all,Sequences)]
  SpeciPeps <- SpeciPeps[,{
    sel <- rep(T,dim(.SD)[1])
    if(.BY$Category=="new_nterm"){
      sel <- which(Position==min(Position))
    }
    if(.BY$Category=="new_cterm"){
      sel <-  which(Position==max(Position))
    }
    
    .SD[sel]
  },.(Category,Sequences)]
  SpeciPeps$IsoFracMW_single <- factor(SpeciPeps$IsoFracMW_single,sort(as.numeric(unique(SpeciPeps$IsoFracMW_single))))
  SpeciPeps <- SpeciPeps[,{
    se <- unlist(strsplit(.BY$Sequences,";"))
    sefun <- lapply(se,function(x){
      xse <- unlist(strsplit(x,split="_"))
      data.table(Sequence=xse[1],cluster_single=xse[2])
    })
    sefun <- rbindlist(sefun)
    sefun
  },names(SpeciPeps)]
  SpeciPeps$gels <- gsub("^.*.#","",SpeciPeps$cluster_single)
  # MSAPlot2 <- ggplot(SpeciPeps,aes(Position,IsoFracMW_single,color=Domain_all,cex=Identified,pch=cluster))+geom_point()+ggplot2::scale_color_viridis_d()+ylab("IsoFrac IsoForms\nIdentified Peptides")
  MSAPlot2 <- ggplot(SpeciPeps,aes(Position,IsoFracMW_single,color=cluster_single,pch=cluster_single))+geom_point(position = position_dodge(width = 0.5))+ylab("IsoFrac IsoForms\nIdentified Peptides")
  SpeciPeps$Category_offset <- as.numeric(as.character(factor(SpeciPeps$Category,c("new_nterm","new_cterm",""),c(-0.5,1,0))))
  # SpeciPeps$Category <- factor(SpeciPeps$Category,c("new_nterm","new_cterm"),c(">","<"))
  
  
  # Domains <- hu[order(Position),{
  #   temp <- .SD
  #   up <- unique(temp$Position)
  #   wifun <- which(diff(up)>1)
  #   if(length(temp$Position)==0){
  #     ra <- as.data.table(list(range(temp$Position)))
  #   }else{
  #     wifun <- c(1,wifun)
  #     la <- lapply(1:(length(wifun)-1),function(it){
  #       temp$Position[c(wifun[it],wifun[it+1])]
  #       
  #     })
  #     ra <- as.data.table(la)
  #   }
  #   ra
  #   
  # },.(Domain_all)]
  
  # DomainSegments <- annotate("segment",)
  ra <- range(hu$Position)
  ano_IF <- annotate("text",SpeciPeps$Position,SpeciPeps$IsoFracMW,label=factor(SpeciPeps$Category,c("new_nterm","new_cterm"),c(">","<")),color="red",cex=5,vjust=SpeciPeps$Category_offset)
  ano_UP <- annotate("text",SpeciPeps$Position,SpeciPeps$variable,label=factor(SpeciPeps$Category,c("new_nterm","new_cterm"),c(">","<")),color="red",cex=5)
  ano_IF <- geom_vline(xintercept = SpeciPeps$Position[grepl("new",SpeciPeps$Category)],lty="dotted",color="red",size=1)
  
  # MSAPlot2 <- MSAPlot2+ano_IF+coord_cartesian(xlim=ra)+ggtitle("Mapped Peptides per IsoFrac Isoform")
  MSAPlot <- MSAPlot+ano_UP+coord_cartesian(xlim=ra)+ggtitle("Proteins and Pfam Domains")
  
  if(gelTraces&allgels&facetPlot){
    # MSAPlot2<<- MSAPlot2
    # MSAPlot2$data
    MSAPlot2  <- MSAPlot2+facet_wrap("gels",ncol = 1)
  }
  MSAPlot2 <<- MSAPlot2
  MSAPlot <<- MSAPlot
  if(invertOrder){
    
    ggout <- plot_grid(MSAPlot2
                       ,
                       MSAPlot,
                       ncol = 1,align="hv",rel_heights = c(2,1.5))
  }else{
    ggout <- plot_grid(MSAPlot+ano_UP+coord_cartesian(xlim=ra)+ggtitle("Proteins and Pfam Domains"),
                       MSAPlot2+ano_IF+coord_cartesian(xlim=ra)+ggtitle("Mapped Peptides per IsoFrac Isoform"),
                       ncol = 1,align="hv")
  }
  
  
  return(list(gg=  ggout,
              table=hu,
              SpecIDs=SpeciPeps,
              MSAPlot=MSAPlot+ano_UP+coord_cartesian(xlim=ra),
              MSAPlot2=MSAPlot2+ano_IF+coord_cartesian(xlim=ra))
         )
  
}

IF_DB_Peptidemapping_modified <- function(x,IsoFrac_Domain_SequenceAnalysis,ProteinSequencesCategories,symbol=19,invertOrder=F,geli=1,allgels=T,facetPlot=T,gelTraces=T){
  
  cat("\r",x)
  subsetTable <<- IsoFrac_Domain_SequenceAnalysis[ENS==x]
  
  if(allgels){
    subsetTable$HDBscanCluster <- paste(subsetTable$HDBscanCluster,gsub("_",".",subsetTable$gels),sep="#")
    
  }else{
    subsetTable <- subsetTable[grep(c("5per","8per","10per")[geli],gels)]
  }
  # Domain Plot:
  # p <- (plotDomains(x=x,subset_dt = IsoFrac_Domain_SequenceAnalysis[ENS==x]))
  # MSA Alignment
  SEt <- ProteinSequencesCategories[ENSG==x,{paste(paste(unique(ProteinID),collapse=";"),round(mw(.BY$SEQUENCE)),sep=" MW:")},.(SEQUENCE)]
  SE <- SEt$SEQUENCE
  names(SE) <-gsub("_HUMAN","",SEt$V1)
  dfAlign <- msa::msa(SE,type = "protein",verbose=F,method = "ClustalOmega")
  
  dff <- as.data.frame(dfAlign@unmasked)
  M <- matrix(NA,ncol=nchar(dff$x[1]),nrow=dim(dff)[1])
  
  MatchFun <- lapply(strsplit((dff$x),""),function(x){
    x <- x
    sapply(1:length(x),function(y){
      if(x[y]!="-"){
        s <- sum(x[1:y]!="-")
      }else{
        s <- 0
      }
      
    })
  })
  if(length(names(SE))>0){
    names(MatchFun) <- names(dfAlign@unmasked)
  }
  
  MatchFun <- lapply(MatchFun,function(x){x[x==0] <- NA;x})
  hu <- as.data.table(MatchFun)
  hu$Position <- 1:dim(hu)[1]
  
  hu <- melt(hu,id.vars = "Position")
  hu[,MW:={as.numeric(gsub(".*.MW:","",variable))},value]
  hu$variable  <- factor(as.character(hu$variable),unique(hu$variable[order(hu$MW)]))
  
  
  DomainInfo <- PfamDomainMappingsDomainSequence[ensembl_gene_id==x]
  hu$Domain <- NULL
  hu$Domain <- ""
  

  for(i in 1:dim(DomainInfo)){
    tempDomain <- DomainInfo[i,]
  }
  
  
  
  print("Mapping Domains and Peptide to Protein Sequence")
  for(i in unique(hu$variable)){
    # Extracting Sequence
    SEi <- which(names(SE)==i)
    SEsel <- SE[SEi]
    # Adding Domain Information
    if(length(SEsel)>0){
      DomainInfoi <- DomainInfo[SEQUENCE==SEsel]
      if(dim(DomainInfoi)[1]>0){
        # hu <<- hu
        # stop()
        Dfun <<- DomainInfoi[,{
          sel <- (!is.na(match(hu$value,.BY$pfam_start:.BY$pfam_end))&hu$variable==i)
          # print(any(sel))
          hu$Domain[sel] <<- paste(hu$Domain[sel],.BY$Pfam_Description)
          NULL
        },.(pfam_start,pfam_end,Pfam_Description)]
        if(dim(Dfun)[1]>0){
          stop()
        }
        # stop()
        
      }
      
      
    }
    ## Mapping Peptide Sequences
    hgnc <- SemiTrypticInfo[ENSG==x]$hgnc
    selectedP <- subsetTable[ENS==x,]
    SEsel_IL <- gsub("I","L",SEsel)
    PeptidesMappings <- selectedP[,{
      gr <- aregexec(.BY$sequence,SEsel_IL)
      gr <- gr[gr>0]
      if(length(gr)>0){
        sapply(gr,function(x){
          x:(x+nchar(.BY$sequence)-1)
        })
      }else{
        # print(.BY$sequence)
        as.integer(NA)
      }
      
      
    },.(sequence,DomainCategory,ClusterType,Category,IsoFrac_MW,HDBscanCluster)]
    PeptidesMappings$sequence <- paste(PeptidesMappings$sequence,PeptidesMappings$HDBscanCluster,sep="_")
    # PeptidesMappings[Category!="tryptic"]
    PeptidesMappingsCollapsed <- PeptidesMappings[,.(Category=paste(unique(Category[Category!="tryptic"]),collapse=";"),Mapped = length(Category),IsoFrac=paste(sort(unique(IsoFrac_MW)),collapse=";"),Sequences=paste(unique(sequence),collapse=";")),V1]
    # Matching to bigger table
    Posis <- hu$value
    # Posis[hu$variable!=i] <- NA # removing no protein specific entries
    PeptidesMappingsCollapsedMatched <- PeptidesMappingsCollapsed[match(Posis,V1)]
    hu$Category[hu$variable==i] <- PeptidesMappingsCollapsedMatched$Category[hu$variable==i]
    hu$Identified_by_Peptide[hu$variable==i] <- PeptidesMappingsCollapsedMatched$V1[hu$variable==i]
    hu$IsoFracMW[hu$variable==i] <- PeptidesMappingsCollapsedMatched$IsoFrac[hu$variable==i]
    hu$Identified <- !is.na(hu$Identified_by_Peptide)
    hu$Sequences[hu$variable==i] <- PeptidesMappingsCollapsedMatched$Sequences[hu$variable==i]
    
  }
  
  hu[,Domain_all :={
    di <- unique(.SD$Domain)
    di <- paste(unique(di),collapse=";")
  },Position]
  #### Extracting nterminal specific data 
  selectedP_terminal <- hu[Category=="new_nterm"|Category=="new_cterm",.(value=unique(value)),.(IsoFracMW,variable,Category,Position)]
  hu$Domain_all <- gsub("^; ","",hu$Domain_all)
  
  MSAPlot <- ggplot(hu[!is.na(value)][order(MW)],aes(Position,variable,color=Domain_all,size=Identified))+geom_point(pch=symbol,alpha=0.5)+ggplot2::scale_color_viridis_d()+ylab("DataBase Entries")
  
  SpeciPeps <- hu[!is.na(value)&Identified==TRUE,.(IsoFracMW_single={unlist(strsplit(IsoFracMW,";"))}),.(Position,variable,value,Domain,Category,Identified_by_Peptide,Identified,Domain_all,Sequences)]
  SpeciPeps <- SpeciPeps[,{
    sel <- rep(T,dim(.SD)[1])
    if(.BY$Category=="new_nterm"){
      sel <- which(Position==min(Position))
    }
    if(.BY$Category=="new_cterm"){
      sel <-  which(Position==max(Position))
    }
    
    .SD[sel]
  },.(Category,Sequences)]
  SpeciPeps$IsoFracMW_single <- factor(SpeciPeps$IsoFracMW_single,sort(as.numeric(unique(SpeciPeps$IsoFracMW_single))))
  SpeciPeps <- SpeciPeps[,{
    se <- unlist(strsplit(.BY$Sequences,";"))
    sefun <- lapply(se,function(x){
      xse <- unlist(strsplit(x,split="_"))
      data.table(Sequence=xse[1],cluster_single=xse[2])
    })
    sefun <- rbindlist(sefun)
    sefun
  },names(SpeciPeps)]
  SpeciPeps$gels <- gsub("^.*.#","",SpeciPeps$cluster_single)
  # MSAPlot2 <- ggplot(SpeciPeps,aes(Position,IsoFracMW_single,color=Domain_all,cex=Identified,pch=cluster))+geom_point()+ggplot2::scale_color_viridis_d()+ylab("IsoFrac IsoForms\nIdentified Peptides")
  # SpeciPeps <<- SpeciPeps
  
  SpeciPeps <- SpeciPeps[!(IsoFracMW_single=="20157"&
                             Sequence!="EDGPEGLDK"&
                             Sequence!="AMNLDQVSK"&
                             Sequence!="QVSSHLQVLAR")]
  
  MSAPlot2 <- ggplot(SpeciPeps,aes(Position,IsoFracMW_single,color=cluster_single,pch=cluster_single))+geom_point(position = position_dodge(width = 0.5))+ylab("IsoFrac IsoForms\nIdentified Peptides")
  SpeciPeps2 <- SpeciPeps[,.(confidence=dim(.SD)[1]),.(Position,IsoFracMW_single)]
  SpeciPeps2$MWkd <- as.character(round(as.numeric(as.character(SpeciPeps2$IsoFracMW_single))/1000,1))
  
  Domains <-   SpeciPeps[,{as.list(range(Position))},Domain_all]
  
  
  library(ggplot2)
  
  huha <<- hu[,as.list(range(Position)),.(variable,Domain)]
  huha <- huha[Domain!=""]
  # Create a data frame for the colored rectangles
  rects <- data.table(
    xmin = huha$V1,  # Start of the x ranges
    xmax = huha$V2,  # End of the x ranges
    ymin = -Inf,     # Extend to full y-axis (negative infinity)
    ymax = Inf,      # Extend to full y-axis (positive infinity)
    fill = huha$Domain  # Colors for each range
  )
  
  
  # Plot with ggplot2
  MSAPlot3 <- ggplot()+geom_rect(data = rects[fill!=""], aes(xmin = xmin, xmax = xmax, ymin = ymin, ymax = ymax, fill = fill), alpha = 0.5)+ 
    geom_point(data=SpeciPeps2,legend="fwieufh",aes(Position,MWkd,color="identified peptides"),pch="|",position = position_dodge(width = 0.5),size=10)+ylab("IsoFrac MW [kD]")+
    xlab("Meta Gene Position")+
    scale_color_manual(name = "identified peptides", values = c("black")) +theme(legend.title = element_blank())
    #+ggtitle("Protein Isoform Coverage")
  
  SpeciPeps$Category_offset <- as.numeric(as.character(factor(SpeciPeps$Category,c("new_nterm","new_cterm",""),c(-0.5,1,0))))
  # SpeciPeps$Category <- factor(SpeciPeps$Category,c("new_nterm","new_cterm"),c(">","<"))
  
  
  # Domains <- hu[order(Position),{
  #   temp <- .SD
  #   up <- unique(temp$Position)
  #   wifun <- which(diff(up)>1)
  #   if(length(temp$Position)==0){
  #     ra <- as.data.table(list(range(temp$Position)))
  #   }else{
  #     wifun <- c(1,wifun)
  #     la <- lapply(1:(length(wifun)-1),function(it){
  #       temp$Position[c(wifun[it],wifun[it+1])]
  #       
  #     })
  #     ra <- as.data.table(la)
  #   }
  #   ra
  #   
  # },.(Domain_all)]
  
  # DomainSegments <- annotate("segment",)
  ra <- range(hu$Position)
  ano_IF <- annotate("text",SpeciPeps$Position,SpeciPeps$IsoFracMW,label=factor(SpeciPeps$Category,c("new_nterm","new_cterm"),c(">","<")),color="red",cex=5,vjust=SpeciPeps$Category_offset)
  ano_UP <- annotate("text",SpeciPeps$Position,SpeciPeps$variable,label=factor(SpeciPeps$Category,c("new_nterm","new_cterm"),c(">","<")),color="red",cex=5)
  ano_IF <- geom_vline(xintercept = SpeciPeps$Position[grepl("new",SpeciPeps$Category)],lty="dotted",color="red",size=1)
  
  # MSAPlot2 <- MSAPlot2+ano_IF+coord_cartesian(xlim=ra)+ggtitle("Mapped Peptides per IsoFrac Isoform")
  MSAPlot <- MSAPlot+ano_UP+coord_cartesian(xlim=ra)+ggtitle("Proteins and Pfam Domains")
  
  if(gelTraces&allgels&facetPlot){
    # MSAPlot2<<- MSAPlot2
    # MSAPlot2$data
    MSAPlot2  <- MSAPlot2+facet_wrap("gels",ncol = 1)
  }
  MSAPlot2 <<- MSAPlot2
  MSAPlot <<- MSAPlot
  if(invertOrder){
    
    ggout <- plot_grid(MSAPlot2
                       ,
                       MSAPlot,
                       ncol = 1,align="hv",rel_heights = c(2,1.5))
  }else{
    ggout <- plot_grid(MSAPlot+ano_UP+coord_cartesian(xlim=ra)+ggtitle("Proteins and Pfam Domains"),
                       MSAPlot2+ano_IF+coord_cartesian(xlim=ra)+ggtitle("Mapped Peptides per IsoFrac Isoform"),
                       ncol = 1,align="hv")
  }
  
  
  return(list(gg=  ggout,
              table=hu,
              SpecIDs=SpeciPeps,
              MSAPlot=MSAPlot+ano_UP+coord_cartesian(xlim=ra),
              MSAPlot2=MSAPlot2+ano_IF+coord_cartesian(xlim=ra),
              MSAPlot3=MSAPlot3+ano_IF+coord_cartesian(xlim=ra))
  )
  
}


Traces_from_DB <- function(x,dbpath="../FinalViewer/IsoFrac_0.84/TableListNormEnsHdbScanMW_Peaks.sqlite",gel=2){
  db <- dbConnect(SQLite(),dbpath) # from IsoFrac Pre Scripts (Hypatia Server) 
  dbl <- dbListTables(db)
  # Extracting
  ensg_vec <- x
  gelSelection <- c("Exp1_5per_Try","Exp1_8per_Try","Exp1_10per_Try")[gel]
  # table name of zscorre
  zn <- grep("zscore",dbl,value=T)
  zn <- grep(gelSelection,zn,value=T)
  # table name of hdbscanTable
  
  hr <- grep("hdbscanTables_",dbListTables(db),value=T)
  hr <- grep(gelSelection,hr,value=T)
  # table name of MW
  mw <- grep("Fractions_MW",dbl,value=T)
  mw <- grep(gelSelection,mw,value=T)
  
  if(length(hr)>1){
    print("Too many selected candidates, picking first entry.")
    hr <- hr[1]
    zn <- zn[1]
  }
  hdbscans <- data.table(dbGetQuery(db,paste("SELECT * FROM",paste("'",hr,"'",sep = "") ,"WHERE EnsG = ",paste("'",ensg_vec,"'",sep=""))))
  
  
  IdentifiedPeaks_Table <- data.table(dbGetQuery(db,paste("SELECT * FROM IdentifiedPeaks_Table WHERE Ens=",paste("'",ensg_vec,"'",sep=""))))
  
  # Extract only these zscores:
  #"SELECT * FROM your_table LIMIT 1 OFFSET 2;"
  zscore <- lapply(hdbscans$id,function(x){
    limit <- 1
    offset <- x-1
    dbGetQuery(db,paste("SELECT * FROM",paste("'",zn,"'",sep = ""),"LIMIT",limit,"OFFSET",offset))
    
  })
  
  zscoredt <- rbindlist(zscore)
  zscoredt$id <- hdbscans$id
  zscoredt$cl <- hdbscans$cl_2
  zscoredt$Seq <- hdbscans$Seq
  
  zscoredt_melt <- melt(zscoredt,id.vars = c("id","cl","Seq"))
  
  # Adding MW 
  mwDT <- dbReadTable(db,mw)
  zscoredt_melt$MW <- mwDT[as.numeric(gsub("V","",zscoredt_melt$variable)),1]
  CPP <<- data.table(dbGetQuery(db,paste("SELECT * FROM Compressed_Pairs WHERE ENS=",paste("'",ensg_vec,"'",sep=""))))
  dbDisconnect(db)
  plotlyScale <- c("#1f77b4", "#ff7f0e", "#2ca02c", "#d62728", "#9467bd", "#8c564b", "#e377c2", "#7f7f7f", "#bcbd22", "#17becf")
  zscoredt_melt$color <- rev(plotlyScale)[as.numeric(zscoredt_melt$cl)+1]
  ggplot(zscoredt_melt,aes(10^MW,value,group=id,color=cl))+geom_line(lwd=1)+  
    annotate("segment",x=10^CPP$MW_lm_gel_lo,xend=10^CPP$MW_lm_gel_hi,y=CPP$Height,yend = CPP$Height,cex=1,color="grey")+
    annotate("text",10^CPP$MW_lm,CPP$Height,label="*",cex=15,color="red")
  
  
  
  
}
```

## SimilarityTables across all Gels

-   Enrichment analysis dhyper of peptides enriched in all Uniprot entries for a gene


```{r,cache=F}
ProteinSequences <- PacBIO
ipr_PeakIdentity <- Results_CombinedGels$Compressed_Pairs[AboveZscore==TRUE]
# Start of analysis ------------- 
outfilestr <- "PacBio_Enrichment"
rerun <- T
rdaName <- "ENSG_Similarity_AllGelsCombined_PacBioEnrichment.rda"
rdaName <- "ENSG_Similarity_AllGelsCombined_PacBioEnrichment_woSemiTryptic.rda"
if(!file.exists(rdaName)|rerun){
  unlink(outfilestr)
  gc()
  
  hdbscanList <- lapply(TableListNormEnsHdbScanMW_Peaks,function(x){gel = x$xname;x$hdbscanTables$gel <- gel;x$hdbscanTables})
  hdbscan <- rbindlist(hdbscanList,fill=T)
  table(hdbscan$gel)
  threads <- 50
  cl <- makeCluster(threads,outfile=outfilestr)
  input <- ipr_PeakIdentity[AboveZscore=="TRUE"]
  ENSG <- unique(ipr_PeakIdentity$ENS)
  zscoreCutoff <- 1
  
  ProteinSequencesCategories <- ProteinSequences
  clusterExport(cl,c("ProteinSequences","ProteinSequencesCategories","input","zscoreCutoff","hdbscan","traces_info_crawler_compressed_gels","nw_plot","NetworkEnrichment","nw_assembler_compressed_gels","ENSG","hyper_equal_or_more"))
  
ENSG_Similarity_PacBIO <- parLapply(cl,ENSG,Network_DB_Enrichment_Wrapper,UseRandomPeak=F,RemoveFalseRandoms=F)
  if(0){
    ENSG_Similarity_PacBIO <-   ENSG_Similarity <- parLapply(cl,ENSG,function(ensVec){
    library(igraph)
    library(data.table)
    library(visNetwork)
    Randomize <- F
    seedInitSetList <- c(1234,0000,4321,9786,6789,2341)
    seedIT <<- 0
    FVec <- c(F,T,T,T,T,T)
    EnrichmentResults <<- lapply(1:length(FVec),function(itrandom){
         Randomize <<- FVec[itrandom]
      set.seed(NULL)
      seedIT <<- seedIT+1

      try(seedInitSet <- seedInitSetList[seedIT],silent = T)
      if(is.na(seedInitSet)){
        seedInitSet <- NULL
      }
 
      Randomize <<- Randomize
      v <- NULL
      {
        traces_info <- traces_info_crawler_compressed_gels(input,ensVec,mpwCut = min(zscoreCutoff),random = F)
        traces_info <- traces_info[PeakfinderSet=="Cl_2"]
        NW_list <- nw_assembler_compressed_gels(input = input[PeakfinderSet=="Cl_2"],hdbscan = hdbscan,traces_info = traces_info,EnsID = ensVec,applyCutoffs = T,ProteinSequences = ProteinSequencesCategories[Canonical_in_gff==FALSE],random = F)
  
        temp <- NW_list$AssembledTable
  
  
        NW_list$edges <- unique(NW_list$edges)
        NW_list$AssembledTable <- unique(NW_list$AssembledTable)
  
  
  
        if(length(NW_list)==0){
          print(paste("No Entry for",ensVec))
          return("NoEntry")
        }
        options(warn=-1)
        v <- nw_plot(NW_list,type = "igraph")

        OriNames <- attributes(V(v$v))$name
        ReplNames <- gsub("|","_",OriNames,fixed = T)
  
        V(v$v)$name <- ReplNames

        e <- NetworkEnrichment(nw = simplify(v$v),NW_list = NW_list,check="MQ",
                               ctrl="UniProt",wholetable = F,Randomize = Randomize,seedInit=seedInitSet
        )
        e$seed <- seedInitSet
        v$enrichmentTable <- e
  
        v$simiList[,names(e):={
          gr <- .BY
          sel <-
            (gr$I1==e$IsoFrac|gr$I2==e$IsoFrac)&
            (gr$I1==e$Uniprot|gr$I2==e$Uniprot)&
            as.character(e$IsoFrac)!=as.character(e$Uniprot)
  
          if(length(sel)>0){
            ef <- e[sel,]
            ef <- ef[ef$dhyper_p==min(ef$dhyper_p,na.rm = T),]
            ef <- ef[ef$Success_Uniprot_Peptides==max(ef$Success_Uniprot_Peptides,na.rm = T),]
            ef <- ef[ef$UniprotOverlap==max(ef$UniprotOverlap,na.rm = T),]
            ef <- ef[ef$IsofracOverlap==max(ef$IsofracOverlap,na.rm = T),]
  
  
            ef <- ef[!is.na(ef$dhyper_p),]
            ef <- ef[1,]
          }else{
            print("no match")
            ef <- NULL
          }
        },.(I1,I2)]
        v$IsofracInfo <- NW_list$AssembledTable
        v$v <- simplify(v$v)
  
        v}
  
    })
  
    options(warn=1)

    names(EnrichmentResults) <- FVec
    EnrichmentResults <<- EnrichmentResults
    fw <- EnrichmentResults[names(EnrichmentResults)=="FALSE"][[1]]
    ra <- EnrichmentResults[names(EnrichmentResults)=="TRUE"]
  
    list(fw=fw,rev=ra)
  
  })
  }


  
  names(ENSG_Similarity_PacBIO) <- ENSG
  plot(ENSG_Similarity_PacBIO[[1]]$fw$v)
  
  stopCluster(cl)

  teno <- ENSG_Similarity_PacBIO$ENSG00000026025$fw$simiList
  
  save(ENSG_Similarity_PacBIO,file=rdaName)
}else{
  load(rdaName)
}


```

## Pairs Unfiltered

```{r}

library(h2o)
h2o.init()

# rerun <-T

ENSG_Similarity_PacBIO <- lapply(ENSG_Similarity_PacBIO,function(y){
  y <<- y
  
  x <- y$rev
  library(data.table)
  Prepare_EnrichmentTable <- function(x){
    x$enrichmentTable$IsoFrac_MW <- gsub(".*.# ","",x$enrichmentTable$IsoFrac)
    x$enrichmentTable$IsoFrac_MW <- gsub(" .*.","",x$enrichmentTable$IsoFrac_MW)
    x$enrichmentTable$IsoFrac_MW <- as.numeric(x$enrichmentTable$IsoFrac_MW)
    
    x$enrichmentTable$Uniprot_MW <- gsub(".*.# ","",x$enrichmentTable$Uniprot)
    x$enrichmentTable$Uniprot_MW <- gsub(" .*.","",x$enrichmentTable$Uniprot_MW)
    x$enrichmentTable$Uniprot_MW <- as.numeric(x$enrichmentTable$Uniprot_MW)  
    x$enrichmentTable$DeltaMass <- x$enrichmentTable$Uniprot_MW-x$enrichmentTable$IsoFrac_MW
    
    IF <- x$IsofracInfo
    if(length(IF$ENS)==0){
      IF$ENS <- IF$EnsG
    }
    IF[,c("nClusters","nGels"):={
      cl <- strsplit(.BY$clusters,";")
      gel <- strsplit(.BY$gels,";")
      temp <- unique(cbind(unlist(cl),unlist(gel)))
      list(length(unique(temp[,1])),length(unique(temp[,2])))
    },.(clusters,gels,ENS)]
    x$enrichmentTable <- cbind(x$enrichmentTable,IF[match(x$enrichmentTable$IsoFrac,IF$LabelID),.SD,.SDcols=c("nClusters","nGels","gel")])
    
    x
    
  }
  
  Prepare_EnrichmentTable <- function(x){
    x$enrichmentTable$IsoFrac_MW <- gsub(".*.# ","",x$enrichmentTable$IsoFrac)
    x$enrichmentTable$IsoFrac_MW <- gsub(" .*.","",x$enrichmentTable$IsoFrac_MW)
    x$enrichmentTable$IsoFrac_MW <- as.numeric(x$enrichmentTable$IsoFrac_MW)
    
    x$enrichmentTable$Uniprot_MW <- gsub(".*.# ","",x$enrichmentTable$Uniprot)
    x$enrichmentTable$Uniprot_MW <- gsub(" .*.","",x$enrichmentTable$Uniprot_MW)
    x$enrichmentTable$Uniprot_MW <- as.numeric(x$enrichmentTable$Uniprot_MW)  
    x$enrichmentTable$DeltaMass <- x$enrichmentTable$Uniprot_MW-x$enrichmentTable$IsoFrac_MW
    
    IF <- x$IsofracInfo
      if(length(IF$ENS)==0){
      IF$ENS <- IF$EnsG
    }
    IF[,c("nClusters","nGels"):={
      cl <- strsplit(.BY$clusters,";")
      gel <- strsplit(.BY$gels,";")
      temp <- unique(cbind(unlist(cl),unlist(gel)))
      list(length(unique(temp[,1])),length(unique(temp[,2])))
    },.(clusters,gels,ENS)]
   x$enrichmentTable <- cbind(x$enrichmentTable,IF[match(x$enrichmentTable$IsoFrac,IF$LabelID),.SD,.SDcols=c("nClusters","nGels","gel","heights","Height","Height_lm","Fraction_sd_80")])
    
   x
    
  }

  
  if(length(y$fw)>1&all(y$fw!="NoEntry")){
    y$fw <-Prepare_EnrichmentTable(y$fw)
  }
  if(length(y$rev)>1&all(y$rev!="NoEntry")){
    
    
    
    try({
    y$rev$enrichmentTable <- NULL
    y$rev <- lapply(y$rev,function(x){Prepare_EnrichmentTable(x)})
    et <- lapply(y$rev,function(x){x$enrichmentTable})
    et <- rbindlist(et,fill=T)
    if(length(et$V1)>0){
          et$V1 <- NULL
    }
    et <- et[!is.na(nClusters )]
    if(length(et$seed )>0){
    et$seed <- NULL
    }
    y$rev$enrichmentTable <- unique(et)
    })
    
  }
  
  y
  
})


it <- 5
x <- ENSG_Similarity_PacBIO
print(it)
er <- try({
  ensg <- names(x)
  fw <- lapply(1:length(x),function(i){
    i <- i
    y <- x[[i]]
    if(length(y$fw)>1){
      hu <- y$fw$enrichmentTable;
      hu$ENSG <- ensg[[i]]
    }else{hu <- NULL}
    
    hu
  })
  fw <- rbindlist(fw[lengths(fw)>1],fill=T)
  
  rv <- lapply(1:length(x),function(i){
    y <- x[[i]]
    if(length(y$rev)>1){
      
      hu <- y$rev$enrichmentTable;
      
      hu$ENSG <- ensg[[i]]
    }else{
      hu <- NULL
    }
    hu
  })
  rv <- rbindlist(rv[lengths(rv)>1],fill=T)
  rv$IsoFrac_MW <- gsub(".*.# ","",rv$IsoFrac)
  rv$IsoFrac_MW <- gsub(" .*.","",rv$IsoFrac_MW)
  rv$IsoFrac_MW <- as.numeric(rv$IsoFrac_MW)
  
  rv$Uniprot_MW <- gsub(".*.# ","",rv$Uniprot)
  rv$Uniprot_MW <- gsub(" .*.","",rv$Uniprot_MW)
  rv$Uniprot_MW <- as.numeric(rv$Uniprot_MW)  
  rv$DeltaMass <- 1
  fw$type <- "fw"
  rv$type <- "rv"
  # rv$DeltaMass_abs <- NA
  setdiff(names(rv),names(fw))
  huPacBio <- rbind(fw,rv,fill=T)
  huPacBio$gel_MainPeak <- huPacBio$gel
  huPacBio$gel <- it
  tabiAllGels_PacBio <- huPacBio
  
})
tabiAllGels_PacBio$DeltaMass <- tabiAllGels_PacBio$IsoFrac_MW-as.numeric(tabiAllGels_PacBio$Uniprot_MW)
tabiAllGels_PacBio$QuotientMass <- tabiAllGels_PacBio$IsoFrac_MW/as.numeric(tabiAllGels_PacBio$Uniprot_MW)

tabiAllGels_PacBio$MultiMapping <- c("One candidate","More candidates (Median)")[as.numeric( grepl(";",tabiAllGels_PacBio$Uniprot))+1]

tabiAllGels_PacBio$IsoFrac_MW_round <- round(tabiAllGels_PacBio$IsoFrac_MW,-4)
tabiAllGels_PacBio$Uniprot_MW_round <- round(tabiAllGels_PacBio$Uniprot_MW,-4)
Results_CombinedGels_PacBio$Pairs_Unfiltered_PacBio <- tabiAllGels_PacBio

```

## AI Comparison:
```{r}
AddTotalUniprotFraction <- function(x){
   x$TotalUniprotFraction <- as.numeric(x$All_Uniprot_Peptides)/as.numeric(x$all_sequences)
   x
}
Results_CombinedGels_PacBio$Pairs_Unfiltered_PacBio <-    AddTotalUniprotFraction(Results_CombinedGels_PacBio$Pairs_Unfiltered_PacBio)
Results_CombinedGels$Pairs_Unfiltered <-    AddTotalUniprotFraction(Results_CombinedGels$Pairs_Unfiltered)

library(h2o)
a <- Results_CombinedGels_PacBio$Pairs_Unfiltered_PacBio
a$DataType <- "PB"
b <- Results_CombinedGels$Pairs_Unfiltered
b$DataType <- "UP"
ab <- rbind(a,b,fill=T)

Results_CombinedGels_PacBio$Pairs_Unfiltered_PacBio$DataType <- "PB"
Results_CombinedGels$Pairs_Unfiltered$DataType <- "UP"
h2o.init()
unlink("AI_PACBIO_comparison.rda")
rdaname <- "AI_PACBIO_comparison.rda"
rdaname <- "AI_PACBIO_comparison_woSemitryptic.rda"

if(!file.exists(rdaname)|rerun){
  Resultfun <- lapply(list(PB=Results_CombinedGels_PacBio$Pairs_Unfiltered_PacBio,UP=Results_CombinedGels$Pairs_Unfiltered,UPPB=rbind(ab)),function(tabiAllGels){
    tabiAllGels$IsoFrac_MW_round <- round(tabiAllGels$IsoFrac_MW,-4)
    tabiAllGels$Uniprot_MW_round <- round(tabiAllGels$Uniprot_MW,-4)
    
    FeatureTable <- tabiAllGels[,.SD,.SDcols=c("dhyper_p","IsofracOverlap","TotalUniprotFraction","UniprotOverlap","all_sequences","type","IsoFrac_MW_round")]
    FeatureTable <- tabiAllGels[,.SD,.SDcols=c("dhyper_p","IsofracOverlap","UniprotOverlap","type")]

    temp <- tabiAllGels[ENSG=="ENSG00000189079"]
   
    FeatureTable$type <- factor(FeatureTable$type,c("fw","rv"),c(1,0))
    FeatureTable <- apply(FeatureTable,2,as.numeric)
    FeatureTable <- data.table(FeatureTable)
    print("FeatureTabling")
    h2o.FeatureTable <- as.h2o(FeatureTable)
        print("Building Model RF,GB and DL")

    md_rf<- h2o.randomForest(training_frame = h2o.FeatureTable,y="type",seed = 1234,nfolds = 10)
    md_gb<- h2o.gbm(training_frame = h2o.FeatureTable,y="type",seed = 1234,nfolds = 10)
    md_dl<- h2o.deeplearning(training_frame = h2o.FeatureTable,y="type",seed = 1234,nfolds = 10)
        print("SAVING Model RF,GB and DL")

    h2o.save_mojo(md_rf,paste(dim(tabiAllGels)[1],"rf_test",sep="_"))
    h2o.save_mojo(md_gb,paste(dim(tabiAllGels)[1],"gb_test",sep="_"))
    h2o.save_mojo(md_dl,paste(dim(tabiAllGels)[1],"dl_test",sep="_"))
    varimp_heatmap <- h2o.varimp_heatmap(list(md_rf,md_gb,md_dl))
    varimp <- lapply(list(md_rf,md_gb,md_dl),h2o.varimp)
    # md_svn<- h2o.modelSelection(training_frame = h2o.FeatureTable,y="type",seed = 1234)
    
    # h2o.save_mojo(md_rf,"CombinedGels_UniProtAssignmentModel")
    print("Predicting RegressionScores")
    score_rf <- list(predict=as.double(NA))
    score_gb <- list(predict=as.double(NA))
    score_dl <- list(predict=as.double(NA))
    try({score_rf <- as.data.frame(predict(md_rf,h2o.FeatureTable))})
    try({score_gb <- as.data.frame(predict(md_gb,h2o.FeatureTable))})
    try({score_dl <- as.data.frame(predict(md_dl,h2o.FeatureTable))})
    
    hm <- h2o.performance(md_rf,h2o.FeatureTable)
    tabiAllGels$scoreTemp <- score_rf$predict
    tabiAllGels$scoreGB <- score_gb$predict
    tabiAllGels$scoreDL <- score_dl$predict
    
    # h2o.varimp_plot(md_rf)
    temp <- tabiAllGels[ENSG=="ENSG00000189079"&type=="fw"&grepl("20",IsoFrac)]
    temp[scoreTemp==max(scoreTemp)]
    tabiAllGelsFiltered <- tabiAllGels[,{
      .SD[scoreTemp==max(scoreTemp)]
    },.(type,ENSG,IsoFrac)]
    tabiAllGelsFiltered_GB <- tabiAllGels[,{
      .SD[scoreGB==max(scoreGB)]
    },.(type,ENSG,IsoFrac)]
    tabiAllGelsFiltered_DL <- tabiAllGels[,{
      .SD[scoreDL==max(scoreDL)]
    },.(type,ENSG,IsoFrac)]
    
    tabiAllGels$ScoreType <- "all"
    tabiAllGelsFiltered$ScoreType <- "RF"
    tabiAllGelsFiltered_GB$ScoreType <- "GB"
    tabiAllGelsFiltered_DL$ScoreType <- "DL"
    
    temp <- rbind(tabiAllGels,tabiAllGelsFiltered,tabiAllGelsFiltered_GB,tabiAllGelsFiltered_DL)
    list(g=ggplot(temp,aes((as.numeric(IsoFrac_MW)-as.numeric(Uniprot_MW)),color=ScoreType))+geom_density(lwd=2)+xlim(c(-50000,50000)),
         table=temp,varimp=varimp,varimp_plot=varimp_heatmap)
  })
  names(Resultfun) <- c("PB","UP","UPPB")
  
  Resultfun <- lapply(Resultfun,function(x){x$table <- unique(x$table[type=="fw"]);x})
  Resultfun <- lapply(Resultfun,function(x){
    # x <<- x
    x$tablemin <- x$table[,.SD[DeltaMass==min(DeltaMass)],.(ENSG,DataType,ScoreType,IsoFrac)]
    x
  })
  save(Resultfun,file=rdaname)
}else{
  load(rdaname)#Resultfun
}

### Density Plot
Resultfung <- lapply(1:length(Resultfun),function(x){
  temp <- Resultfun[[x]]$table
  na <- names(Resultfun)[x]
  names(temp) <- make.unique(names(temp))
  if(any(names(temp)=="DataType")){
    ggplot(temp,aes((as.numeric(IsoFrac_MW)-as.numeric(Uniprot_MW)),color=ScoreType))+geom_freqpoly(lwd=0.5,bins=10000)+
      coord_cartesian(xlim=c(-100000,100000))+facet_wrap("DataType")+xlab("Delta Mass IsoFrac - DB")+geom_vline(xintercept =0, lty = "dotted")
    
  }else{
    ggplot(temp,aes((as.numeric(IsoFrac_MW)-as.numeric(Uniprot_MW)),color=ScoreType))+geom_freqpoly(lwd=0.5,bins=10000)+coord_cartesian(xlim=c(-100000,100000))+ggtitle(na)
    
  }

})
Resultfung_min <- lapply(1:length(Resultfun),function(x){
  temp <- Resultfun[[x]]$tablemin
  na <- names(Resultfun)[x]
  names(temp) <- make.unique(names(temp))
  if(any(names(temp)=="DataType")){
    ggplot(temp,aes((as.numeric(IsoFrac_MW)-as.numeric(Uniprot_MW)),color=ScoreType))+geom_freqpoly(lwd=0.5,bins=10000)+
      coord_cartesian(xlim=c(-100000,100000))+facet_wrap("DataType")+xlab("Delta Mass IsoFrac - DB")+geom_vline(xintercept =0, lty = "dotted")
    
  }else{
    ggplot(temp,aes((as.numeric(IsoFrac_MW)-as.numeric(Uniprot_MW)),color=ScoreType))+geom_freqpoly(lwd=0.5,bins=10000)+coord_cartesian(xlim=c(-100000,100000))+ggtitle(na)
    
  }
  # ggplot(temp,aes((as.numeric(IsoFrac_MW)-as.numeric(Uniprot_MW)),color=ScoreType))+geom_freqpoly(lwd=0.5,bins=10000)+coord_cartesian(xlim=c(-100000,100000))+ggtitle(na)
  
})
Resultfung2_min <- lapply(1:length(Resultfun),function(x){
  temp <- Resultfun[[x]]$tablemin
  na <- names(Resultfun)[x]
  names(temp) <- make.unique(names(temp))
  if(any(names(temp)=="DataType")){
    ggplot(temp,aes((as.numeric(IsoFrac_MW)-as.numeric(Uniprot_MW)),color=ScoreType))+geom_density(lwd=0.5,bw=0.05)+
      coord_cartesian(xlim=c(-100000,100000))+facet_wrap("DataType")+xlab("Delta Mass IsoFrac - DB")+geom_vline(xintercept =0, lty = "dotted")
    
  }else{
    ggplot(temp,aes((as.numeric(IsoFrac_MW)-as.numeric(Uniprot_MW)),color=ScoreType))+geom_density(lwd=0.5,bw=0.05)+coord_cartesian(xlim=c(-100000,100000))+ggtitle(na)
    
  }

})
### Poly Plot
Resultfung2 <- lapply(1:length(Resultfun),function(x){
  temp <- Resultfun[[x]]$table
  na <- names(Resultfun)[x]
  names(temp) <- make.unique(names(temp))
  print(names(temp))
  if(any(names(temp)=="DataType")){
    ggplot(temp,aes((as.numeric(IsoFrac_MW)-as.numeric(Uniprot_MW)),color=ScoreType))+geom_density(lwd=0.5,bw=0.05)+
      coord_cartesian(xlim=c(-100000,100000))+facet_wrap("DataType")+xlab("Delta Mass IsoFrac - DB")+geom_vline(xintercept =0, lty = "dotted")
    
  }else{
    ggplot(temp,aes((as.numeric(IsoFrac_MW)-as.numeric(Uniprot_MW)),color=ScoreType))+geom_density(lwd=0.5,bw=0.05)+coord_cartesian(xlim=c(-100000,100000))+ggtitle(na)
    
  }
  
})
### CountPlot
CountPLots <- lapply(1:length(Resultfun),function(x){
  temp <- Resultfun[[x]]$table
  na <- names(Resultfun)[x]
  names(temp) <- make.unique(names(temp))
  count <- temp[,{dim(.SD)[1]},.(ENSG,IsoFrac,DataType,ScoreType)]
  countdt <- count[,{
    td <- table(V1);
  list(n=as.numeric(names(td)),count=as.numeric(td))
  },.(DataType,ScoreType)]
  ggplot(countdt,aes(as.numeric(n),count,color=DataType))+geom_point()+facet_wrap("ScoreType",nrow=1)
  
  
})
CountPLots_min <- lapply(1:length(Resultfun),function(x){
  print(x)
  temp <- Resultfun[[x]]$tablemin
  na <- names(Resultfun)[x]
  names(temp) <- make.unique(names(temp))
  count <- temp[,{
    hu <- .SD
    hu$V1 <- NULL
    dim(unique(hu))[1]
    },.(ENSG,IsoFrac,DataType,ScoreType)]
  countdt <- count[,{
    td <- table(V1);
    list(n=as.numeric(names(td)),count=as.numeric(td))
  },.(DataType,ScoreType)]
  ggplot(countdt,aes(as.numeric(n),count,color=DataType))+geom_point(size=1)+ggtitle("MinDelta")+facet_wrap("ScoreType",nrow=1)+ylim(c(0,22000))
  
  
})


Resultfung$ncol=3
Resultfung2$ncol=3
Resultfung_min$ncol=3
Resultfung2_min$ncol=3
CountPLots$ncol=3
CountPLots_min$ncol = 3
pdf("DM_ML.pdf",width=10,height=5)
do.call(plot_grid,Resultfung2)
do.call(plot_grid,Resultfung)
do.call(plot_grid,CountPLots)

do.call(plot_grid,Resultfung2_min)
do.call(plot_grid,Resultfung_min)
do.call(plot_grid,CountPLots_min)


dev.off()
system("open DM_ML.pdf")


```


## Trains Based on gbm Model

```{r}
library(h2o)
h2o.init()
# loading model
if(!exists("gbmodel_UP")){
  gbmodel_UP <- h2o.import_mojo("./CombinedGels_UniProtAssignmentModelgbm/") # Has been assigned to Protein
}

AddTotalUniprotFraction <- function(x){
   x$TotalUniprotFraction <- as.numeric(x$All_Uniprot_Peptides)/as.numeric(x$all_sequences)
   x
}

Results_CombinedGels_PacBio$Pairs_Unfiltered_PacBio <-    AddTotalUniprotFraction(Results_CombinedGels_PacBio$Pairs_Unfiltered_PacBio)
Results_CombinedGels$Pairs_Unfiltered <-    AddTotalUniprotFraction(Results_CombinedGels$Pairs_Unfiltered)

PredictScores <- function(x,md){
  x <<- x
  names(x) <- make.unique(names(x))
  x$V1 <- NULL
  xh <- as.h2o(x)
  p <- predict(md,xh)
  as.data.frame(p)$predict
}

Results_CombinedGels$Pairs_Unfiltered_combi <- Results_CombinedGels$Pairs_Unfiltered
Results_CombinedGels$Pairs_Unfiltered_combi$score_ref_gbm <- NULL
Results_CombinedGels$Pairs_Unfiltered_combi$score_ref <- NULL

Results_CombinedGels$Pairs_Unfiltered_combi$score_UP_gb <- PredictScores(Results_CombinedGels$Pairs_Unfiltered,gbmodel_UP)

Results_CombinedGels_PacBio$Pairs_Unfiltered_PacBio$score_ref <- NULL

Results_CombinedGels_PacBio$Pairs_Unfiltered_PacBio$score_UP_gb <- PredictScores(Results_CombinedGels_PacBio$Pairs_Unfiltered_PacBio,gbmodel_UP)

ReferenceScore <- "GBM"
if(ReferenceScore=="GBM"){
  Results_CombinedGels_PacBio$Pairs_Unfiltered_PacBio$score_ref <- Results_CombinedGels_PacBio$Pairs_Unfiltered_PacBio$score_UP_gb
Results_CombinedGels$Pairs_Unfiltered$score_ref <- Results_CombinedGels$Pairs_Unfiltered$score_ref_gbm
}

Results_CombinedGels$Pairs_Unfiltered_combi$DataType <- "UP"
Results_CombinedGels_PacBio$Pairs_Unfiltered_PacBio$DataType <- "PB"

CombiScoreTable <- rbind(Results_CombinedGels_PacBio$Pairs_Unfiltered_PacBio,Results_CombinedGels$Pairs_Unfiltered_combi,fill=T)
table(CombiScoreTable$DataType)
save(Results_CombinedGels,file="Results_CombinedGels.rda")
save(Results_CombinedGels_PacBio,file="Results_CombinedGels_PacBio.rda")

```

## Filtering original

```{r}
BestMatch <- function(x,dt_gn_uni=dt_gn_uni,scoreType="score_ref"){
    x <<- x
    scoreType <<- scoreType
    x$id <- 1:dim(x)[1]
    
    scoreSel <- which(names(x)==scoreType)
    if(length(scoreSel)==0){
      print("NO selected in Score Column")
      return(NULL)
    }
    x$selectedScore <- as.data.frame(x)[,scoreSel]
    
    BestMatchx <- x[,{
  cat("\r",.GRP)
  temp <- .SD
  temp$score_ref <- as.numeric(temp$selectedScore)
  tempResults <- temp[selectedScore==max(selectedScore)]
  tempResults_DM <- temp[abs(DeltaMass)==min(abs(DeltaMass))]
  tempResults_random <- temp[sample(1:dim(temp)[1],1)]
  tempResults_random2 <- temp[sample(1:dim(temp)[1],1)]
  tempResults_random3 <- temp[sample(1:dim(temp)[1],1)]
  
  if(any(tempResults$Uniprot==unique(tempResults_DM$Uniprot))){
    Category <- "Score_equals_DM"
  }
  if(all(tempResults$Uniprot!=unique(tempResults_DM$Uniprot))){
    Category <- "Score_unequals_DM"
  }
  # up <- tempResults$Uniprot
  if(is.vector(tempResults)){
    tempResults <- as.data.table(as.list(tempResults))
  }else{
    tempResults <- as.data.table(tempResults)
    
  }
  if(is.vector(tempResults_DM)){
    tempResults_DM <- as.data.table(as.list(tempResults_DM))
  }else{
    tempResults_DM <- as.data.table(tempResults_DM)
  }
  # tempResults$Uniprot <- up
  tempResults$MultiMapping <- NULL
  tempResults_DM$MultiMapping <- NULL
  tempResults$ScoreFilteredEntries <- dim(tempResults)[1]
  tempResults_DM$ScoreFilteredEntries <- dim(tempResults_DM)[1]
  
  tempResults$Category <- Category
  tempResults_DM <- apply(tempResults_DM,2,function(x){paste(unique(x),collapse=";")})
  tempResults_DM <- as.list(tempResults_DM)
  tempResults$DMFiltered_DeltaMass <- tempResults_DM$DeltaMass
  tempResults$DMFiltered_Uniprot <- tempResults_DM$Uniprot
  tempResults$DMFiltered_UniprotOverlap <- tempResults_DM$UniprotOverlap
  tempResults$DMFiltered_IsofracOverlap <- tempResults_DM$IsofracOverlap
  tempResults$Random_DeltaMass_1 <- tempResults_random$DeltaMass
  tempResults$Random_DeltaMass_2 <- tempResults_random2$DeltaMass
  tempResults$Random_DeltaMass_3 <- tempResults_random3$DeltaMass
  
  tempResults <- apply(tempResults,2,function(x){paste(unique(x),collapse=";")})
  as.list(tempResults)
},.(gel,type,ENSG,IsoFrac)]

    BestMatchx$ScFiltered_DeltaMass <- BestMatchx$DeltaMass
    BestMatchx$ScFiltered_Uniprot <- BestMatchx$Uniprot
    BestMatchx$ScFiltered_UniprotOverlap <- BestMatchx$UniprotOverlap
    BestMatchx$ScFiltered_IsofracOverlap <- BestMatchx$IsofracOverlap
    
    # FinalSelectedTable_rf_allgels[ENSG=="ENSG00000026025"]
    BestMatchx$hgnc_symbol <- dt_gn_uni$hgnc_symbol[match(BestMatchx$ENSG,dt_gn_uni$ens)]
    BestMatchx$Use <- dt_gn_uni$Use[match(BestMatchx$ENSG,dt_gn_uni$ens)]
    BestMatchx$ScoreType <- scoreType
    
    
    ##### Defining Identificationstatus
    BestMatchx$IdentificationStatus  <- "Undefined"
    # FinalSelectedTable_rf_allgels[Uniprot==DMFiltered_Uniprot]$IdentificationStatus <- "sc_and_dm"
    BestMatchx[selectedScore>=0.75]$IdentificationStatus <- "Good"
    BestMatchx[selectedScore>0.5&selectedScore<0.75]$IdentificationStatus <- "Mediocre"
    BestMatchx[selectedScore<0.5]$IdentificationStatus <- "Unsufficient"
    table(BestMatchx$IdentificationStatus)
    
    ######
    BestMatchx[,c("ScFiltered_DM_min"):={
  temp <- .BY
  UniFun <- as.numeric(unique(unlist(strsplit(temp$Uniprot_MW,";"))))
  Iso <- as.numeric(temp$IsoFrac_MW)
  Diffis <- Iso-UniFun
  Quoties <- log10(Iso/UniFun)
  QuotiesDiff <- diff(range(Quoties,na.rm=T))
  QuotiesMin <- Quoties[abs(Quoties)==min(abs(Quoties),na.rm=T)]
  QuotiesAV <- median(Quoties,na.rm=T)
  rangeDiff <- diff(range(Diffis,na.rm=T))
  AverageMass <- median(Diffis,na.rm=T)
  MinMass <- Diffis[abs(Diffis)==min(abs(Diffis),na.rm=T)]
  .(min_DM=MinMass)
},.(Uniprot_MW,Uniprot,IsoFrac,IsoFrac_MW,score_ref,type)]

    BestMatchx$DM_min <- BestMatchx$ScFiltered_DM_min
    # Final Filtering Decision

    BestMatchx$IdentificationStrategy <- "SequenceBased"
    BestMatchx <- BestMatchx[,{
      label <- "Undefined"
      gr <- .BY
      temp <- .SD
      gr <- gr
      gr$DM_min <- as.numeric(gr$DM_min)
      gr$DMFiltered_DeltaMass <- as.numeric(gr$DMFiltered_DeltaMass)
      
      if(gr$ScFiltered_Uniprot==gr$DMFiltered_Uniprot){
        label <- "Single Candidate"
      }else{
        if(abs(abs(gr$ScFiltered_DM_min)-abs(gr$DMFiltered_DeltaMass))<5000){
          label <- "SequenceBased, >5kD<"
        }else{
          if(gr$score_ref>0.5){
            if(gr$ScFiltered_DM_min>0){
              label <- "SequenceBased, DM>5kD"
              
            }
            if(gr$ScFiltered_DM_min<0){
              label <- "SequenceBased, DM<5kD"
              
            }
          }else{
              label <- "DM_Based"
              temp$Uniprot <- gr$DMFiltered_Uniprot
              temp$DeltaMass <- (gr$DMFiltered_DeltaMass)
              temp$DM_min <- (gr$DMFiltered_DeltaMass)
              
              temp$Uniprot_MW <- (gsub(" Da","",gsub(".*. # ","",gr$DMFiltered_Uniprot)))
    
            
      
          }
        }
        
      }
      temp$DeltaMass <- as.numeric(temp$DeltaMass)
      temp$DM_min <- as.numeric(temp$DM_min)
      temp$Uniprot_MW <- as.numeric(temp$Uniprot_MW)
      
      temp$label <- label
      # print(paste(.GRP,dim(temp),label))
      
      temp
      
    },.(ScFiltered_Uniprot,ScFiltered_DM_min,IdentificationStatus,Category,DMFiltered_Uniprot,DMFiltered_DeltaMass,score_ref,type)]
    table(BestMatchx$label)

    BestMatchx
  }
BestMatch_UPPB <- function(x,dt_gn_uni=dt_gn_uni,scoreType="score_ref"){
  # x <<- x
  scoreType <<- scoreType
  x$id <- 1:dim(x)[1]
  
  scoreSel <- which(names(x)==scoreType)
  if(length(scoreSel)==0){
    print("NO selected in Score Column")
    return(NULL)
  }
  x$selectedScore <- as.data.frame(x)[,scoreSel]
  # Searching Seperately for Uniprot or PACBIO
  print("Pickying Best DB specific candidate")
  BestMatchx <- x[,{
    # cat("\r",.GRP)
    temp <- .SD
    tempResults <- temp[selectedScore==max(selectedScore,na.rm=T)]
    up <- tempResults$Uniprot
    if(is.vector(tempResults)){
      tempResults <- as.data.frame(as.list(tempResults))
    }else{
      tempResults <- as.data.frame(tempResults)
      
    }
    tempResults$Uniprot <- up
    tempResults$MultiMapping <- NULL
    tempResults <- apply(tempResults,2,function(x){paste(unique(x),collapse=";")})
    as.list(tempResults)
  },.(gel,type,ENSG,IsoFrac,DataType)]
  # Picking Best Entry from PACBIO and UNIPROT
  print("Pickying Best overall candidate")
  BestMatchxCombi <- BestMatchx[type=="fw",{
    cat("\r",.GRP)
    temp <- .SD
    UPNames <- c("Uniprot","Success_ISOFRAC_Peptides_UP","Success_Uniprot_Peptides","score_ref_rf","score_UP_gb","score_ref_gb","score_ref_dl","selectedScore","Uniprot_MW")
    UPNames <- intersect(UPNames,names(temp))
    PB <- temp[DataType =="PB",.SD,.SDcols = UPNames]
    UP <- temp[DataType =="UP",.SD,.SDcols = UPNames]
    PB_single <- apply(PB,2,paste,sep=";")
    PB_single <- as.data.table(as.list(PB_single))
    UP_single <- apply(UP,2,paste,sep=";")
    UP_single <- as.data.table(as.list(UP_single))
    Best_single <- temp[as.numeric(selectedScore)==max(as.numeric(selectedScore)),{apply(.SD,2,paste,collapse="###")},]
    Best_single <- as.data.table(as.list(Best_single))[,.SD,.SDcols=c("Uniprot","Uniprot_MW","DataType","selectedScore")]
    setnames(Best_single,c("Uniprot","Uniprot_MW"),c("Best_Accession","Best_MW"))
    if(length(UP_single)==0){
      rep_c <- rep("",length(UPNames))
      names(rep_c) <- UPNames
      UP_single <- as.data.table(as.list(rep_c))
    }
    if(length(PB_single)==0){
      rep_c <- rep("",length(UPNames))
      names(rep_c) <- UPNames
      PB_single <- as.data.table(as.list(rep_c))
    }
    PBnamesOri <- 
               c("Uniprot","Success_ISOFRAC_Peptides_UP","Success_Uniprot_Peptides","score_ref_rf","score_ref_gb","score_ref_dl","selectedScore","Uniprot_MW","score_UP_gb")
    wh <- !is.na(match(PBnamesOri,names(PB_single)))
    PBnamesOriNew <- c("PACBIO_ACC","Success_ISOFRAC_Peptides_PB","Success_PB_Peptides","score_ref_rf_PB","score_ref_gb_PB","score_ref_dl_PB","selectedScore_PB","PB_MW","score_UP_gb_PB")
    
    setnames(PB_single,PBnamesOri[wh],PBnamesOriNew[wh])
    
    UPnamesOri <- 
               c("Uniprot","Success_ISOFRAC_Peptides_UP","Success_Uniprot_Peptides","score_ref_rf","score_ref_gb","score_ref_dl","selectedScore","Uniprot_MW","score_UP_gb")
    UPnamesOriNew    <-     c("Uniprot","Success_ISOFRAC_Peptides_UP","Success_Uniprot_Peptides","score_ref_rf_UP","score_ref_gb_UP","score_ref_dl_UP","selectedScore_UP","Uniprot_MW","score_UP_gb_UP")
    wh2 <- !is.na(match(UPnamesOri,names(UP_single)))

    setnames(UP_single,UPnamesOri[wh2],UPnamesOriNew[wh2])
    
    if(any(temp$DataType=="UP")){
      Info <- temp[DataType =="UP",.SD,.SDcols =  setdiff(names(temp),c(PBnamesOri,UPnamesOri))]
      Info <- unique(Info)
    }else{
      Info <- temp[DataType =="PB",.SD,.SDcols =  setdiff(names(temp),c(PBnamesOri,UPnamesOri))]
      Info <- unique(Info)
    }
    Info$Score_ref <- NULL
    
    if(dim(Info)[1]==1){
      tempResults <-       cbind(Best_single,Info,UP_single,PB_single)
      #apply(tempResults,2,function(x){paste(unique(x),collapse=";")})
      
    }else{
      
      stop("Output!=1")
    }
    tempResults
  },.(gel,type,ENSG,IsoFrac)]
  
  table(BestMatchxCombi$DataType)
  # names(BestMatchxCombi)
  table(BestMatchxCombi$DataType)
  print("Adding Best minimal Deltamass entry.")
  BestMatchxCombi[,c("DM_UP_min","DM_UP","DM_PB_min","DM_PB"):={
    DMUP <- as.numeric(.BY$IsoFrac_MW)-as.numeric(unlist(strsplit(Uniprot_MW,";")))
    DMUP_min <- DMUP[abs(DMUP)==min(abs(DMUP))]
    DMPB <- as.numeric(.BY$IsoFrac_MW)-as.numeric(unlist(strsplit(PB_MW,";")))
    DMPB_min <- DMPB[abs(DMPB)==min(abs(DMPB))]
    list(as.numeric(unique(DMUP_min)),paste(DMUP,collapse=";"),as.numeric(unique(DMPB_min)),paste(DMPB,collapse=";"))
    
  },.(IsoFrac_MW,ENSG)]
  
  # FinalSelectedTable_rf_allgels[ENSG=="ENSG00000026025"]
  BestMatchxCombi$hgnc_symbol <- dt_gn_uni$hgnc_symbol[match(BestMatchxCombi$ENSG,dt_gn_uni$ens)]
  BestMatchxCombi$Use <- dt_gn_uni$Use[match(BestMatchxCombi$ENSG,dt_gn_uni$ens)]
  BestMatchxCombi$ScoreType <- scoreType
  BestMatchxCombi
}
    cl <- makeCluster(4)
    clusterExport(cl,c("CombiScoreTable","dt_gn_uni","BestMatch_UPPB","BestMatch"))
    ScoresToAnalyze <- c("score_UP_gb")
    BestMatchList <- parLapply(cl,ScoresToAnalyze,function(x){
      library(data.table)
      print(x)
      test <-F
      if(test){
            tempi <- BestMatch_UPPB(CombiScoreTable[sample(1:dim(CombiScoreTable)[1],1000)],dt_gn_uni,scoreType =x )
            table(tempi$DataType)

      }else{
            tempi <- BestMatch_UPPB(CombiScoreTable,dt_gn_uni,scoreType =x )

      }
      tempi
    })
 
  selectedScore <- "score_UP_gb"
  Results_CombinedGels_PBUP <- list()

  Results_CombinedGels_PBUP$BestMatch_PB <- BestMatch(CombiScoreTable[DataType=="PB"],dt_gn_uni,scoreType =selectedScore )
  Results_CombinedGels_PBUP$BestMatch_UP <- BestMatch(CombiScoreTable[DataType=="UP"],dt_gn_uni,scoreType =selectedScore )

  
  names(BestMatchList) <- ScoresToAnalyze
  BestMatchdt <- rbindlist(BestMatchList)
  names(BestMatchdt) <- make.unique(names(BestMatchdt))
Results_CombinedGels_PBUP$Pairs_Unfiltered <- CombiScoreTable
Results_CombinedGels_PBUP$Pairs_Filtered_list <- BestMatchList
Results_CombinedGels_PBUP$Pairs_Filtered <- BestMatchdt[ScoreType=="score_UP_gb"]

dim(Results_CombinedGels_PBUP$Pairs_Filtered$Score )
table(BestMatchdt$DataType)/3
lapply(BestMatchList,function(x){table(x$DataType)})
# BestMatchdt$DeltaMass
BestMatchdt[,"DeltaMass_min":={
  dm <- as.numeric(unlist(strsplit(DeltaMass,";")))
  dm <- dm[abs(dm)==min(abs(dm),na.rm=T)]
  dm
},DeltaMass] 

names(BestMatchdt) <- make.unique(names(BestMatchdt))

  pdf("DM_COMBI.pdf",width=10)
  ggplot(BestMatchdt[type=="fw"],aes(as.numeric(DeltaMass_min),color=ScoreType))+geom_freqpoly(bins=1000)+coord_cartesian(xlim=c(-150000,150000))
plot_grid(ggplot(BestMatchdt[type=="fw"],aes(as.numeric(DeltaMass_min),color=ScoreType))+geom_freqpoly(bins=1000)+coord_cartesian(xlim=c(-150000,150000)),
          ggplot(BestMatchdt[type=="fw"],aes(as.numeric(DeltaMass_min),color=ScoreType))+geom_freqpoly(bins=1000))
dev.off()
    # stop()
system("open ./DM_COMBI.pdf")

save(Results_CombinedGels_PBUP,file="Results_CombinedGels_PBUP.rda")
```

## ROC Combi

-   ROC curves based on Random Forest scores

```{r ROC allGels before filtering}
BestMatchdt <- rbindlist(Results_CombinedGels_PBUP$Pairs_Filtered_list)
RocGenerator <- function(FeatureInput,SelectedColumns=c("dhyper_p","IsofracOverlap","UniprotOverlap","all_sequences","type","score_ref_PacBIO")){
  FeatureInput <<- FeatureInput
  SelectedColumns <<- SelectedColumns
  FeatureSelect <- FeatureInput[,.SD,.SDcols=SelectedColumns]
  setnames(FeatureSelect,
           c("dhyper_p","IsofracOverlap","UniprotOverlap","score_UP_gb"),
           c("Overlap","Coverage Isofrac","Coverage Uniprot","ML Score"))
  typevec <- FeatureSelect$type
  FeatureSelect$type <- NULL
  typevec <- typevec=="fw"
  it <<- 0
  rocs <- apply(FeatureSelect,2,function(x){
    it <<- it+1
   
    pos <- pROC::roc(typevec,as.numeric(x))
    
    aucs <- sapply(1:10,function(i){
      p <- as.double(NA)
      sa <- sample( 1:length(x),length(x),replace = T)
      if(length(unique(typevec[sa]))==2){
        pos <- pROC::roc(typevec[sa],as.numeric(x[sa]))
        p <- pos$auc
      }
      p
    })
    list(q=quantile(aucs,probs=c(0.05,0.5,0.95),na.rm = T),pos=pos,feature=    colnames(FeatureSelect)[it])
    
  })

  rocs_d <- rbindlist(lapply(rocs,function(x){
    data.table(Specificity=x$pos$specificities,Sensitivity=x$pos$sensitivities,feature=paste(x$feature,round(x$pos$auc,2)))
    
  }))
  
  
}

ROCSall <- lapply(c("score_UP_gb"),function(x){
  ROC_Combi <<- RocGenerator(CombiScoreTable,SelectedColumns = c("dhyper_p","IsofracOverlap","UniprotOverlap","type",x))
  ROC_Combi$ROC_TYPE <- x
 ROC_Combi
})
names(ROCSall) <- c("score_UP_gb")

x <- ROCSall$score_UP_gb
table(x$feature)
ROCplot <- function(x){
x$feature <- factor(x$feature,c(
  grep("Isofrac",(x$feature),value=T)[1],
  grep("Uniprot",(x$feature),value=T)[1],
  grep("Overlap",(x$feature),value=T)[1],
  grep("Score",(x$feature),value=T)[1]
  ))
  
  ggplot(x,aes((Specificity),(Sensitivity),color=feature))+geom_line(lwd=1)+xlim(c(1,0))+geom_abline(slope=1,intercept = 1,lty="dotted")+facet_wrap("ROC_TYPE")
}

pdf("ROCS.pdf",width=10)
do.call(plot_grid,lapply(ROCSall,ROCplot))
dev.off()
system("open ROCS.pdf")
ROCSalldt <- rbindlist(ROCSall)
library(pracma)
BP_AUC <- ROCSalldt[,{
  temp <<- .SD
  trapz(temp$Specificity,temp$Sensitivity)
},.(ROC_TYPE,feature)]
Results_CombinedGels_PBUP$ROC_BP <- list(table=ROCSalldt,gg=ggplot(BP_AUC,aes(feature,V1,fill=ROC_TYPE))+geom_bar(stat="identity",position = "dodge")+theme(axis.text.x = element_text(angle=45,hjust = 1,vjust = 1)))


ROCtable <- ROCSall$score_UP_gb[grep("all_sequences",feature,invert = T)]
ROCtable$feature <- gsub("UniprotOverlap" ,"Database Overlap",ROCtable$feature)


Results_CombinedGels_PBUP$ROC <- ggplot(ROCtable,aes((Specificity),(Sensitivity),color=feature))+geom_line()+xlim(c(1,0))+geom_abline(slope=1,intercept = 1,lty="dotted")
save(Results_CombinedGels_PBUP,file="Results_CombinedGels_PBUP.rda")
save(ROCSall,file="ROCSall.rda")

```

## Trying with orthogonal distance calculation
```{r}

UP <- Results_CombinedGels$Pairs_Filtered
PB <- Results_CombinedGels_PBUP$Pairs_Filtered

UPPB <- rbind(UP,PB,fill=T)
ggplot(UPPB[type=="fw"],aes(Category,fill=DataType))+geom_bar(stat="count",position = "dodge")
plotCollecter$PBUP_IsoFormCountCategories <- ggplot(UPPB[type=="fw"],aes(label,fill=DataType))+geom_bar(stat="count",position = "dodge")+theme(axis.text.x = element_text(angle=45,hjust=1,vjust=1))
plotCollecter$PBUP_IsoFormCountCategories$Caption <- "Protein Isoform Counts in across different Identification Categories using either PacBIO (PB) or UniProt (UP). PB substantially increases the number of single candidate hits due the database specificity."

plotCollecter$PBUP_IsoFormCountCategoriesBoxplots <-ggplot(UPPB[type=="fw"&!is.na(score_UP_gb_PB)],aes(label,as.numeric(score_UP_gb_PB),fill=DataType))+geom_boxplot()+theme(axis.text.x = element_text(angle=45,hjust=1,vjust=1))

plotCollecter$PBUP_IsoFormCountCategoriesBoxplots$Caption <- "Boxplots between PB and UP within different IsoformQuality bins. No strong differences in can observed."



RowWiseCompare <- Results_CombinedGels_PBUP$Pairs_Filtered[type=="fw"]

RowWiseCompare$Uniprot_Isoform_Category <- Results_CombinedGels$Pairs_Filtered$Uniprot_Isoform_Category[match(RowWiseCompare$Uniprot,Results_CombinedGels$Pairs_Filtered$Uniprot)] 


table(RowWiseCompare$DataType)
names(RowWiseCompare)
# RowWiseCompare$score_ref

RowWiseCompare$hgnc <- dt_gn_uni$uni_hgnc[match(RowWiseCompare$ENSG,dt_gn_uni$ens)]
RowWiseCompare$score_ref_Uniprot  <- sapply(strsplit(RowWiseCompare$score_UP_gb_UP,";"),function(x){max(as.numeric(x),na.rm=T)})
RowWiseCompare$score_ref_PacBIO  <- sapply(strsplit(RowWiseCompare$score_UP_gb_PB,";"),function(x){max(as.numeric(x),na.rm=T)})

scoreCuroff <- c(0.75,2)
dmDiff <- 10000

RowWiseCompare$CombiScore_Uniprot <- as.numeric(RowWiseCompare$score_ref_Uniprot)
RowWiseCompare$Combiscore_ref <- as.numeric(RowWiseCompare$score_ref_PacBIO)

score_Ref_cutoff <- (RowWiseCompare$CombiScore_Uniprot>scoreCuroff[1]|RowWiseCompare$Combiscore_ref>scoreCuroff[1])&
  (RowWiseCompare$CombiScore_Uniprot<scoreCuroff[2]&RowWiseCompare$CombiScore_Uniprot<scoreCuroff[2])&
  1

score_Ref_cutoff[is.na(score_Ref_cutoff)] <- F

table(score_Ref_cutoff)
# plotting Deltas:
RowWiseCompare$Label <- paste(RowWiseCompare$hgnc_symbol,round(as.numeric(RowWiseCompare$IsoFrac_MW)/1000),RowWiseCompare$Uniprot_Uniprot,RowWiseCompare$Uniprot,sep="\n")
RowWiseCompare$NFKB1 <- grepl("NFKB1|CEBPB|^GARS|DAG1|CRK|GART|^ARSB|CFDP1",RowWiseCompare$hgnc_symbol)

naorinf <- function(x){is.na(x)|is.infinite(x)}
sel <- naorinf(RowWiseCompare$DM_PB_min)|naorinf(RowWiseCompare$DM_UP_min)
y  <- RowWiseCompare$DM_UP_min[!sel]
x <- RowWiseCompare$DM_PB_min[!sel]
#add force model
x <- c(x,y,y,y,y,y)
y <- c(y,y,y,y,y,y)
library(pracma)
odifun <- odregress(x = x,y=y)

RowWiseCompare$Residuals <- NA
RowWiseCompare$Residuals[!sel] <-  odifun$resid[1:(length(x)/6)]


### Applying DM Categories



lmcheck <- function(x,y,n1,m1,n2,m2){
  y1 <- m1*x+n1
  y2 <- m2*x+n2
  yall <<- sort(c(y1,y2))
  y>y1&y<y2
  
}
a <- RowWiseCompare$DM_UP_min
b <- RowWiseCompare$DM_PB_min

windowSize <- dmDiff
fac <- 1.4
Dia <- lmcheck(a,b,-windowSize*fac,1,windowSize*fac,1)
AntiDia <- lmcheck(a,b,-windowSize*fac,-1,windowSize*fac,-1)
xequal <- lmcheck(a,b,-windowSize,0,windowSize,0)
yequal <- (a>-windowSize)&a<windowSize


DiffCategory <- paste(as.numeric(Dia),
                      as.numeric(AntiDia),
                      as.numeric(xequal),
                      as.numeric(yequal),sep="")
table(DiffCategory)

point_in_circle <- function(x, y, radius) {
  # Function from ChatGPT
  distance <- sqrt(x^2 + y^2)
  return(distance <= radius)
}
DiffCategoryNew <- "other"
WithinCircle <- point_in_circle(a,b,windowSize*1.5)
DiffCategoryNew[DiffCategory=="1000"&!WithinCircle] <- "Equal MassShift"
table(DiffCategoryNew)

DiffCategoryNew[!WithinCircle&DiffCategory=="0010"&a>0] <- "PACBIO correct+"
DiffCategoryNew[!WithinCircle&DiffCategory=="0001"&b>0] <- "UniProt correct+"
DiffCategoryNew[!WithinCircle&DiffCategory=="0010"&a<0] <- "PACBIO correct-"
DiffCategoryNew[!WithinCircle&DiffCategory=="0001"&b<0] <- "UniProt correct-"
DiffCategoryNew[is.na(DiffCategoryNew)] <- "Other"
DiffCategoryNew[WithinCircle] <- "Correct Mass"
RowWiseCompare$Categories_Uniprot_Uniprot_fixed <- NULL
RowWiseCompare$DM_min_categories <- DiffCategoryNew

######
UPsplit <- strsplit(RowWiseCompare$Uniprot,";")#[[1]]
if(!exists("UP_L_CombiScores")|0){
  # Ccl <- makeCluster(30)
  # clusterExport(cl,"ProteinSequencesCategories")
  load("ProteinSequencesCategories2.rda")
  PSCnames <- names(ProteinSequencesCategories)
  ProteinSequencesCategories <- rbind(ProteinSequencesCategories,PacBIO,fill=T)
  ProteinSequencesCategories <- ProteinSequencesCategories[,.SD,.SDcols = PSCnames]
  
  # parLapply(cl,UPsplit,function(x){})
  UP_L_CombiScores <- lapply(UPsplit,function(x){
    x <<- x
    x <- gsub(" # .*.","",x)
    xtemp <- ProteinSequencesCategories[match(x,ProteinID)]
    
    # xtemp <- xtemp[Canonical_in_gff==FALSE,]
    if(dim(xtemp)[1]>1){
      xtemp <- apply(xtemp,2,function(y){
        y <- gsub("^ ","",y)
        # combining multimappers into one entry
        if(all(!is.na(as.logical(y)))){
          # if logical take any
          y <- any(as.logical(y))
        }else{
          # if not logical, collapse unique information content
          y <- paste(unique(y),collapse=";")
        }
        y
      })
      xtemp <- data.table(xtemp)
      xtemp <- t(xtemp)
      xtemp <- as.data.table(xtemp)
      colnames(xtemp) <- colnames(ProteinSequencesCategories)
      cleanedstring <- gsub("_.*.","",lelu <- strsplit(xtemp$ID,";")[[1]])
      cleanedstring <- gsub("-.*.","",cleanedstring)
      if(length(unique(cleanedstring))==1){
        xtemp$Multimapper <- F
        xtemp$Variants <- length(cleanedstring)
        
      }else{  xtemp$Multimapper <- T
      xtemp$Variants <- length(lelu)
      }
      
      
    }else{
      xtemp$Multimapper <- F
      xtemp$Variants <- 1
      
    }
    xtemp
  })
  # stopCluster(cl)
  UP_L_CombiScoresList<- UP_L_CombiScores
  
}
UP_L_CombiScores <- lapply(UP_L_CombiScoresList,function(x){if(dim(x)[1]==0){
  x <<- x
  x <-rbind(x,"",fill=T)
}
  x})
UP_Matched <- rbindlist(UP_L_CombiScores,fill=T)
UP_Matched[AS==TRUE&(sp_canonical==TRUE|tr_canonical==TRUE)&!Multimapper]$sp_canonical <- F
UP_Matched[AS==TRUE&(sp_canonical==TRUE|tr_canonical==TRUE)&!Multimapper]$tr_canonical <- F
UP_Matched$cP <- as.logical(UP_Matched$cP)
UP_Matched$nP <- as.logical(UP_Matched$nP)
UP_Matched$M_Cleavage <- as.logical(UP_Matched$M_Cleavage)

UP_Matched$sp_canonical <- as.logical(UP_Matched$sp_canonical)
UP_Matched$tr_canonical <- as.logical(UP_Matched$tr_canonical)
UP_Matched$Multimapper <- as.logical(UP_Matched$Multimapper)
UP_Matched$Multimapper <- NULL

UP_Matched[,Processing:=cP|nP,]
UP_Matched[,Canonical:=sp_canonical|tr_canonical|M_Cleavage,]
Categories_CombiScore <- UP_Matched[,.SD,.SDcols=c("Canonical","Processing","AS")]


CatNames <- names(Categories_CombiScore)

Categories_CombiScore$Categories_Uniprot <- apply(Categories_CombiScore,1,function(x){
  x <- x
  x <- as.logical(gsub(" ","",x))
  x[is.na(x)] <- F
  paste((CatNames)[as.logical(x)],collapse="; ")
})
######

RowWiseCompare$Uniprot_Isoform_Category <- Results_CombinedGels$Pairs_Filtered$Uniprot_Isoform_Category[match(RowWiseCompare$Uniprot,Results_CombinedGels$Pairs_Filtered$Uniprot)] 
table(RowWiseCompare$Uniprot_Isoform_Category)

RowWiseCompare$Categories_Uniprot_Uniprot_fixed[RowWiseCompare$Categories_Uniprot_Uniprot_fixed=="AS; AS"] <- "AS"
RowWiseCompare$Categories_Uniprot_Uniprot_fixed[RowWiseCompare$Categories_Uniprot_Uniprot_fixed=="AS; Processing; AS"] <- "AS; Processing"
RowWiseCompare$Categories_Uniprot_Uniprot_fixed[RowWiseCompare$Categories_Uniprot_Uniprot_fixed=="Processing; AS"] <- "AS; Processing"
table(RowWiseCompare$Categories_Uniprot_Uniprot_fixed)

CombiScoreList <- list(
                       CombiScores=c("score_ref_rf_PB","score_ref_rf_UP"))

BoxplotGG_List <- lapply(1:length(CombiScoreList),function(i){
  i <<- i
  RowWiseComparedMelt <- melt(RowWiseCompare[score_Ref_cutoff],measure.vars = CombiScoreList[[i]])
  RowWiseComparedMelt$variable <- factor(RowWiseComparedMelt$variable,CombiScoreList[[i]],c("PACBIO","UniProt"))
  RowWiseComparedMelt$value <- as.numeric(RowWiseComparedMelt$value)
  ggt <- names(CombiScoreList)[i]
  cates <- unique(RowWiseComparedMelt$DM_min_categories)
  
  cates <- c(grep("UniProt",cates,value=T),grep("PACBIO",cates,value=T),"Correct Mass","Equal MassShift","Other")
  cates <- rev(cates)
  RowWiseComparedMelt$DM_min_categories <- factor(RowWiseComparedMelt$DM_min_categories,cates,cates)
  
  table(is.na(RowWiseComparedMelt$variable))
  BoxplotGG <- ggplot(RowWiseComparedMelt,aes(x=DM_min_categories,y=value,
                                              group=paste(DM_min_categories,variable),
                                              fill=variable
  ))+
    geom_boxplot(alpha=1,outlier.shape = NA)+xlab("RF Score")+labs(color="db")+ylab("")+coord_flip()+ggtitle(ggt)
})
names(BoxplotGG_List) <- names(CombiScoreList)

pdf("Test.pdf",width=10)
BoxplotGG_List$CombiScores
graphics.off()
system("open ./Test.pdf")

RowWiseCompare$UniprotIsoformType <- Results_CombinedGels$Pairs_Filtered$Uniprot_Isoform_Category[match(RowWiseCompare$Uniprot,Results_CombinedGels$Pairs_Filtered$Uniprot)]
RowWiseCompare$UniprotIsoformType[grep("Alternative [ip]",RowWiseCompare$UniprotIsoformType)]<- "Other Processes"
RowWiseCompare$UniprotIsoformType["Alternative splicing;Alternative splicing"==RowWiseCompare$UniprotIsoformType]<- "Other Processes"
table(RowWiseCompare$UniprotIsoformType)

######
# Delta Mass Plots
#####
subset_table<- RowWiseCompare[(as.numeric(selectedScore_UP)>0.8|as.numeric(selectedScore_PB)>0.8)]
gfu2 <- ggplot(subset_table,aes(DM_PB_min,DM_UP_min,color=DM_min_categories,label=paste(PACBIO_ACC,Uniprot)))+geom_point()+xlab("IsoFrac-PACBIO MW [D]")+ylab("IsoFrac-UniProt MW [D]")
gfu2_Counts <- ggplot(subset_table,aes(DM_min_categories,fill=DM_min_categories))+geom_bar(stat="count")

Results_CombinedGels_PBUP$DM_plots <- list(DM=gfu2,counts=gfu2_Counts)

#####
# Fraction Plots
#####

Results_CombinedGels_PBUP$Category_BP <- ggplot(subset_table,aes(DM_min_categories,log2(as.numeric(score_UP_gb_UP)/as.numeric(score_UP_gb_PB)),fill=DataType))+geom_boxplot(orientation="x")+ylab("UP/PB")+coord_flip()
Results_CombinedGels_PBUP$ProteinCategory_BP <- ggplot(subset_table[grep("UniProt",Categories_Uniprot_Uniprot_fixed)],aes(Categories_Uniprot_Uniprot_fixed,log2(as.numeric(score_UP_gb_UP)/as.numeric(score_UP_gb_PB))))+geom_boxplot(orientation="x")+ylab("UP/PB")+coord_flip()


####
# Database Counts
####

DB_counts <- as.data.table(table(RowWiseCompare$DataType))
RowWiseCompare[,table(UniprotIsoformType),DataType]


DB_counts <- RowWiseCompare[,{
  ta <-  table(UniprotIsoformType)
  list(Isoform_Type=names(ta),count=as.vector(ta),count_rel=as.vector(ta)/sum(as.vector(ta)))
},DataType]


DB_counts[,p:={
  
  temp <- .SD
  blackPicked <- sum(temp$count)
  AllCount <- sum(DB_counts$count)
  ps <<- temp[,{
    White <- count
    Black <- blackPicked-White
    AllWhite <- sum(DB_counts[DB_counts$Isoform_Type==.BY$Isoform_Type]$count)
    AllBlack <- AllCount-AllWhite
    hyper_equal_or_more(White,AllWhite,AllBlack,White+Black)
    
  },Isoform_Type]
  ps$V1
  
  
},.(DataType)]

DB_counts[,count_rel:=count/sum(count),DataType]

DB_counts$p.adjust <- p.adjust(DB_counts$p,"BH")
DB_counts$p.adjust[DB_counts$p.adjust>0.05] <- as.double(NA)

add_Label <- function(gg,hj=0,vj=1,angle=0,repel=T,setdirection="both"){
  if(!repel){
    gg+geom_label(aes(label=signif(p.adjust,2)),position = position_stack(vjust = vj),hjust=hj ,angle=angle,size=2,show.legend = FALSE)+
      guides(fill = guide_legend(override.aes = list(label="",angle=angle)),
             lwd=guide_legend(override.aes=list(lwd=NA)))+theme(legend.title = element_blank())
    
  }else{
    gg+geom_label_repel(aes(label=signif(p.adjust,2)),position = position_stack(vjust = vj),hjust=hj ,angle=angle,size=2,direction = setdirection,show.legend = FALSE)+
      guides(fill = guide_legend(override.aes = list(label="",angle=angle)))+theme(legend.title = element_blank())
    
  }
}

if(class(DB_counts$DataType)!="factor"){
  DB_counts$DataType <- factor(DB_counts$DataType,c("PB","PB###UP","UP"),c("PACBIO","PACBIO\nand\nUniProt","UniProt"))
  
}

Results_CombinedGels_PBUP$FractionOfIdentifed <- ggplot(DB_counts,aes(DataType,count))+geom_bar(stat="identity",position="stack")
Results_CombinedGels_PBUP$FractionOfIdentifed_rel <- add_Label(ggplot(DB_counts,aes(DataType,count_rel,fill=Isoform_Type))+geom_bar(stat="identity",position="stack"),hj=0.5,vj=c(0.5,0.5,0),repel = F)

subset_table_counts <- subset_table[score_UP_gb_PB!=score_UP_gb_UP,{
 ta <-  table(UniprotIsoformType)
 list(Isoform_Type=names(ta),count=as.vector(ta),count_rel=as.vector(ta)/sum(as.vector(ta)))
},DM_min_categories]
subset_table_countsAll <- subset_table[,{
  ta <-  table(UniprotIsoformType)
  list(Isoform_Type=names(ta),count=as.vector(ta),count_rel=as.vector(ta)/sum(as.vector(ta)))
},]
subset_table_countsAll$DM_min_categories <- "ALL"

####
# Fraction Enrichment
###
subset_table_counts[,p:={
  
  temp <- .SD
  blackPicked <- sum(temp$count)
  AllCount <- sum(subset_table_countsAll$count)
  ps <<- temp[,{
    White <- count
    Black <- blackPicked-White
    AllWhite <- subset_table_countsAll[subset_table_countsAll$Isoform_Type==.BY$Isoform_Type]$count
    AllBlack <- AllCount-AllWhite
    hyper_equal_or_more(White,AllWhite,AllBlack,White+Black)
    
  },Isoform_Type]
  ps$V1

  
},.(DM_min_categories)]


subset_table_counts$p.adjust <- p.adjust(subset_table_counts$p,"BH")
subset_table_counts$p.adjust[subset_table_counts$p.adjust>0.05] <- as.double(NA)
Results_CombinedGels_PBUP$IsoformFractions <-
  ggplot(rbind(subset_table_counts[grep("PACBIO",DM_min_categories,invert = T)],subset_table_countsAll,fill=T),
              aes(DM_min_categories   ,count_rel,fill=Isoform_Type ))+
  geom_bar(stat="identity")+ggplot2::coord_flip()+ylab("Fractions")+xlab("")+
  guides(fill = guide_legend(override.aes = list(shape = NA)))
Results_CombinedGels_PBUP$IsoformFractions_ALL <-
  ggplot(rbind(subset_table_counts,subset_table_countsAll,fill=T),
              aes(DM_min_categories   ,count_rel,fill=Isoform_Type ))+
  geom_bar(stat="identity")+ggplot2::coord_flip()+ylab("Fractions")+xlab("")+
  guides(fill = guide_legend(override.aes = list(shape = NA)))
library(ggrepel)
Results_CombinedGels_PBUP$IsoformFractions <- Results_CombinedGels_PBUP$IsoformFractions+geom_label_repel(aes(label=signif(p.adjust,2)),position = position_stack(vjust = 1),hjust=0 ,angle=90,size=2,direction = "x")+
    guides(fill = guide_legend(override.aes = list(label="")))+theme(legend.title = element_blank())
Results_CombinedGels_PBUP$IsoformFractions_ALL <- Results_CombinedGels_PBUP$IsoformFractions_ALL+geom_label_repel(aes(label=signif(p.adjust,2)),position = position_stack(vjust = 1),hjust=0 ,angle=90,size=2,direction = "x")+
    guides(fill = guide_legend(override.aes = list(label="")))+theme(legend.title = element_blank())

#####
# PacBio Categories 
#####
# Adding PacBio Categories
subset_table$PACBIO_Categories <- sapply(strsplit(subset_table$PACBIO_ACC,"ENSG"),function(x){
  x <<- x
  x <- x[x!=""]
  x <- gsub("#","",x)
  xi <- as.numeric(sapply(strsplit(x," "),"[[",2))
  pb <- PacBIO[unique(xi)]$PacBioNames
  if(any(is.na(pb))){
    stop()
    
  }
  paste(sort(unique(pb)),collapse=";")
})
table(is.na(subset_table$PACBIO_Categories))
subset_table$Multiple <- grepl(";",subset_table$PACBIO_Categories)

PacBioCat_COUNTS  <- subset_table[,{
  ta <- table(unlist(strsplit(.SD$PACBIO_Categories,";")))
  data.table(names=unlist(names(ta)),Counts=unlist(ta),Counts_rel=unlist(ta)/sum(unlist(ta)))
},.(DM_min_categories,Multiple)]

PacBioCat_COUNTSList <- lapply(c(T,F), function(MultiX){
PacBioCat_COUNTSAll <- PacBioCat_COUNTS[Multiple==MultiX,.(sum(Counts.N)),.(Counts.V1,names,Multiple)]
PacBioCat_COUNTSAll$DM_min_categories <- "ALL"

PacBioCat_COUNTS[Multiple==MultiX,p:={
  
  temp <<- .SD
  blackPicked <- sum(temp$Counts.N)
  AllCount <- sum(PacBioCat_COUNTSAll$V1)
  ps <<- temp[,{
    White <- Counts.N
    Black <- blackPicked-White
    AllWhite <- PacBioCat_COUNTSAll[PacBioCat_COUNTSAll$names==.BY$names]$V1
    AllBlack <- AllCount-AllWhite
    hyper_equal_or_more(White,AllWhite,AllBlack,White+Black)
    
  },names]
  ps$V1

  
},.(DM_min_categories,Multiple)]
PacBioCat_COUNTS$p.adjust <- p.adjust(PacBioCat_COUNTS$p,"BH")
PacBioCat_COUNTSAll$Counts.N  <- PacBioCat_COUNTSAll$V1
PacBioCat_COUNTSAll$Counts_rel.N <- PacBioCat_COUNTSAll$Counts.N/sum(PacBioCat_COUNTSAll$Counts.N)
PacBioCat_COUNTSAll$p <- as.double(NA)
rbind(PacBioCat_COUNTS,PacBioCat_COUNTSAll,fill=T)
})

PacBioCat_COUNTS <- rbindlist(PacBioCat_COUNTSList)
PacBioCat_COUNTS <- unique(PacBioCat_COUNTS)
g1 <-ggplot(unique(PacBioCat_COUNTS),aes(DM_min_categories,Counts.N,fill=names))+geom_bar(stat="identity")+facet_wrap("Multiple")+ggtitle("Multimapping")+theme(axis.text.x = element_text(angle=45,vjust=1,hjust=1))
PacBioCat_COUNTS$p[PacBioCat_COUNTS$p>0.01] <- NA
PacBioCat_COUNTS$p.adjust[PacBioCat_COUNTS$p.adjust>0.01] <- NA

g2 <- ggplot((PacBioCat_COUNTS[Multiple==FALSE]),aes(DM_min_categories,Counts_rel.N,fill=names,label=signif(p.adjust,2)))+geom_bar(stat="identity")+theme(axis.text.x = element_text(angle=45,hjust=1,vjust=1))+geom_text(angle=90,size=3,position = position_stack(vjust = 0.5),hjust=0.5 ,size=4)+xlab("")+ylab("Fraction")



pdf("PACBIO_Cat_Enrichment.pdf",width=12)
plot_grid(g1,g2)
dev.off()
system("open PACBIO_Cat_Enrichment.pdf")
```


## Subfigures 1

```{r}

subset_table$UniprotIsoformType[is.na(subset_table$UniprotIsoformType)] <- "UNKNOWN"
pg <- plot_grid(
  ggplot(subset_table,aes(DM_PB_min,color=UniprotIsoformType))+geom_density(alpha=0.5)+coord_cartesian(xlim=c(-500000,100000))+theme(legend.title = element_blank())+xlab("PACBIO Delta Mass [Da]"),
  ggplot(subset_table,aes(DM_UP_min,color=UniprotIsoformType))+geom_density(alpha=0.5)+coord_cartesian(xlim=c(-500000,100000))+theme(legend.title = element_blank())+xlab("UniProt Delta Mass [Da]"),ncol = 1
)




AUC <- Results_CombinedGels_PBUP$ROC
AUC <- AUC+theme(legend.position = "bottom", legend.box="vertical", legend.margin=margin(),legend.title = element_blank())+guides(color=guide_legend(nrow=2,byrow=TRUE))

ra <- c(-10^5,10^5)*c(5,2)
DeltaMassFig <- Results_CombinedGels_PBUP$DM_plots$DM+theme(legend.position = "right",legend.title =element_blank())+coord_cartesian(xlim=ra,ylim=ra)+guides(color=guide_legend(byrow=TRUE))
DeltaMassFigCounts <- Results_CombinedGels_PBUP$DM_plots$counts

ProteinCategory_BP <- Results_CombinedGels_PBUP$IsoformFractions
Category_BP <- Results_CombinedGels_PBUP$FractionOfIdentifed+ylab("# Isoforms")+xlab("Best Scoring DB")
Category_BP_rel <- Results_CombinedGels_PBUP$FractionOfIdentifed_rel+ylab("Isoform Fraction")+xlab("Best Scoring DB")

Category_BP2 <- pg+theme(legend.position = "none")

Results_CombinedGels_PBUP$Category_BP

upperrow <- plot_grid(AUC+xlab("Specificity")+ggtitle("ROC PACBIO Best Matching Entry"),
          Category_BP+theme(legend.position = "none"),Category_BP_rel,
          
          rel_widths = c(0.8,0.7,1),labels = c("A","B"),ncol=3)

lowerrow <- plot_grid(DeltaMassFig+ggtitle("Isoform Pairs UniProt vs PACBIO\n Uniprot or PACBIO Score > 0.8")+update_geom_defaults(geom="point",new =list(size=0.5) ),DeltaMassFigCounts+ggtitle("Counts")+xlab("")+theme(legend.position = "none",axis.text.x=element_text(angle=45,vjust=1,hjust=1)),ProteinCategory_BP
          ,rel_widths = c(1.3,0.5,1),nrow=1,rel_heights = c(1,0.7),labels = c("C","","D"))


```

## Updating SQLITE Database

```{r}
IsofracCandidates <- list.files(pattern="IsoFrac_\\d")
IsofracCandidates<- grep("7z$",IsofracCandidates,invert=T,value=T)
pa <- sort(IsofracCandidates,decreasing = T)[1]
setwd(pa)

library(RSQLite)
dbpath <- "./TableListNormEnsHdbScanMW_Peaks.sqlite"
if(file.exists(dbpath)){
db <- dbConnect(SQLite(),dbpath)
dbListTables(db)
Results_CombinedGels_PBUP$Pairs_Unfiltered$V1 <- NULL
names(Results_CombinedGels_PBUP$Pairs_Unfiltered) <- make.unique(names(Results_CombinedGels_PBUP$Pairs_Unfiltered))
dbWriteTable(db,"CombiTable_PBUP",Results_CombinedGels_PBUP$Pairs_Unfiltered,overwrite=T)
dbWriteTable(db,"CombiTableFiltered_PBUP",Results_CombinedGels_PBUP$Pairs_Filtered,overwrite=T)

df <- Results_CombinedGels_PacBio$Pairs_Unfiltered_PacBio
class(df)
df$V1 <- NULL
dful <- apply(df,2,unlist)
dful <- as.data.frame(dful)

colnames(dful) <- make.unique(colnames(dful))

RowWiseCompare$PeakScore <- 100
names(RowWiseCompare)

RowWiseCompare_PB <- RowWiseCompare[,.SD,.SDcols = c("gel","type","ENSG","IsoFrac","dhyper_p","Success_PB_Peptides","All_Uniprot_Peptides","Non_Uniprot_Peptides","IsofracOverlap","UniprotOverlap","PACBIO_ACC","PB_MW","IsoFrac_MW","Uniprot_MW","DM_PB_min","nClusters","nGels","heights","Height","Height_lm","Fraction_sd_80","gel_MainPeak","score_ref_rf_PB")]
setnames(RowWiseCompare_PB,c("score_ref_rf_PB","Success_PB_Peptides","DM_PB_min","PACBIO_ACC"),c("score_ref","Success_Uniprot_Peptides","DeltaMass","Uniprot"))

RowWiseCompare_UP <- RowWiseCompare[,.SD,.SDcols = c("gel","type","ENSG","IsoFrac","dhyper_p","Success_Uniprot_Peptides","All_Uniprot_Peptides","Non_Uniprot_Peptides","IsofracOverlap","UniprotOverlap","Uniprot","IsoFrac_MW","DM_UP","DM_UP_min","DeltaMass","nClusters","nGels","heights","Height","Height_lm","Fraction_sd_80","gel_MainPeak","score_ref_rf_UP")]
setnames(RowWiseCompare_UP,c("score_ref_rf_UP","DM_UP_min"),c("score_ref","DeltaMass"))

names(db)
# dbWriteTable(db,"CombinedGels_Pairs_Unfiltered_PACBIO",Results_CombinedGels_PacBio$Pairs_Unfiltered_PacBio[type=="fw"&DataType=="PB"],overwrite=T)
# dbWriteTable(db,"CombinedGels_Pairs_Unfiltered",Results_CombinedGels_PacBio$Pairs_Unfiltered_PacBio[type=="fw"&DataType=="UP"],overwrite=T)

PBfun <- as.data.frame(Results_CombinedGels_PBUP$BestMatch_PB)[,!duplicated(names(Results_CombinedGels_PBUP$BestMatch_PB))]
UPfun <- as.data.frame(Results_CombinedGels_PBUP$BestMatch_UP)[,!duplicated(names(Results_CombinedGels_PBUP$BestMatch_UP))]
PBfun$score_ref <- PBfun$score_UP_gb
UPfun$score_ref <- UPfun$score_
PBfun$score_ref_PacBIO <- PBfun$score_ref
PBfun <- as.data.table(PBfun)
UPfun <- as.data.table(UPfun)

# PBfun[ENSG=="ENSG00000109320"&type=="fw",]$
# UPfun[ENSG=="ENSG00000109320"&type=="fw",]$score_ref


dbWriteTable(db,"CombinedGels_Pairs_Filtered_PACBIO",PBfun[PBfun$type=="fw",],overwrite=T)
dbWriteTable(db,"CombinedGels_Pairs_Filtered",UPfun[UPfun$type=="fw",],overwrite=T)

# dbWriteTable(db,"CombinedGels_Compressed_Pairs_PACBIO",as.data.table(apply(Results_CombinedGels_PacBio$Compressed_Pairs,2,unlist),integer64="double"),overwrite=T)

# dbWriteTable(db,"CombinedGels_Compressed_Pairs_PB_UP",as.data.table(apply(RowWiseCompare[type=="fw"&ScoreType=="score_ref_rf"],2,unlist),integer64="double"),overwrite=T)
}

# dbWriteTable(db,name"test"="Version",value = Version,overwrite=T)

GN <- lapply(TableListNormEnsHdbScanMW_Peaks,function(x){
  gel <- x$xname
  dt_gn <- x$dt_gn
  dt_gn$gel <- gel
  dt_gn
})
GN <- rbindlist(GN)

hdbscanTables <- lapply(TableListNormEnsHdbScanMW_Peaks,function(x){
  xtemp <- x$hdbscanTables
  xtemp$Abu <- as.numeric(xtemp$Abu)
  xtemp$Gel <- x$xname
  xtemp
})
hdbscanTables <- rbindlist(hdbscanTables)
hdbscanTables[EnsG=="ENSG00000127603",length(unique(BestCluster)),.(Gel)]

TMTSubset_final_singlenorm_zscore <- lapply(TableListNormEnsHdbScanMW_Peaks,function(x){
  xtemp <- x$TMTSubset_final_singlenorm_zscore
  xtemp
})

IsoFrac_PeakResultsTable <- lapply(TableListNormEnsHdbScanMW_Peaks,function(x){
  xtemp <- x$IsoFrac_PeakResultsTable
    xtemp$gel <- x$xname

  xtemp
})
IsoFrac_PeakResultsTable <- rbindlist(IsoFrac_PeakResultsTable)
IdentifiedPeaks_Table <- lapply(TableListNormEnsHdbScanMW_Peaks,function(x){
  xtemp <- x$IdentifiedPeaks_Table
  xtemp$gel <- x$xname
  xtemp
})
IdentifiedPeaks_Table <- rbindlist(IdentifiedPeaks_Table)

IsoFrac_PeakResultsTable <- apply(IsoFrac_PeakResultsTable,2,unlist)
IsoFrac_PeakResultsTable <- as.data.table(IsoFrac_PeakResultsTable,integer64="double")

IdentifiedPeaks_Table <- apply(IdentifiedPeaks_Table,2,unlist)
IdentifiedPeaks_Table <- as.data.table(IdentifiedPeaks_Table,integer64="double")
gelnames <- sapply(TableListNormEnsHdbScanMW_Peaks,function(x){x$xname})
dbWriteTable(db,"GN",GN,overwrite=T)
dbWriteTable(db,"Gels",data.frame(gels=gelnames),GN,overwrite=T)

dbWriteTable(db,"hdbscanTables",hdbscanTables,overwrite=T)
dbWriteTable(db,"IsoFrac_PeakResultsTable",IsoFrac_PeakResultsTable,overwrite=T)
dbWriteTable(db,"IdentifiedPeaks_Table",IdentifiedPeaks_Table,overwrite=T)
for(x in TableListNormEnsHdbScanMW_Peaks){
  temp <- x$TMTSubset_final_singlenorm_zscore
  dbWriteTable(db,paste("TMTSubset_final_singlenorm_zscore",x$xname,sep="_#_"),temp,overwrite=T)
  dbWriteTable(db,paste("Protein_MW",x$xname,sep="_#_"),x$Protein_MW,overwrite=T)
  dbWriteTable(db,paste("Fractions_MW",x$xname,sep="_#_"),data.frame(x$Fractions_predicted_MW),overwrite=T)
  dbWriteTable(db,paste("hdbscanTables",x$xname,sep="_#_"),data.frame(x$hdbscanTables),overwrite=T)


}
dbListTables(db)
df <- Results_CombinedGels$Pairs_Unfiltered
class(df)
dful <- apply(df,2,unlist)
dful <- as.data.frame(dful)

colnames(dful) <- make.unique(colnames(dful))
Results_CombinedGels$Pairs_Unfiltered_combi$score_ref <- Results_CombinedGels$Pairs_Unfiltered_combi$score_ref_rf

df <- Results_CombinedGels$Pairs_Unfiltered_combi
names(df) <- make.unique(names(df))
dbWriteTable(db,"CombinedGels_Pairs_Unfiltered",df,overwrite=T)
dbWriteTable(db,"CombinedGels_Compressed_Pairs",as.data.table(apply(Results_CombinedGels$Compressed_Pairs,2,unlist),integer64="double"),overwrite=T)
colnames(Results_CombinedGels$Pairs_Filtered) <- make.unique(colnames(Results_CombinedGels$Pairs_Filtered))
dbWriteTable(db,"CombinedGels_Pairs_Filtered",Results_CombinedGels$Pairs_Filtered,overwrite=T)
apply(Results_CombinedGels$Pairs_Filtered,2,class)
#colnames(Results_CombinedGels$DeltaMassComparison) <- make.unique(colnames(Results_CombinedGels$DeltaMassComparison))
#dbWriteTable(db,"CombinedGels_DeltaMassComparison",Results_CombinedGels$DeltaMassComparison,overwrite=T)
# dbWriteTable(db,"CombinedGels_DeltaMassComparison",Results_CombinedGels$CountTable_FDR_sg)
# al <- data.table(al)

  IsoformCount <- lapply(TableListNormEnsHdbScanMW_Peaks,function(x){
    x<<- x
    IP <- x$IdentifiedPeaks_Table
    # IP <- x$IdentifiedPeaks_Table
    IP <- IP[FDR_RF<0.01,,]
    IP <- IP[unlist(Height)>=0,,]
    IP <- IP[unlist(Fraction_width)<2,]
    IP <- IP[grep(";",ENS,invert = T),{list(MWlog10=(unique(MW_lm_Compressed_)),
                                            mpw=min(unlist(Height)),
                                            maxFracWidth = max(unlist(Fraction_width),na.rm = T),
                                            MW_lm_lo = median(MW_lm_lo),
                                            MW_lm_hi=median(MW_lm_lo))},ENS]
    IP$Experiment <- x$xname
    IP
  })
  IsoformCountTable <- rbindlist(IsoformCount)
  table(IsoformCountTable$Experiment)
  IsoformCountTable$Experiment <-  gsub("/peptides.txt","",IsoformCountTable$Experiment)
  IsoformCountTable$Experiment <- gsub("/peptides_TryChymCOMBINED.txt","_ChyTry",IsoformCountTable$Experiment)
  IsoformCountTable$Experiment <-  gsub("/peptides.txt","",IsoformCountTable$Experiment)
  IsoformCountTable$Experiment <- gsub("/peptides_TryChymCOMBINED.txt","_ChyTry",IsoformCountTable$Experiment)
  IsoformCountTable$Experiment<- gsub("2020_12_23_TMTpro_","gel1_",IsoformCountTable$Experiment)
  table(IsoformCountTable$Experiment)
  IsoformCountTable$Experiment<- gsub("2021_10_09_TM.*.pro_","gel2_",IsoformCountTable$Experiment)
  dbWriteTable(db,"IsoformCountTable",IsoformCountTable,overwrite=T)

dbDisconnect(db)
  system("open ./")
  
file.copy("../Results_CombinedGels.rda",paste(dirname(dbpath),"Results_CombinedGels.rda",sep="/"),overwrite = T)
setwd("../")
```

# Selecting Specific Proteins

```{r}
library(plotly)
library(ggrepel)
RowWiseCompare$PickedGenes <- !is.na(match(RowWiseCompare$hgnc_symbol,c(names(GeneList),"CEBPB","NFKB1")))
subset_table<- RowWiseCompare[(selectedScore_UP>0.8|selectedScore_PB>0.8)&!grepl(";",ENSG)&selectedScore_UP<selectedScore_PB]
subset_table_MorePB<- RowWiseCompare[(selectedScore_UP>0.8|selectedScore_PB>0.8)&!grepl(";",ENSG)&as.numeric(Success_ISOFRAC_Peptides_PB)>as.numeric(Success_ISOFRAC_Peptides_UP)]

# size <- as.nu
gfu2_PacbioBetter <- ggplot(subset_table,aes(DM_PB_min,DM_UP_min,color=DM_min_categories,label=paste(hgnc_symbol,PACBIO_ACC,Uniprot,size=as.numeric(Success_PB_Peptides)/as.numeric(Success_ISOFRAC_Peptides_UP),sep="\n")))+geom_point()+xlab("IsoFrac-PACBIO MW [D]")+ylab("IsoFrac-UniProt MW [D]")
gfu2_PacbioMORE <- ggplot(subset_table_MorePB,aes(DM_PB_min,DM_UP_min,color=DM_min_categories,size=as.numeric(Success_PB_Peptides)/as.numeric(Success_ISOFRAC_Peptides_UP),label=paste(hgnc_symbol,PACBIO_ACC,Uniprot,sep="\n")))+geom_point()+xlab("IsoFrac-PACBIO MW [D]")+ylab("IsoFrac-UniProt MW [D]")
ggplotly(gfu2_PacbioMORE)

subset_table_MorePB[DM_min_categories=="PACBIO correct+",,]
# 
subset_table_MorePB$Success_PB_Peptides/subset_table_MorePB$Success_ISOFRAC_Peptides_UP

ggplotly(gfu2_PacbioBetter)
pdf("Candidate_GEnes.pdf",width=20,height=20)
gfu2_PickedGenes <- ggplot(RowWiseCompare[TRUE==PickedGenes],aes(DM_PB_min,DM_UP_min,color=DM_min_categories,label=paste(hgnc_symbol),sep="\n"))+geom_point()+xlab("IsoFrac-PACBIO MW [D]")+ylab("IsoFrac-UniProt MW [D]")+geom_text_repel()
print(gfu2_PickedGenes)
dev.off()
system("open Candidate_GEnes.pdf")


ggplot(RowWiseCompare[score_Ref_cutoff],aes(DM_PB_min,DM_UP_min,color=DM_min_categories,label=paste(hgnc,PACBIO_ACC,Uniprot,sep="\n")))+geom_point()+xlab("IsoFrac-PACBIO MW [D]")+ylab("IsoFrac-UniProt MW [D]")+facet_wrap("Categories_Uniprot_Uniprot_fixed")+coord_cartesian(xlim=ra,ylim=ra)

g <- ggplot(RowWiseCompare[score_Ref_cutoff],aes(DM_PB_min,DM_UP_min,color=DM_min_categories,label=paste(hgnc,PACBIO_ACC,Uniprot,sep="\n")))+geom_point()+xlab("IsoFrac-PACBIO MW [D]")+ylab("IsoFrac-UniProt MW [D]")+facet_wrap("Categories_Uniprot_Uniprot_fixed")+coord_cartesian(xlim=ra,ylim=ra)


ggplotly(g)
RowWiseCompare[all_sequences>20&DM_min_categories=="Equal MassShift"]


```

## PACBIO Exclusive Peaks With ALL Entries

- filtering PacBIo to match known technical limitations of proteomic data which are:
  - sensitivity (e.g. TPM> 1)
  - ORF redundancy due to many UTRs
  - MW Resolution of IsoFrac

```{r}
print("PACBIO DB:")

# Assigns Each IsoFrac Peak to the best mapping Database entry, either with or without removing the sequence from the selection:
DB_IF_hits <- function(DB,ALLPAIRS,rdaname="Pacbio_Identified.rda",RemoveSequences=F,Rerun=F){
  
  if(!file.exists(rdaname)|Rerun){
 
    PacBIOMatched <- DB[,{
      cat("\r",.GRP)
      gr <- .BY
      temp <- .SD
      IDfun <- paste(temp$ProteinID,paste(temp$MW,"Da"),sep=" # ")
      FUNNI <- ALLPAIRS[!is.na(match(Uniprot,IDfun)),]
      if(dim(FUNNI)[1]==0){
        FUNNI <- FUNNI[1,]
      }
      FUNNI$TPM <- sum(temp$Abundance_TPM)
      FUNNI
    },.(SEQUENCE)]
    
    
    PacBIOMatched<- PacBIOMatched[order(score_ref_gb,decreasing = T)]
    
    
    
    OutFun <- data.table()
    PacBIOMatched$IsoFracID <- paste(PacBIOMatched$ENSG,PacBIOMatched$IsoFrac_MW)
    IsoFracID <- unique(PacBIOMatched$IsoFracID)
    PacBIOMatchedTemp <- copy(PacBIOMatched)
    
    while(length(IsoFracID)>0){
      cat("\r",length(IsoFracID))
      temp <- PacBIOMatchedTemp[1,]
      PacBIOMatchedTemp <- PacBIOMatchedTemp[IsoFracID!=temp$IsoFracID,]
      if(RemoveSequences){
        PacBIOMatchedTemp <- PacBIOMatchedTemp[SEQUENCE!=temp$SEQUENCE,]
      }
      
      IsoFracID <- IsoFracID[IsoFracID!=temp$IsoFracID]
      IsoFracID <- IsoFracID[!is.na(IsoFracID)]
      OutFun <- rbind(OutFun,temp)
      
    }
    
    save(OutFun,RemoveSequences,PacBIOMatched,file=rdaname)
    
  }
  rdaname
}
# plotting of this data
PB_IsoFrac_Evidences <- function(rdaname){
  load(rdaname)
  MappedTable <- OutFun
  MappedTable$id <- 1:dim(MappedTable)[1]
  MappedTable$score_ref_rf[is.na(MappedTable$score_ref_rf)] <- -1
  scorediffs <- c(0.8,0.5,-1,-1)
  MappedTable$score_ref_rf <- as.numeric(MappedTable$score_ref_rf)
  IDs <- list(
    "High ML Score" = MappedTable[score_ref_rf>scorediffs[1]]$SEQUENCE,
    "Middle ML Score" = MappedTable[score_ref_rf<=scorediffs[1]&score_ref_rf>scorediffs[2]]$SEQUENCE,
    "Low ML Score" = MappedTable[score_ref_rf<=scorediffs[2]&score_ref_rf>scorediffs[3]]$SEQUENCE,
    No_Evidence = setdiff(unique(PacBIOMatched$SEQUENCE),unique(MappedTable$SEQUENCE))
  )
  IDs <- lapply(IDs,unique)
  names(IDs) <- paste(names(IDs),c(paste(">",scorediffs[1],collapse=" "),
                                   paste(scorediffs[c(1,2)],collapse=" to "),
                                   paste(scorediffs[c(2,3)],collapse=" to "),""))
  PB_identified <- data.table(lengths(IDs),names(lengths(IDs)))
  PB_identified$V2 <- factor(PB_identified$V2,names(IDs))
  gg <- ggplot(PB_identified,aes(V2,V1,fill=V2,label=V1))+geom_bar(stat="identity",
                                                             position="dodge")+theme(axis.text.x = element_text(angle=45,hjust=1,vjust=1))+
    geom_text(vjust=-1)+ylab("PACBIO IsoForm proteomic Evidences")
  
  list(gg=gg,table=PB_identified)
}
# unlink("Pacbio_Identified_ALL.rda")
P1 <- PacBIO
hu <- DB_IF_hits(P1,Results_CombinedGels_PBUP$Pairs_Unfiltered[type=="fw"],
           rdaname = "Pacbio_Identified_ALL.rda",RemoveSequences=T)
hm <- load("Pacbio_Identified_ALL.rda")
P2 <- PacBIO[SEQUENCE==SequenceMerged]
DB_IF_hits(P2,Results_CombinedGels_PBUP$Pairs_Unfiltered[type=="fw"],
           rdaname = "Pacbio_Identified_SizeCondensed.rda",RemoveSequences=T)
P3 <- PacBIO[SEQUENCE==SequenceMerged&Abundance_TPM>0]
DB_IF_hits(P3,Results_CombinedGels_PBUP$Pairs_Unfiltered[type=="fw"],
           rdaname = "Pacbio_Identified_SizeCondensed_TPM.rda",RemoveSequences=T)
P4 <- PacBIO[SEQUENCE==SequenceMerged&Abundance_TPM>5]
DB_IF_hits(P4,Results_CombinedGels_PBUP$Pairs_Unfiltered[type=="fw"],
           rdaname = "Pacbio_Identified_SizeCondensed_TPM5.rda",RemoveSequences=T)
# Size and TPM
P4.1 <- PacBIO[SEQUENCE==SequenceMerged&Abundance_TPM>1]
DB_IF_hits(P4.1,Results_CombinedGels_PBUP$Pairs_Unfiltered[type=="fw"],
           rdaname = "Pacbio_Identified_SizeCondensed_TPM1.rda",RemoveSequences=T)
P4.2 <- PacBIO[SEQUENCE==SequenceMerged&Abundance_TPM>2]
DB_IF_hits(P4.2,Results_CombinedGels_PBUP$Pairs_Unfiltered[type=="fw"],
           rdaname = "Pacbio_Identified_SizeCondensed_TPM2.rda",RemoveSequences=T)
P4.3 <- PacBIO[SEQUENCE==SequenceMerged&Abundance_TPM>3]
DB_IF_hits(P4.3,Results_CombinedGels_PBUP$Pairs_Unfiltered[type=="fw"],
           rdaname = "Pacbio_Identified_SizeCondensed_TPM3.rda",RemoveSequences=T)
# Only TPM:
P4.1.0 <- PacBIO[Abundance_TPM>1]
DB_IF_hits(P4.1.0,Results_CombinedGels_PBUP$Pairs_Unfiltered[type=="fw"],
           rdaname = "Pacbio_Identified_TPM1.rda",RemoveSequences=T)
P4.2.0 <- PacBIO[Abundance_TPM>2]
DB_IF_hits(P4.2.0,Results_CombinedGels_PBUP$Pairs_Unfiltered[type=="fw"],
           rdaname = "Pacbio_Identified_TPM2.rda",RemoveSequences=T)
P4.3.0 <- PacBIO[Abundance_TPM>3]
DB_IF_hits(P4.3.0,Results_CombinedGels_PBUP$Pairs_Unfiltered[type=="fw"],
           rdaname = "Pacbio_Identified_TPM3.rda",RemoveSequences=T)


DB_IF_hits_figure <- lapply(c("Pacbio_Identified_ALL.rda",
                              "Pacbio_Identified_SizeCondensed.rda",
                              "Pacbio_Identified_SizeCondensed_TPM1.rda",  
                              "Pacbio_Identified_SizeCondensed_TPM2.rda", 
                              "Pacbio_Identified_SizeCondensed_TPM3.rda",
                              "Pacbio_Identified_TPM1.rda",  
                              "Pacbio_Identified_TPM2.rda", 
                              "Pacbio_Identified_TPM3.rda"
                              ),PB_IsoFrac_Evidences)

names(DB_IF_hits_figure) <- c("All","Size Restricted","Size & TPM > 1","Size & TPM > 2","Size & TPM > 3","TPM > 1","TPM > 2","TPM > 3")

DB_IF_hits_list <- lapply(1:length(DB_IF_hits_figure),function(it){
  x <- DB_IF_hits_figure[[it]]
  xt <- x$table
  xt$type <- names(DB_IF_hits_figure)[it]
  xt
})
DB_IF_hits_dt <- rbindlist(DB_IF_hits_list)
files <- c("Pacbio_Identified_ALL.rda",
                              "Pacbio_Identified_SizeCondensed.rda",
                              "Pacbio_Identified_SizeCondensed_TPM1.rda",  
                              "Pacbio_Identified_SizeCondensed_TPM2.rda", 
                              "Pacbio_Identified_SizeCondensed_TPM3.rda",
                              "Pacbio_Identified_TPM1.rda",  
                              "Pacbio_Identified_TPM2.rda", 
                              "Pacbio_Identified_TPM3.rda"
                              )
PB_Matched <- lapply(1:length(files),function(it){
  x <- files[it]
                                hu <-load(x)
                                PacBIOMatched$SubsetType <- names(DB_IF_hits_figure)[it]
                                names(DB_IF_hits_figure) 
                                PacBIOMatched
                              })

PB_Matcheddt <- rbindlist(PB_Matched)
PB_Matcheddt$ScoreType <- NULL
PB_Matcheddt$ScoreType <- "No Evidence"
scorediffs <- c(0.8,0.5,-1,-1)

PB_Matcheddt$ScoreType[PB_Matcheddt$score_UP_gb >= scorediffs[1]] <- "High ML Score"
PB_Matcheddt$ScoreType[PB_Matcheddt$score_UP_gb < scorediffs[1]&PB_Matcheddt$score_UP_dl > scorediffs[2]] <- "Middle ML Score"
PB_Matcheddt$ScoreType[PB_Matcheddt$score_UP_gb < scorediffs[2]] <- "Low ML Score"

PB_Matcheddt$ScoreType <- factor(PB_Matcheddt$ScoreType,sapply(c("High","Middle","Low","No"),function(x){grep(x,PB_Matcheddt$ScoreType,value=T)[1]}))
names(PB_Matcheddt) <- make.unique(names(PB_Matcheddt))

PB_Matcheddt$SubsetType <- factor(PB_Matcheddt$SubsetType,names(DB_IF_hits_figure))

DB_IF_hits_dt$Group <- "All Abundances"
DB_IF_hits_dt$Group[grep("Size &",DB_IF_hits_dt$type)] <- "Size & TPM"
DB_IF_hits_dt$Group[grep("^TPM",DB_IF_hits_dt$type)] <- "TPM"
ra <- c(0,max(DB_IF_hits_dt$V1))
PB_ExclusiveBestPeak <- ggplot(DB_IF_hits_dt[Group=="All Abundances"],aes(V2,V1,fill=type,label=V1,group=type))+geom_bar(stat="identity",position="dodge")+theme(axis.text.x = element_text(angle=45,hjust=1,vjust=1))+geom_text(position=position_dodge(width = 1),size=3,vjust = 0.5,hjust=1,angle=90,color="black")+ylab("PACBIO IsoForms")+ylim(ra)+xlab("")+ggtitle("")
# PB_ExclusiveBestPeakC

PB_ExclusiveBestSizeTPM<- ggplot(DB_IF_hits_dt[Group=="Size & TPM"],aes(V2,V1,fill=type,label=V1,group=type))+geom_bar(stat="identity",position="dodge")+theme(axis.text.x = element_text(angle=45,hjust=1,vjust=1))+geom_text(position=position_dodge(width = 1),size=3,vjust = 0.5,hjust=1,angle=90,color="black")+ylab("PACBIO IsoForms")+ylim(ra)+xlab("")+ggtitle("with TPM CutOff &\nSize Restriction")
# PB_ExclusiveBestSizeTPMC
PB_ExclusiveBestSize<- ggplot(DB_IF_hits_dt[Group=="TPM"],aes(V2,V1,fill=type,label=V1,group=type))+geom_bar(stat="identity",position="dodge")+theme(axis.text.x = element_text(angle=45,hjust=1,vjust=1))+geom_text(position=position_dodge(width = 1),vjust = 0.5,size=3,hjust=1,angle=90,color="black")+ylab("PACBIO IsoForms")+ylim(ra)+xlab("")+ggtitle("with TPM CutOff")
# PB_ExclusiveBestSize


Dgg <- ggplot(PacBIO,aes(log10(Abundance_TPM)))+geom_density(fill="skyblue")+geom_vline(xintercept = log10(c(1,2,3)),color="orange",lty="solid",lwd=1)+ggplot2::coord_cartesian(xlim=c(-1,4))+annotate("text",x=c(log10(1:3)),y=c(0,0,0),label=paste("Cutoff:",1:3),hjust=0,vjust=1,angle=90)+ylab("Density")

pdf("PB_Exclusive_Peaks.pdf",width=12,height=12)


DB_IF_hits_dt[,V1_rel :={V1/sum(V1)},.(Group,type)]
AbundanceBP <- ggplot(PB_Matcheddt,aes(log10(TPM),ScoreType,color=ScoreType))+geom_boxplot()+facet_wrap("SubsetType",ncol=1)+coord_cartesian(xlim=c(0,2))


DB_FUN <- DB_IF_hits_dt[as.character(V2)!="No_Evidence ",{list(V1=sum(V1[V2!="No_Evidence"]),V1_rel=sum(V1_rel[V2!="No_Evidence"]))},.(type,Group)]
FractionCheck <- ggplot(DB_FUN,aes(V1,V1_rel,label=paste(type),pch=Group))+geom_point(size=3)+geom_text_repel()+xlab("Identified Isoforms")+ylab("Identified Fraction")

p1 <- plot_grid(PB_ExclusiveBestPeak+theme(legend.position = "bottom")+guides(fill=guide_legend(nrow=3,byrow=TRUE,size=0.5,title = element_blank())),
          PB_ExclusiveBestSize+theme(legend.position = "bottom")+guides(fill=guide_legend(nrow=3,byrow=TRUE,size=0.5,title = element_blank())),
          PB_ExclusiveBestSizeTPM+theme(legend.position = "bottom")+guides(fill=guide_legend(nrow=3,byrow=TRUE,size=0.5,title = element_blank())),nrow=1,align = "hv")
p2 <- FractionCheck+ggtitle("Identified Fraction\n")+theme(legend.position = "right")+guides(pch=guide_legend(nrow=3,byrow=TRUE,size=1,title = element_blank()))

ExclusivePeaks_GG <- plot_grid(plot_grid(Dgg,AbundanceBP,ncol=1,rel_heights =  c(1,2),labels = c("A","B")  ),plot_grid(p1,p2,ncol=1,rel_heights = c(1,0.8),labels=c("C","D")),ncol=2,rel_widths = c(0.7,1))
plot(ExclusivePeaks_GG)

dev.off()
system("open PB_Exclusive_Peaks.pdf")

ggsave("PB_Exclusive_Peaks.png",ExclusivePeaks_GG,height = 13,width=12 )
system("open PB_Exclusive_Peaks.png")

```

## PACBIO Exclusive Peaks With Only 1+ Entries

```{r}
print("PACBIO DB:")

# Assigns Each IsoFrac Peak to the best mapping Database entry, either with or without removing the sequence from the selection:
DB_IF_hits <- function(DB,ALLPAIRS,rdaname="Pacbio_Identified.rda",RemoveSequences=F,Rerun=F){
  
  if(!file.exists(rdaname)|Rerun){
    
    PacBIOMatched <- DB[,{
      cat("\r",.GRP)
      gr <- .BY
      temp <- .SD
      IDfun <- paste(temp$ProteinID,paste(temp$MW,"Da"),sep=" # ")
      FUNNI <- ALLPAIRS[!is.na(match(Uniprot,IDfun)),]
      if(dim(FUNNI)[1]==0){
        FUNNI <- FUNNI[1,]
      }
      FUNNI$TPM <- sum(temp$Abundance_TPM)
      FUNNI
    },.(SEQUENCE)]
    
    
    PacBIOMatched<- PacBIOMatched[order(score_ref_gb,decreasing = T)]
    
    
    
    OutFun <- data.table()
    PacBIOMatched$IsoFracID <- paste(PacBIOMatched$ENSG,PacBIOMatched$IsoFrac_MW)
    IsoFracID <- unique(PacBIOMatched$IsoFracID)
    PacBIOMatchedTemp <- copy(PacBIOMatched)
    
    while(length(IsoFracID)>0){
      cat("\r",length(IsoFracID))
      temp <- PacBIOMatchedTemp[1,]
      PacBIOMatchedTemp <- PacBIOMatchedTemp[IsoFracID!=temp$IsoFracID,]
      if(RemoveSequences){
        PacBIOMatchedTemp <- PacBIOMatchedTemp[SEQUENCE!=temp$SEQUENCE,]
      }
      
      IsoFracID <- IsoFracID[IsoFracID!=temp$IsoFracID]
      IsoFracID <- IsoFracID[!is.na(IsoFracID)]
      OutFun <- rbind(OutFun,temp)
      
    }
    
    save(OutFun,RemoveSequences,PacBIOMatched,file=rdaname)
    
  }
  rdaname
}
# plotting of this data
PB_IsoFrac_Evidences <- function(rdaname,SelectENSGs=NULL){
  load(rdaname)
  if(length(SelectENSGs)>0&is.character(SelectENSGs)){
    OutFun <- OutFun[!is.na(match(ENSG,SelectENSGs))]
    PacBIOMatched <- PacBIOMatched[!is.na(match(ENSG,SelectENSGs))]
  }
  MappedTable <- OutFun
  MappedTable$id <- 1:dim(MappedTable)[1]
  MappedTable$score_ref_rf[is.na(MappedTable$score_ref_rf)] <- -1
  scorediffs <- c(0.8,0.5,-1,-1)
  MappedTable$score_ref_rf <- as.numeric(MappedTable$score_ref_rf)
  IDs <- list(
    "High_Evidence" = MappedTable[score_ref_rf>scorediffs[1]]$SEQUENCE,
    Good_Evidence = MappedTable[score_ref_rf<=scorediffs[1]&score_ref_rf>scorediffs[2]]$SEQUENCE,
    Low_Evidences = MappedTable[score_ref_rf<=scorediffs[2]&score_ref_rf>scorediffs[3]]$SEQUENCE,
    No_Evidence = setdiff(unique(PacBIOMatched$SEQUENCE),unique(MappedTable$SEQUENCE))
  )
  IDs <- lapply(IDs,unique)
  names(IDs) <- paste(names(IDs),c(paste(">",scorediffs[1],collapse=" "),
                                   paste(scorediffs[c(1,2)],collapse=" to "),
                                   paste(scorediffs[c(2,3)],collapse=" to "),""))
  PB_identified <- data.table(lengths(IDs),names(lengths(IDs)))
  PB_identified$V2 <- factor(PB_identified$V2,names(IDs))
  gg <- ggplot(PB_identified,aes(V2,V1,fill=V2,label=V1))+geom_bar(stat="identity",
                                                             position="dodge")+theme(axis.text.x = element_text(angle=45,hjust=1,vjust=1))+
    geom_text(vjust=-1)+ylab("PACBIO IsoForm proteomic Evidences")
  
  list(gg=gg,table=PB_identified)
}
Selection <- PacBIO[,length(unique(SEQUENCE)),ENSG]
ENSGSelection <- Selection[V1>1]$ENSG

DB_IF_hits_figure <- lapply(c("Pacbio_Identified_ALL.rda",
                              "Pacbio_Identified_SizeCondensed.rda",
                              "Pacbio_Identified_SizeCondensed_TPM1.rda",  
                              "Pacbio_Identified_SizeCondensed_TPM2.rda", 
                              "Pacbio_Identified_SizeCondensed_TPM3.rda",
                              "Pacbio_Identified_TPM1.rda",  
                              "Pacbio_Identified_TPM2.rda", 
                              "Pacbio_Identified_TPM3.rda"
                              ),PB_IsoFrac_Evidences,SelectENSGs=ENSGSelection)

names(DB_IF_hits_figure) <- c("All","Size Restricted","Size & TPM > 1","Size & TPM > 2","Size & TPM > 3","TPM > 1","TPM > 2","TPM > 3")

DB_IF_hits_list <- lapply(1:length(DB_IF_hits_figure),function(it){
  x <- DB_IF_hits_figure[[it]]
  xt <- x$table
  xt$type <- names(DB_IF_hits_figure)[it]
  xt
})

DB_IF_hits_dt <- rbindlist(DB_IF_hits_list)
files <- c("Pacbio_Identified_ALL.rda",
                              "Pacbio_Identified_SizeCondensed.rda",
                              "Pacbio_Identified_SizeCondensed_TPM1.rda",  
                              "Pacbio_Identified_SizeCondensed_TPM2.rda", 
                              "Pacbio_Identified_SizeCondensed_TPM3.rda",
                              "Pacbio_Identified_TPM1.rda",  
                              "Pacbio_Identified_TPM2.rda", 
                              "Pacbio_Identified_TPM3.rda"
                              )
PB_Matched <- lapply(1:length(files),function(it){
  x <- files[it]
                                hu <<-load(x)
                                PacBIOMatched$SubsetType <- names(DB_IF_hits_figure)[it]
                                names(DB_IF_hits_figure) 
                                PacBIOMatched
                              })

PB_Matcheddt <- rbindlist(PB_Matched)
PB_Matcheddt$ScoreType <- "Not Mapped"
PB_Matcheddt$ScoreType[PB_Matcheddt$score_UP_gb >= scorediffs[1]] <- "Very Good"
PB_Matcheddt$ScoreType[PB_Matcheddt$score_UP_gb < scorediffs[1]&PB_Matcheddt$score_UP_dl > scorediffs[2]] <- "Good"
PB_Matcheddt$ScoreType[PB_Matcheddt$score_UP_gb < scorediffs[2]] <- "Low"

PB_Matcheddt$ScoreType <- factor(PB_Matcheddt$ScoreType,c("Very Good","Good","Low","Not Mapped"))

names(PB_Matcheddt) <- make.unique(names(PB_Matcheddt))

PB_Matcheddt$SubsetType <- factor(PB_Matcheddt$SubsetType,names(DB_IF_hits_figure))

DB_IF_hits_dt$Group <- "All Abundances"
DB_IF_hits_dt$Group[grep("Size &",DB_IF_hits_dt$type)] <- "Size & TPM"
DB_IF_hits_dt$Group[grep("^TPM",DB_IF_hits_dt$type)] <- "TPM"
ra <- c(0,max(DB_IF_hits_dt$V1))
PB_ExclusiveBestPeak <- ggplot(DB_IF_hits_dt[Group=="All Abundances"],aes(V2,V1,fill=type,label=V1,group=type))+geom_bar(stat="identity",position="dodge")+theme(axis.text.x = element_text(angle=45,hjust=1,vjust=1))+geom_text(position=position_dodge(width = 1),size=3,vjust = 0.5,hjust=1,angle=90,color="black")+ylab("PACBIO IsoForms")+ylim(ra)+xlab("")+ggtitle("")
# PB_ExclusiveBestPeakC

PB_ExclusiveBestSizeTPM<- ggplot(DB_IF_hits_dt[Group=="Size & TPM"],aes(V2,V1,fill=type,label=V1,group=type))+geom_bar(stat="identity",position="dodge")+theme(axis.text.x = element_text(angle=45,hjust=1,vjust=1))+geom_text(position=position_dodge(width = 1),size=3,vjust = 0.5,hjust=1,angle=90,color="black")+ylab("PACBIO IsoForms")+ylim(ra)+xlab("")+ggtitle("with TPM CutOff &\nSize Restriction")
# PB_ExclusiveBestSizeTPMC
PB_ExclusiveBestSize<- ggplot(DB_IF_hits_dt[Group=="TPM"],aes(V2,V1,fill=type,label=V1,group=type))+geom_bar(stat="identity",position="dodge")+theme(axis.text.x = element_text(angle=45,hjust=1,vjust=1))+geom_text(position=position_dodge(width = 1),vjust = 0.5,size=3,hjust=1,angle=90,color="black")+ylab("PACBIO IsoForms")+ylim(ra)+xlab("")+ggtitle("with TPM CutOff")
# PB_ExclusiveBestSize


Dgg <- ggplot(PacBIO,aes(log10(Abundance_TPM)))+geom_density(fill="skyblue")+geom_vline(xintercept = log10(c(1,2,3)),color="orange",lty="solid",lwd=1)+ggplot2::coord_cartesian(xlim=c(-1,4))+annotate("text",x=c(log10(1:3)),y=c(0,0,0),label=paste("Cutoff:",1:3),hjust=0,vjust=1,angle=90)+ylab("Density")

pdf("PB_Exclusive_Peaks_MultiIsoforms.pdf",width=12,height=12)


DB_IF_hits_dt[,V1_rel :={V1/sum(V1)},.(Group,type)]
DB_FUN <- DB_IF_hits_dt[as.character(V2)!="No_Evidence ",{list(V1=sum(V1[V2!="No_Evidence"]),V1_rel=sum(V1_rel[V2!="No_Evidence"]))},.(type,Group)]

FractionCheck <- ggplot(DB_FUN,aes(V1,V1_rel,label=paste(type),pch=Group))+geom_point(size=3)+geom_text_repel()+xlab("Identified Isoforms")+ylab("Identified Fraction")
AbundanceBP <- ggplot(PB_Matcheddt,aes(log10(TPM),ScoreType,color=ScoreType))+geom_boxplot()+facet_wrap("SubsetType",ncol=1)+coord_cartesian(xlim=c(0,2))


DB_FUN <- DB_IF_hits_dt[as.character(V2)!="No_Evidence ",{list(V1=sum(V1[V2!="No_Evidence"]),V1_rel=sum(V1_rel[V2!="No_Evidence"]))},.(type,Group)]
p1 <- plot_grid(PB_ExclusiveBestPeak+theme(legend.position = "bottom")+guides(fill=guide_legend(nrow=3,byrow=TRUE,size=0.5,title = element_blank())),
          PB_ExclusiveBestSize+theme(legend.position = "bottom")+guides(fill=guide_legend(nrow=3,byrow=TRUE,size=0.5,title = element_blank())),
          PB_ExclusiveBestSizeTPM+theme(legend.position = "bottom")+guides(fill=guide_legend(nrow=3,byrow=TRUE,size=0.5,title = element_blank())),nrow=1,align = "hv")
p2 <- FractionCheck+ggtitle("Identified Fraction\n")+theme(legend.position = "right")+guides(pch=guide_legend(nrow=3,byrow=TRUE,size=1,title = element_blank()))

plot_grid(plot_grid(Dgg,AbundanceBP,ncol=1,rel_heights =  c(1,2),labels = c("A","B")  ),plot_grid(p1,p2,ncol=1,rel_heights = c(1,0.8),labels=c("C","D")),ncol=2,rel_widths = c(0.7,1))

# ggplot(DB_IF_hits_dt[],aes(V1,V1_rel,color=V2))+geom_point(size=10)+facet_wrap(type~Group)
dev.off()
system("open PB_Exclusive_Peaks_MultiIsoforms.pdf")


Overview_Counts_PB_UP <- DB_IF_hits_dt[!is.na(match(type,c("All","TPM > 3","Size & TPM > 3","Size Restricted"))),{
  sum(V1)
},.(Group,type,V2!="No_Evidence ")]

Overview_Counts_PB_UP$V2 <- factor(Overview_Counts_PB_UP$V2,c("TRUE","FALSE"),c("RNA & Protein","RNA"))
 

setnames(Overview_Counts_PB_UP,"V2","Evidence")

Overview_Counts_PB_UP[,V1_Rel :=V1/sum(V1),.(Group,type)]
Overview_Counts_PB_UP$type <- factor(Overview_Counts_PB_UP$type,c("All","Size Restricted","TPM > 3","Size & TPM > 3"))
Overview_Counts_PB_UP[,"TotalCount":=sum(V1),type]
Overview_Counts_PB_UP$type2 <- paste(Overview_Counts_PB_UP$type,paste("\nn =", Overview_Counts_PB_UP$TotalCount),sep = "\n")
Overview_Counts_PB_UP_FactorTable <- Overview_Counts_PB_UP[,1,.(type,type2)]
Overview_Counts_PB_UP_FactorTable <- Overview_Counts_PB_UP_FactorTable[order(type),]
Overview_Counts_PB_UP$type2 <- factor(Overview_Counts_PB_UP$type2,Overview_Counts_PB_UP_FactorTable$type2)

g <- ggplot(Overview_Counts_PB_UP, aes(x=1, y=V1_Rel*100, fill=Evidence,label=V1))+facet_wrap("type2",nrow = 1)
set.seed(12)
ggpie <- g +geom_col()+geom_text(size=3,check_overlap =F
) + coord_polar("x",clip = "off")+xlab("")+ylab("")
ggpie <- ggpie +geom_col(width=1) + 
  coord_polar("y",clip = "off")+xlab("")+ylab("")+
  geom_text(aes(label = paste(round(V1_Rel*100,1),"%")),size=5, position = position_stack(vjust = 0.5))+
  theme(legend.position = "bottom",axis.text.x = element_text(angle=45,hjust=1,vjust=1))

ggpie <- ggpie+theme_minimal()+theme(
  axis.ticks = element_blank(),
  panel.grid  = element_blank(),
  axis.text = element_blank())
ggpie
  
Results_CombinedGels_PBUP$RNAIdentifiedFraction$Pie <- ggpie


```


## Potentially new Splice Variants
```{r}
library(plotly)
RowWiseCompare$PickedGenes <- !is.na(match(RowWiseCompare$hgnc,c(names(GeneList),"CEBPB","NFKB1")))
subset_table<- RowWiseCompare[(selectedScore_UP>0.8|selectedScore_PB>0.8)&!grepl(";",ENSG)&selectedScore_UP<selectedScore_PB]
subset_table_MorePB<- RowWiseCompare[(selectedScore_UP>0.8|selectedScore_PB>0.8)&!grepl(";",ENSG)&as.numeric(Success_ISOFRAC_Peptides_PB)>as.numeric(Success_ISOFRAC_Peptides_UP)]
names(subset_table_MorePB) <- make.unique(names(subset_table_MorePB))

subset_table_MorePB$PB_better <- F
subset_table_MorePB[Success_PB_Peptides>Success_Uniprot_Peptides]$PB_better <- T

if(length(subset_table_MorePB$hgnc_symbol)==0){
 subset_table_MorePB$hgnc_symbol <- subset_table_MorePB$hgnc
}
gfu2_PacbioMORE <- ggplot(subset_table_MorePB[Success_PB_Peptides<Success_Uniprot_Peptides],aes(DM_PB_min,DM_UP_min,color=DM_min_categories,size=as.numeric(Success_PB_Peptides)/as.numeric(Success_ISOFRAC_Peptides_UP),label=paste(IsoFrac_MW,hgnc_symbol,PACBIO_ACC,Uniprot,sep="\n")))+geom_point()+xlab("IsoFrac-PACBIO MW [D]")+ylab("IsoFrac-UniProt MW [D]")+facet_grid("PB_better")
ggplotly(gfu2_PacbioMORE)

CandidateIsoforms <- subset_table_MorePB[Success_PB_Peptides>Success_Uniprot_Peptides]
CandidateIsoformsSel <- CandidateIsoforms[,{
  temp <- .SD
  UnfilUP<- Results_CombinedGels$Pairs_Unfiltered[ENSG==.BY$ENSG&type=="fw"]
  UnfilPB<- Results_CombinedGels_PacBio$Pairs_Unfiltered_PacBio[ENSG==.BY$ENSG&type=="fw"]
  
  list(DM_FC= log2(min(abs(as.numeric(DM_UP_min)))/min(abs(as.numeric(UnfilUP$DM_UP_min)))),Pep_FC= log2(as.numeric(Success_ISOFRAC_Peptides_PB)/as.numeric(Success_ISOFRAC_Peptides_UP)))
  
},.(hgnc_symbol,ENSG,IsoFrac_MW,DM_min_categories,PACBIO_ACC)]

fwrite(CandidateIsoformsSel,"Potential_NewSpliceVariants.txt",sep = "\t")


```

# Domain Analysis

```{r}

library(RSQLite)
pa <- "IsoFrac_0.84"
Cp <- Results_CombinedGels$Compressed_Pairs
Pf <- Results_CombinedGels$Pairs_Filtered

dhb <- lapply(1:length(TableListNormEnsHdbScanMW_Peaks),function(i){
  x <- TableListNormEnsHdbScanMW_Peaks[[i]]
  df <- x$hdbscanTables
  df$gel <- x$xname
  data.table(df)
})
dhb_dt <- rbindlist(dhb)
Cp <- as.data.table(Cp)
Pf <- data.table(Pf)
10^Cp[ENS=="ENSG00000007866"]$MW_lm_Compressed_
if(!file.exists("IsoFrac_with_Sequences_v1.rda")){
  # Compressed Proteins Table
  
  cPtemp <- Cp[,{
    temp <- .SD
    
    tempi <- temp[,{
      tempie <- .SD
      (apply(tempie,2,function(x){unlist(strsplit(x,";"))}))
    }
    ,.SDcols=c("gels","heights","clusters")]
    
    tempi <<- tempi
    
    if(length(dim(tempi)[1])==0){
      tempi <- as.list(tempi)
    }else{
      tempi <- as.list(data.frame(tempi))
      
    }
    tempi
  },.(Peak,n_Peaks,ENS,MW_lm_Compressed_)]
  cPtemp[ENS=="ENSG00000007866"]
  
  # extracting Sequences
  
  cPtemp2 <- cPtemp[,{
    cat("\r",.BY$ENS)
    gr <- .BY
    dsacl <-dhb_dt[EnsG==gr$ENS&(cl_2==gr$clusters|cl_2==0)]
    dsacl <- unique(dsacl)
    dsacl
  },.(gels,Peak,n_Peaks,heights,clusters,MW_lm_Compressed_,ENS)]

  # Map UniprotIsoforms
  
  cPtemp2$IsoFracID <- paste(cPtemp2$ENS,round(10^as.numeric(cPtemp2$MW_lm_Compressed_)))
  Pf$IsoFracID <- paste(Pf$ENSG,Pf$IsoFrac_MW)
  MatchVec1 <- match(cPtemp2$IsoFracID,Pf$IsoFracID)
  IsoFrac_with_Sequences <- cbind(cPtemp2,Pf[MatchVec1])
  IsoFrac_with_Sequences_for_Ellie <- IsoFrac_with_Sequences[Use==T,1,.(IsoFracID,IsoFrac_MW,hgnc_symbol,Uniprot,Seq)]
  save(IsoFrac_with_Sequences,file="IsoFrac_with_Sequences_v2.rda")
}else{
  load("IsoFrac_with_Sequences_v2.rda")
}
library(data.table)
if(!file.exists("DomainMappings.rda")){
  #### Reading Mart Results:
li <- list.files(pattern="^Mart",recursive = T,full.names = T)

lili <- lapply(li,function(x){
  x <- x
  fi <- fread(x)
  fi
  
})
DomainMappings <-rbindlist(lili,fill=T)
save(DomainMappings,file="DomainMappings.rda")
}else{
  load("DomainMappings.rda")
}
load("ProteinSequencesCategories2.rda")

#### Reading Mart Results:
li2 <- list.files(pattern="^PFAM",recursive = T,full.names = T)

lili2 <- lapply(li2[grepl("Archive",li2)],function(x){
  x <- x
  fi <- fread(x)
  fi
  
})

PFAM_Desc <- fread("PFAM_Desc_entries.tsv")
# Converting Pfam Position to sequence information:
if(!file.exists("PfamDomainMappings.rda")){
  PfamDomainMappings <-rbindlist(lili2,fill=T)
  PfamDomainMappings <- PfamDomainMappings[!is.na(ensembl_gene_id)]
  P <- (ProteinSequencesCategories$ProteinID)
  P_acc <- sapply(strsplit(P,"|",fixed=T),function(x){x[2]})
  table(is.na(P_acc))
  P_acc[is.na(P_acc)] <- P[is.na(P_acc)]
  P_acc <- gsub("_.*.","",P_acc)
  
  m1 <- match(PfamDomainMappings$uniprotswissprot,P_acc)
  ## Loading Database
  PfamDomainMappings$SEQUENCE <- ProteinSequencesCategories[m1]$SEQUENCE
  ###
  # Adding Pfam Descriptions
  ###
  PfamDomainMappings <- PfamDomainMappings[pfam!=""]
  pfamDes <- fread("PFAM_Desc_entries.tsv")
  PfamDomainMappings$Pfam_Description <- pfamDes[match(PfamDomainMappings$pfam,pfamDes$Accession)]$Name
  
  save(PfamDomainMappings,pfamDes,file="PfamDomainMappings.rda")
}else{
  load("PfamDomainMappings.rda")
}
  PfamDomainMappingsDomainSequence <- PfamDomainMappings[,{
    gr <- .BY
    DomainSequence <- substr(gr$SEQUENCE,gr$pfam_start,gr$pfam_end)
    SequenceBefore <- substr(gr$SEQUENCE,1,gr$pfam_start)
    SequenceAfter <- substr(gr$SEQUENCE,gr$pfam_end,nchar(gr$SEQUENCE))
    list(DomainSequence=DomainSequence,SequenceBefore=SequenceBefore,SequenceAfter=SequenceAfter)
  },.(ensembl_gene_id,SEQUENCE,uniprotswissprot,pfam,pfam_start,pfam_end,Pfam_Description)]

PfamDomainMappingsDomainSequence <- PfamDomainMappings 
DESC_m1 <- match(PfamDomainMappingsDomainSequence$pfam,PFAM_Desc$Accession) 
PfamDomainMappingsDomainSequence$Pfam_Description <- PFAM_Desc$Name[DESC_m1]
PfamDomainMappingsDomainSequence$Pfam_Description <- PFAM_Desc$Name[DESC_m1]
save(PfamDomainMappingsDomainSequence,file="PfamDomainMappingsDomainSequence.rda")

####
# Mapping Semitryptic Information
####
load("SemiTrypticInfo.rda")
SemiTrypticInfo <- SemiTrypticInfo[grep("^REV_",ENSG,invert=T)]
SemiTrypticInfo <- unique(SemiTrypticInfo)
SemiTrypticInfo <- SemiTrypticInfo[,{
  if(any(SearchSpecificity=="Tryptic")){
    .SD[SearchSpecificity=="Tryptic"]
  }else{
    .SD
  }
},Sequence]
SemiTrypticInfo$Category <- "UNKNOWN"
SemiTrypticInfo[grepl("K$|R$",Sequence)&AA_after!="-"]$Category <- "tryptic"

SemiTrypticInfo[AA_after=="-"&(AA_Before=="K"|AA_Before=="R")]$Category <- "cterm"
SemiTrypticInfo[AA_Before=="-"&grepl("K$|R$",Sequence)]$Category <- "nterm"
table(SemiTrypticInfo$Category)
# SemiTrypticInfo[!(AA_Before!="K"&AA_Before!="R"&AA_Before!="-")]$Category <- "new_nterm"
SemiTrypticInfo[SearchSpecificity=="Semitryptic"&(!grepl("K$|R$",Sequence))&AA_after!="-"&AA_after!="K"&AA_after!="R"&SearchSpecificity!="Tryptic"&(AA_Before=="K"|AA_Before=="R")]$Category <- "new_cterm"

SemiTrypticInfo[SearchSpecificity=="Semitryptic"&AA_Before!="K"&AA_Before!="R"&AA_Before!="-"&AA_Before!="M"&SearchSpecificity!="Tryptic"]$Category <- "new_nterm"
SemiTrypticInfo[Category=="new_cterm"&SearchSpecificity=="Tryptic"]
SemiTrypticInfo[Category=="cterm"&SearchSpecificity=="Semitryptic"]
SemiTrypticInfo[Sequence=="AASAPAKE"]
table(SemiTrypticInfo$Category)
SemiTrypticInfo[Category=="UNKNOWN"]
save(SemiTrypticInfo,file="SemiTrypticInfo2.rda")

####  
# Starting compilation and analysis:
###
library(parallel)
library(ggplot2)
if(!file.exists("IsoFrac_Domain_SequenceAnalysis_woSemi.rda")){
  cl <- makeCluster(9)
  clusterExport(cl,c("IsoFrac_with_Sequences","ProteinSequencesCategories","PfamDomainMappings"))
  clusterExport(cl,"PfamDomainMappingsDomainSequence")
  clusterExport(cl,"SemiTrypticInfo")
  ENSGvec <- unique(IsoFrac_with_Sequences$ENSG)
  ENSGvec <- ENSGvec[!is.na(ENSGvec)]
  xensg <- ENSGvec[1]
  xensg <- "ENSG00000007866"
  ENSGvec <- GeneList
  LaFun <- parLapply(cl,ENSGvec[!is.na(ENSGvec)],function(xensg){
    # LaFun <- lapply(ENSGvec,function(x){
    
    library(data.table)
    library(Peptides)
    xensg <<- xensg
    # xensg <- "ENSG00000007866"
    IsoFrac_with_Sequences[hgnc_symbol=="NFKB1"]
    
    IsoFrac_with_Sequences_DomainMatch <- IsoFrac_with_Sequences[ENSG==xensg,{
      
      temp <- .SD
      gr <- .BY
      gr2 <- gr
      temp2 <- temp
      if(0){
        temp <- temp2
        gr <- gr2
      }
      
      USxdt <- NULL
      ({
        cat(paste("\r",.GRP,unlist(gr$Uniprot)))
        # print(gr)
        if(!is.na(gr$Uniprot)){
          gr$Uniprot <- strsplit( as.character(gr$Uniprot),";")
          US <- sapply(unlist(gr$Uniprot),function(x){
            g <- unlist(strsplit(x,"|",fixed=T))
            g[2]
          })
        }else{
          US <- gr$ENS
        }
        
        x <- US[1]
        USx <- lapply(US,function(x){
          # stop()
          # x <<- x
          if(grepl("_",x)){
            # x <<- x
            xs <- unlist(strsplit(x,"_"))
            CutPositions <- as.numeric(xs[-1])
            x <- xs[1]
          }else{
            CutPositions <- NULL
          }
          if(!grepl("ENSG",x)){
            seqi <- ProteinSequencesCategories[match(x,ID)]$SEQUENCE
            
          }else{
            seqi <- ProteinSequencesCategories[match(x,ENSG)]$SEQUENCE
            # seqi <-seqi[nchar(seqi)==max(nchar(seqi))]
          }
          
          humpi <- lapply(temp$Seq,function(se){
            re <- regexec(se,seqi)
            re <- re[!sapply(re,is.na)]
            if(length(re)>0){
              unlist(lapply(re[re!=-1],function(it){
                it:(it+nchar(se))
              }))
            }else{
              NULL
            }
            
          })
          humpi <- unlist(humpi)
          # Map Domains
          if(!grepl("ENSG",x)){
            MappingType <- "Specific"
            
            dm <- PfamDomainMappings[ensembl_gene_id==gr$ENS&uniprotswissprot==x]
          }else{
            MappingType <- "ENSG_Specific"
            
            dm <- PfamDomainMappings[ensembl_gene_id==gr$ENS]
          }
          if(dim(dm)[1]==0){
            MappingType <- "Unspecific"
            
            dm <- PfamDomainMappings[ensembl_gene_id==gr$ENS]
            
          }
          
          ### Checking Direct Sequence Domain Overlap:
          PFAM_Sequences <- PfamDomainMappingsDomainSequence[ensembl_gene_id==gr$ENS]
          PFAM_Sequences$DomainSequence <- PFAM_Sequences$SEQUENCE
          PFAM_Sequences[,DomainSequence:={
            substring(.BY$SEQUENCE,pfam_start,pfam_end)
          },.(SEQUENCE,pfam_start,pfam_end)]
          CheckFun <- lapply(temp$Seq,function(se){
            SimpleMatch <- grepl(se,PFAM_Sequences$DomainSequence)
            AgrepMatch <- agrepl(se,PFAM_Sequences$DomainSequence)
            if(any(AgrepMatch)){
              # se <<- se
              sumpi <- aregexec(se,PFAM_Sequences$DomainSequence)
              Replaced <- regmatches(PFAM_Sequences$DomainSequence,sumpi)[AgrepMatch]
              # MatchLength <- sapply(sumpi[AgrepMatch],function(x){
              #   a <- attributes(x)$match.length
              #   
              # })
              Replaced <- sapply(Replaced,paste,collapse="_")
              
            }else{
              Replaced <- ""
              MatchLength <- NA
            }
            ####
            # Grep in Protein Sequence
            ####
            ProteinPosition <- paste(unlist(aregexec(se,seqi)),collapse=";")
            try(clseq <- temp[Seq==se&round(10^as.numeric(temp$MW_lm_Compressed_))==gr$IsoFrac_MW]$cl_2)
            try(gelseq <- temp[Seq==se&round(10^as.numeric(temp$MW_lm_Compressed_))==gr$IsoFrac_MW]$gels)
            
            li <- data.table(Pfam_cat1=PFAM_Sequences$pfam[SimpleMatch],Pfam_cat2=PFAM_Sequences$pfam[AgrepMatch],SimpleMatch=SimpleMatch,
                             AgrepMatch=AgrepMatch,cterminalSpecific= any(!grepl("K$|R$",se)),sequence=se,Replaced=Replaced,
                             DomainStart_cat1=PFAM_Sequences$pfam_start[SimpleMatch],DomainEnd_cat1=PFAM_Sequences$pfam_end[SimpleMatch],
                             DomainStart_cat2=PFAM_Sequences$pfam_start[AgrepMatch],DomainEnd_cat2=PFAM_Sequences$pfam_end[AgrepMatch],
                             ProteinPosition=ProteinPosition,HDBscanCluster=clseq,gel=gel
                             
            )
            # litem p <<- li
            li
            
          })
          # CheckFun2 <<- CheckFun
          CheckFun_dt <- rbindlist(CheckFun)
          # CheckFun_dt <- CheckFun_dt[!(is.na(CheckFun_dt$Pfam_cat1)&is.na(CheckFun_dt$Pfam_cat2))]
          # CheckFun_dt <- CheckFun_dt[!(SimpleMatch==FALSE&AgrepMatch==FALSE)]
          
          
          
          CheckFun_dt <- unique(CheckFun_dt)
          
          # humpitemp <<- humpi
          # dmtemp <<- dm
          # 1. Checking direct overlap:
          # if(0){
          #   if(dim(dm)[1]>0){
          #     dmList <- apply(dm,1,function(x){
          #       # StartEnd <<- c(x[names(dm)=="interpro_start"],x[names(dm)=="interpro_end"])
          #       StartEnd <- c(x[names(dm)=="pfam_start"],x[names(dm)=="pfam_end"])
          #       
          #       StartEnd <- as.numeric(StartEnd)
          #       Distance <- StartEnd[1]:StartEnd[2]
          #       # gplots::venn(list(Distance,humpi))
          #       # plot(sort(humpi),col=2)
          #       # points(sort(Distance),col=3)
          #       
          #       sc <- any(!is.na(match(humpi,Distance)))
          #       if(all(!sc)){
          #         cd <-  sapply(StartEnd,function(x){min(abs(x-humpi))})
          #       }else{
          #         cd <- 0
          #       }
          #       # if(sc==FALSE){
          #       #   dm <<- dm
          #       #   x <<- x
          #       #   humpi <<- humpi
          #       #   stop()
          #       # }
          #       list(SequenceCovered= sc,ClosestDistance=min(cd),DomainStart=StartEnd[1],DomainEnd=StartEnd[2],DomainLength=diff(StartEnd))
          #     })
          #     dmdt <- rbindlist(dmList)
          #     dmdt$InterPro <- dm$interpro
          #     dmdt$interpro_short_description <- dm$interpro_short_description
          #     dmdt$Pfam <- dm$pfam
          #     dmdt$PfamDescription <- dm$Pfam_Description
          #     
          #     
          #     
          #     dmdt$Protein <- x
          #     dmdt$SequenceLength <- nchar(seqi)
          #     
          #   }else{
          #     
          #     dmdt <- data.table(SequenceCovered= F,ClosestDistance=as.double(NA),DomainStart=as.double(NA),DomainEnd=as.double(NA),DomainLength=as.double(NA)) 
          #     
          #     dmdt$InterPro <- dm$interpro
          #     dmdt$interpro_short_description <- dm$interpro_short_description
          #     dmdt$Pfam <- "No PFAM"
          #     dmdt$PfamDescription <- "No PFAM"
          #     dmdt$Protein <- x
          #     dmdt$SequenceLength <- nchar(seqi)
          #     
          #   }
          # }
          CheckFun_dt$Protein <- x
          CheckFun_dt$SequenceLength <- nchar(seqi)
          CheckFun_dt$MappingType <- MappingType
          
          
          
          
          CheckFun_dt
        })
        
        
        # print("HUI")
        USxdt <-rbindlist(USx)
        
      })
      
      
      USxdt
      
    },.(ENS,IsoFrac,IsoFrac_MW,IsoFrac,Uniprot,gels)]
    IsoFrac_with_Sequences_DomainMatch[ENS=="ENSG00000007866"]$IsoFrac
    # Mapping PfamDescriptions
    
    IsoFrac_with_Sequences_DomainMatch <- unique(IsoFrac_with_Sequences_DomainMatch)
    
    IsoFrac_with_Sequences_DomainMatch$PfamDescription_cat1 <- PfamDomainMappingsDomainSequence[match(IsoFrac_with_Sequences_DomainMatch$Pfam_cat1,pfam)]$Pfam_Description
    IsoFrac_with_Sequences_DomainMatch$PfamDescription_cat2 <- PfamDomainMappingsDomainSequence[match(IsoFrac_with_Sequences_DomainMatch$Pfam_cat2,pfam)]$Pfam_Description
    IsoFrac_with_Sequences_DomainMatch$Category <- "tryptic"
    
    IsoFrac_with_Sequences_DomainMatch$CanonicalPeptidePosition <- NULL
    IsoFrac_with_Sequences_DomainMatch$Max_Protein_MW <- NULL
    IsoFrac_with_Sequences_DomainMatch[,c("Max_Protein_MW"):={
      gr <- .BY
      # print(gr)
      SE <- ProteinSequencesCategories[ENSG==gr$ENS]
      if(dim(SE)[1]==0){
        SEmw <- as.double(NA)
      }else{
        SEmw <- SE[MW==max(MW)]$SEQUENCE
      }
      mw(SEmw[1])
    },.(ENS,sequence)]
    
    
    
    IsoFrac_with_Sequences_DomainMatch
  })
  names(LaFun) <- ENSGvec[!is.na(ENSGvec)]
  AllFun <- rbindlist(LaFun)

  AllFun$DomainCategory <- "Outside Domain"
  AllFun$DomainCategory[!is.na(AllFun$PfamDescription_cat2)] <- "Inside Domain"
  AllFun$ClusterType <- "Noise"
  AllFun$ClusterType[AllFun$HDBscanCluster!=0] <- "Clustered"
  table(AllFun$Category)
  IsoFrac_Domain_SequenceAnalysis <- AllFun
  
  IsoFrac_Domain_SequenceAnalysis[!is.na(Pfam_cat2)]
  table(IsoFrac_Domain_SequenceAnalysis$Category)
  unique(IsoFrac_Domain_SequenceAnalysis[ENS=="ENSG00000007866"]$IsoFrac)
  save(IsoFrac_Domain_SequenceAnalysis,file="IsoFrac_Domain_SequenceAnalysis_woSemi.rda")
  
}else{
  # load("IsoFrac_Domain_SequenceAnalysis_v1.rda")
  
  load("IsoFrac_Domain_SequenceAnalysis_woSemi.rda")
}

fwrite(IsoFrac_Domain_SequenceAnalysis,"IsoFrac_Domain_SequenceAnalysis.txt",sep="\t")
# table(AllFun$ClusterType)
# unique(grep(0,AllFun$HDBscanCluster,value = T))
IsoFrac_Domain_SequenceAnalysis["ENSG00000185049"==ENS]$
NELFA
IF_D_Compiled <- IsoFrac_Domain_SequenceAnalysis[,{
  tc <- table(unlist(Category))
  list(na=names(tc),c=c(tc))
},.(ENS,IsoFrac_MW)]

IFfun <- IF_D_Compiled[c!=0,{
  cn <- any(grepl("new_n",na,fixed=T))
  cc <- any(grepl("new_c",na,fixed=T))

  ca <- c("alternative\nn-term","alternative\nc-term")[c(cn,cc)]
  paste(ca,collapse="\n")

},.(ENS,IsoFrac_MW)]
table(IFfun$V1)
IFfunSum <- IFfun[,.(Peaks=length(IsoFrac_MW),ENSG=length(unique(ENS))),V1]
IFfunSum$V1[IFfunSum$V1==""] <- "tryptic"

plot_collect_peptides$PeaksWithAlternative_ncterms <- ggplot(IFfunSum,aes(V1,Peaks))+geom_bar(stat="sum")+ylab("Isofrac Isoforms")+xlab("")+theme(legend.position = "none")


SumOfPeps <- IsoFrac_Domain_SequenceAnalysis[,{
  length(unique(sequence))
},.(Category,ClusterType)]
SumOfPeps[,V1_rel:={V1/sum(V1)},.(ClusterType)]

hyper_equal_or_more <- function(x,m,n,k){
  dhyper(x,m,n,k) + phyper(x,m,n,k,lower.tail=F)
}

SumOfPeps[,p:={
  temp <<- .SD
  gr <<- .BY
  white <- temp$V1
  AllWhite <- sum(SumOfPeps[Category==gr$Category]$V1)
  AllBlack <- sum(SumOfPeps$V1)-AllWhite
  Picked <- sum(SumOfPeps[ClusterType==gr$ClusterType]$V1)
  hyper_equal_or_more(white,AllWhite,AllBlack,Picked)
  
  
},.(ClusterType,Category)]
SumOfPeps$p.adjust <- p.adjust(SumOfPeps$p,"BH")
SumOfPeps$p.adjust[SumOfPeps$p.adjust>0.01] <- NA
library(cowplot)
library(ggplot2)
library(ggrepel)

plotDomains <- function(x,IsoFrac_with_Sequences_DomainMatch=NULL){
  # x <<- x
  hgnc <- SemiTrypticInfo[ENSG==x]$hgnc
  selectedP <<- IsoFrac_with_Sequences_DomainMatch[ENS==x,]
  selectedP_terminal <- selectedP[Category=="new_nterm"|Category=="new_cterm"]
  if(dim(selectedP)[1]>0){
    hgnc <- paste(hgnc,round(selectedP$Max_Protein_MW[!is.na(selectedP$Max_Protein_MW)][1]))
    g <- ggplot(selectedP,
                aes(x=DomainStart_cat2,xend=DomainEnd_cat2,y=IsoFrac,label=PfamDescription_cat2,
                    group=IsoFrac,yend=IsoFrac,color=PfamDescription_cat2))+
      geom_segment(lwd=3)+
      geom_vline(xintercept = selectedP_terminal$ProteinPosition,lty="dotted",color="orange")+geom_text(selectedP_terminal,aes(x=ProteinPosition,y=IsoFrac,label=paste("---",Category,"---")),color="black",angle=90)+ggtitle(hgnc)+coord_cartesian(xlim=c(0,max(selectedP$SequenceLength,na.rm=T)))
    
  }else{g <- NULL}
  g
}
# plotDomains(x="ENSG00000174444",IsoFrac_with_Sequences_DomainMatch=LaFun$ENSG00000174444)







```


## Second Semitryptic Barplots
```{r}
setwd("E:/MaxQuant/HZ/20220114_Isofrac_v2/IsoFrac_20230217_SimilaritiesTesting_4")
# ENSGvec <- unique(IsoFrac_with_Sequences$ENSG)
library(parallel)
load("ProteinSequencesCategories2.rda")
load("IsoFracPeptides.rda")
# IsoFracPeptides <- unique(IsoFrac_with_Sequences$Seq)
# save(IsoFracPeptides,file="IsoFracPeptides.rda")
print("Generating Clusters")
# print(hu)>

print("EXporting to host")
ProteinSequencesCategories <<- ProteinSequencesCategories
print(exists("ProteinSequencesCategories"))
if(!file.exists("SequenceMappings.rda")){
  cl <- makeCluster(70)
print(cl)
  clusterExport(cl,c("ProteinSequencesCategories"))
  print("Starting Loop")
  SequenceMappings <- parLapply(cl,IsoFracPeptides,function(x){
    tempiall <- NULL
    try({
      library(data.table)
      # if(!exists("ProteinSequencesCategories")){
      #   load("ProteinSequencesCategories2.rda")
      #   ProteinSequencesCategories <<- ProteinSequencesCategories
      # }
      ProteinSequencesCategories <<- ProteinSequencesCategories
      g <- regexec(x,ProteinSequencesCategories$SEQUENCE,fixed=T)
      sel <- which(g!=-1)
      gsel <- g[sel]
      
      tempi <- lapply(1:length(sel),function(i){
        sise <- ProteinSequencesCategories$SEQUENCE[sel[i]]
        tempi <- gsel[i]
        startposition <- tempi[[1]]
        endposition <- (startposition+(attributes(startposition)$match.length-1))
        AA_Before <- substr(sise,startposition-1,startposition-1)
        AA_After <- substr(sise,endposition+1,endposition+1)
        data.table(AA_Before,AA_After,startposition,endposition,ProteinSequence=sise,
                   # Peptide=ProteinSequencesCategories$SEQUENCE[sel[i]],
                   ProteinID=ProteinSequencesCategories$ProteinID[sel[i]],
                   ENSG=ProteinSequencesCategories$ENSG[sel[i]])
      })
      tempiall <- rbindlist(tempi)
      tempiall$Peptide=x
      
    })
    
    
    
    tempiall
  })
  stopCluster(cl)
  print("finished LOOP")
  save(SequenceMappings,file="SequenceMappings.rda")
}else{
  load("SequenceMappings.rda")
  SequenceMappings_dt <- rbindlist(SequenceMappings)

}

SequenceMappings_dt[,nchar:=nchar(.BY$ProteinSequence),ProteinSequence]
IsoFrac_with_Sequences_PeptidesPosUniprot <- IsoFrac_with_Sequences[,{
  gr <- .BY
  tempsel <- SequenceMappings_dt[Peptide==gr$Seq,]
  tempsel[nchar(ProteinSequence)==max(nchar(ProteinSequence))][1]

}
  
,.(Seq,IsoFracID,IsoFrac_MW,hgnc_symbol)]
save(IsoFrac_with_Sequences_PeptidesPosUniprot,file="IsoFrac_with_Sequences_PeptidesPosUniprot.rda")

length(unique(IsoFrac_with_Sequences_PeptidesPosUniprot$IsoFracID))
### Defining Terms
# Tryptics:
SequenceMappings_dt$PeptideEndsWith <- substr(SequenceMappings_dt$Peptide,nchar(SequenceMappings_dt$Peptide),nchar(SequenceMappings_dt$Peptide))
tryptic <- (SequenceMappings_dt$AA_Before=="K"|SequenceMappings_dt$AA_Before=="R")& # AA before is K or R
  (SequenceMappings_dt$AA_After!=""&SequenceMappings_dt$AA_After!="*") & # AA after is a regular AA
  ( SequenceMappings_dt$PeptideEndsWith=="K"|SequenceMappings_dt$PeptideEndsWith=="R") # Peptide ends with K or R
table(tryptic)

# ntermis:
nterm <- (SequenceMappings_dt$AA_Before==""|(SequenceMappings_dt$AA_Before=="M"&SequenceMappings_dt$startposition==2))& # AA before empty or an M and Position is at 2
  (SequenceMappings_dt$AA_After!=""&SequenceMappings_dt$AA_After!="*") & # AA after is a regular AA
  ( SequenceMappings_dt$PeptideEndsWith=="K"|SequenceMappings_dt$PeptideEndsWith=="R") # Peptide ends with K or R
table(nterm)
# ctermis:
cterm <- (SequenceMappings_dt$AA_Before=="K"|(SequenceMappings_dt$AA_Before=="R"))& # AA before is a regular tryptic Peptide and ends with K or R
  (SequenceMappings_dt$AA_After=="" )  # AA after is empty
# ( nchar(SequenceMappings_dt$PeptideEndsWith=="K")|nchar(SequenceMappings_dt$PeptideEndsWith=="R")) # Peptide ends with K or R
# Semitryptics
new_nterm <- (SequenceMappings_dt$AA_Before!=""&SequenceMappings_dt$AA_Before!="K"&SequenceMappings_dt$AA_Before!="R"&SequenceMappings_dt$AA_Before!="*"&SequenceMappings_dt$startposition>10)& # AA before is not empty and not an K or R and the startposition is more than 10 AA from the DB entry aways
  (SequenceMappings_dt$AA_After!=""&SequenceMappings_dt$AA_After!="*") & # AA after is a regular AA
  ( SequenceMappings_dt$PeptideEndsWith=="K"|SequenceMappings_dt$PeptideEndsWith=="R") # Peptide ends with K or R
table(new_nterm)

new_cterm <- (SequenceMappings_dt$AA_Before=="K"|SequenceMappings_dt$AA_Before=="R") &# AA before is K or R  
  (SequenceMappings_dt$AA_After!=""&SequenceMappings_dt$AA_After!="*") & # AA after is a regular AA and not empty and not a star.
  ( SequenceMappings_dt$PeptideEndsWith!="K"&SequenceMappings_dt$PeptideEndsWith!="R")&((nchar(SequenceMappings_dt$ProteinSequence)-SequenceMappings_dt$endposition)>10) # Peptide must not end with K or R end endpostion is more than 10 AA away from regular entry
table(new_cterm)
Peptides <- lapply(list(tryptic,cterm,nterm,new_nterm,new_cterm),function(x){
  # x <<- x
  unique(SequenceMappings_dt$Peptide[x])
})

PeptidesProteins <- lapply(list(tryptic,cterm,nterm,new_nterm,new_cterm),function(x){
  # x <<- x
  unique(paste(SequenceMappings_dt$Peptide[x],SequenceMappings_dt$ProteinSequence[x]))
})

names(PeptidesProteins) <- c("tryptic","C","N","new N","new C")
names(Peptides) <- c("tryptic","C","N","new N","new C")
grep("K$|R$",Peptides$tryptic,value=T,invert = T)
pdf("PeptideCategories_Isofrac.pdf")
gplots::venn(Peptides)
mtext("UniquePeptides")
gplots::venn(PeptidesProteins)
mtext("Unique Peptide Protein Pairs")

dev.off()
system("open PeptideCategories_Isofrac.pdf")
SequenceMappings_dt$Category <- "Undefined"
SequenceMappings_dt$Category[tryptic] <- "tryptic"
SequenceMappings_dt$Category[nterm] <- "N"
SequenceMappings_dt$Category[cterm] <- "C"
SequenceMappings_dt$Category[new_nterm] <- "new N"
SequenceMappings_dt$Category[new_cterm] <- "new C"
table(SequenceMappings_dt$Category)

# Mapping Relevant Informations from IsoFrac_PeptidePositionsUnique
if(!file.exists("SequenceMappings_dt_BIG.rda")){
  SequenceMappings_dt_BIG <- SequenceMappings_dt[,{
    cat("\r",.GRP)
    PE <- .BY$Peptide
    temp <- IsoFrac_PeptidePositionsUnique_combined[Seq==PE]
    temp3 <- NULL
    if(dim(temp)[1]>0){
      temp2 <- .SD
      temp3 <- temp2[,as.list(temp),by=names(temp2)]
    }
    temp3
  },.(Peptide)]
  save(SequenceMappings_dt_BIG,file="SequenceMappings_dt_BIG.rda")
}else{
  load("SequenceMappings_dt_BIG.rda")
}
SequenceMappings_dt[grep("NFKB1",ProteinID,fixed=T)]


# Generate New Mappings based on semitryptic Postions
SequenceMappings_dt_BIG[,SemiTrypticPeaks:= as.numeric(any(grepl("new",Category)) ),.(ENSG,IsoFrac_MW)]

SequenceMappings_dt_Regular <- SequenceMappings_dt_BIG[SemiTrypticPeaks==1&LI_COUNT<0.01&PEPM<0.01,{
  cat("\r",.GRP)
  temp <- .SD
  temp$ProteinPosition <- as.numeric(temp$ProteinPosition)
  temp <- unique(temp)
  tempuni <- temp[,unique(.SD),.SDcols = c("ProteinPosition","ProteinLength","Category","Seq")]
  tempuni$ProteinPosition[tempuni$ProteinPosition<0] <- 1
  tempuni$OriginalProteinLength <- tempuni$ProteinLength
  # Correcting protein length on minimal Datapoint
  if(any(tempuni$Category=="new_nterm")){
    minProteinposition <- min(temp$ProteinPosition,na.rm=T)
    
    tempuni$ProteinPosition <- tempuni$ProteinPosition-minProteinposition+1
    tempuni$ProteinLength <- tempuni$ProteinLength-minProteinposition
  }
  if(any(tempuni$Category=="new_cterm")){
    maxProteinposition <-temp[ProteinPosition== max(ProteinPosition,na.rm=T)]
    maxProteinposition <- (nchar(maxProteinposition$Seq)+maxProteinposition$ProteinPosition)[1]-1
    Leli <- max(temp$ProteinLength)-maxProteinposition
    tempuni$ProteinPosition <- tempuni$ProteinPosition-Leli
    tempuni$ProteinLength <- tempuni$ProteinLength-Leli
  }
  
  tempuni$ProteinPosition <- tempuni$ProteinPosition/tempuni$ProteinLength# Normalising Protein Position to length
  tempuni$ProteinLength <- as.double(tempuni$ProteinLength)
  tempuni$score_ref_gbm <- max(as.numeric(temp$score_ref_gbm),na.rm=T)
  tempuni$Height <- median(as.numeric(temp$Height),na.rm=T)
  tempuni$ProteinLength <- unique(tempuni$ProteinLength)[1]
  tempuni
},.(ENSG,IsoFrac_MW)]

SequenceMappings_dt_Regular$Cut <- cut(round(as.numeric(SequenceMappings_dt_Regular$IsoFrac_MW),1),4)
SequenceMappings_dt_Regular[,MassRange:=paste(range(round(as.numeric(IsoFrac_MW)/1000)),"kD",collapse=" to "),Cut]
SequenceMappings_dt_Regular[,MedianMass:=median(as.numeric(IsoFrac_MW)/1000,na.rm=T),Cut]

SequenceMappings_dt_RegularMedian <- SequenceMappings_dt_Regular[,{median(ProteinPosition)},.(IsoFrac_MW,Category)]
SequenceMappings_dt_Regular$CategoryFac <- factor(SequenceMappings_dt_Regular$Category,c("C","N","new C","new N","tryptic","Undefined"),c("C","new C","N","new N","tryptic","Undefined"))
SequenceMappings_dt_Regular$Group <- SequenceMappings_dt_Regular$Category
SequenceMappings_dt_Regular$Group <- gsub("new ","",SequenceMappings_dt_Regular$Group)
SequenceMappings_dt_Regular$Group[!grepl("^C$|^N$",SequenceMappings_dt_Regular$Group)] <- "Other"
plot_collect_peptides$MetaGene_v1 <- ggplot(SequenceMappings_dt_Regular[!grepl(";",Category)],aes(as.numeric(ProteinPosition),fill=CategoryFac))+geom_density(alpha=0.5,bw=0.05)+xlab("Protein Position [Fraction of estimated Protein Length]")+geom_vline(xintercept = 0.5,color="orange",lwd=2)+facet_wrap("Category",scales = "free_y",nrow = 1)
plot_collect_peptides$MetaGene_v2 <- ggplot(SequenceMappings_dt_Regular[!grepl(";",Category)],aes(as.numeric(ProteinPosition),fill=CategoryFac))+geom_histogram(alpha=0.5,bins=20)+xlab("Protein Position [Fraction of estimated Protein Length]")+geom_vline(xintercept = 0.5,color="orange",lwd=2)

plot_collect_peptides$MetaGene_v1 <- ggplot(SequenceMappings_dt_Regular[!grepl(";",Category)],aes(as.numeric(ProteinPosition),fill=CategoryFac))+geom_density(alpha=0.5,bw=0.05)+xlab("Protein Position [Fraction of estimated Protein Length]")+geom_vline(xintercept = 0.5,color="orange",lwd=2)+facet_wrap("Category",scales = "free_y",nrow = 1)

PeakAllRatios <- SequenceMappings_dt_Regular[,{
  temp <- .SD
  ca <- unique(temp$Category)
  bulk <- lapply(ca[ca!="tryptic"],function(x){
    x <<- x
    reference <- temp[Category==x]
    bulk <- temp[Category=="tryptic"]
    bulk$PositionRatio= median(reference$ProteinPosition,na.rm=T)/median(bulk$ProteinPosition,na.rm=T)
    bulk$PositionDelta= median(reference$ProteinPosition,na.rm=T)-median(bulk$ProteinPosition,na.rm=T)
    
    bulk$RatioType=x
    bulk
  })
  bulk <- rbindlist(bulk)
  bulk
},.(ENSG,IsoFrac_MW)]



IF_D_Compiled <- SequenceMappings_dt_BIG[,{
  tc <- table(unlist(Category2))
  list(na=names(tc),c=c(tc))
},.(ENSG,IsoFrac_MW)]

IFfun <- IF_D_Compiled[c!=0,{
  cn <- any(grepl("new_n",na,fixed=T))
  cc <- any(grepl("new_c",na,fixed=T))
  
  ca <- c("alternative\nn-term","alternative\nc-term")[c(cn,cc)]
  paste(ca,collapse="\n")
  
},.(ENSG,IsoFrac_MW)]
table(IFfun$V1)
IFfunSum <- IFfun[,.(Peaks=length(IsoFrac_MW),ENSG=length(unique(ENSG))),V1]
IFfunSum$V1[IFfunSum$V1==""] <- "tryptic"

plot_collect_peptides$PeaksWithAlternative_ncterms <- ggplot(IFfunSum,aes(V1,Peaks,label=Peaks))+geom_bar(stat="sum")+ylab("Isofrac Isoforms")+xlab("")+theme(legend.position = "none")+geom_text_repel(direction ="y",color="darkorange")



```


### Plotting selected Genes
```{r}

## Loading Required Data
if(!exists("IsoFrac_Domain_SequenceAnalysis")){
  load("IsoFrac_Domain_SequenceAnalysis.rda") # from Domain_Analysis
  load("../FinalViewer/ProteinSequencesCategories2.rda") # from IsoFrac Pre Scripts (Hypatia Server) 
}
library(Peptides)
library(data.table)
library(ggplot2)
# Mapping Peptides
GeneList <- list(
  SMAD4=c("ENSG00000141646",3),
  TEAD3=c("ENSG00000007866"),
  FOXC2=c("ENSG00000176692"),
  RPL3=c("ENSG00000100316"),
  RRP1B=c("ENSG00000160208"),
  NELFA=c("ENSG00000185049"),
  HLTF=c("ENSG00000071794"),
  NOB1=c("ENSG00000141101"),
  SPAST=c("ENSG00000021574"),
  DDX28=c("ENSG00000182810"),
  INCENP=c("ENSG00000149503"),
  GIT1=c("ENSG00000108262"),
  SLAIN2=c("ENSG00000109171"),
  GGA2=c("ENSG00000103365"),
  USP1=c("ENSG00000162607"),
  POLDIP3=c("ENSG00000100227"),
  LRCH4=c("ENSG00000077454"),
  CTTNBP2N=c("ENSG00000143079"),
  GALNS=c("ENSG00000141012"),
  AFAP1=c("ENSG00000196526"),
  NCOA5=c("ENSG00000124160"),
  GART=c("ENSG00000159131"),
  CEBPB=c("ENSG00000172216")
)
```
### Subfigures 2

```{r}
library(grid)
library(msa)
### Western Blots
if(!exists("CollectGG_v2")){
  load("WB_CollectGG_v2.rda")
}
names(CollectGG_v2)
input_i <- CollectGG_v2[[which(names(CollectGG_v2)=="TEAD3_RPE1")]]
gglist1 <- input_i$cterm
gglist2 <- input_i$nterm

#### EF

hu <- plot_grid(gglist1$allinone,gglist2$allinone,nrow=1,labels=c("E","F"))

Pe <- IF_DB_Peptidemapping("ENSG00000007866",IsoFrac_Domain_SequenceAnalysis,ProteinSequencesCategories,invertOrder = T,allgels=T,facetPlot = F)


SequenceInfoTead3 <- plot_grid(Pe$MSAPlot,Pe$MSAPlot2,ncol=1,align="hv",rel_heights = c(1,1))

#### AB
pAB <- plot_grid(plot_collect_peptides$Semitryptic+xlab("")+theme(legend.position = "none"),
                 plot_collect_peptides$PeaksWithAlternative_ncterms,ncol=1,labels = c("A","B"),scale = 0.9)

g1 <- plot_grid(pAB,
                plot_collect_peptides$Semitryptic_Stats,nrow=1,rel_widths   = c(1,2),labels=c("","C")
)

g2 <- plot_grid(SequenceInfoTead3,hu,rel_widths = c(2,1),labels = c("D",""))
```

## Glycosylation

```{r}
# Load Glycosyaltion data:
library(readxl)
Gly <- read_excel("AdditionalData/Sun_et_al_2019_GlycoSiteAtlas/12014_2019_9254_MOESM1_ESM.xlsx",skip = 1)
Gly <- data.table(Gly)
# Gly <- Gly[Database=="Reviewed"]

dt_MW <- Results_CombinedGels$Pairs_Filtered_DM#$Uniprot_MW
DM_cuts <- c(20000,50000,100000)
dt_MW[,c("< -100kD","-100kD to -50kD","-50kD to -20kD","-20kD to 20kD","20kD to 50kD","50kD to 100kd"):={
  list(
    "<-100kD"=DM_min < -DM_cuts[3],
    "-100kD to -50kD"= DM_min >= -DM_cuts[3]&DM_min < -DM_cuts[2],
    "-50kD to -20kD" = DM_min >= -DM_cuts[2]&DM_min < -DM_cuts[1],
    "-20kD to 20kD" = DM_min >= -DM_cuts[1]&DM_min <= DM_cuts[1],
    "20kD to 50kD" = DM_min >= DM_cuts[1]& DM_min <= DM_cuts[2] ,
    "50kD to 100kd"= DM_min >= DM_cuts[2]& DM_min <= DM_cuts[3] 

    
  )
}]

dt_MW$Glyco <- !is.na(match(dt_MW$hgnc_symbol,Gly$`Gene names`))
exon_pm <- exons[grep("endoplasmic|lysosome|vacuole|plasma membrane",goslim_goa_description)]
exon_pm <- exon_pm[NameSpace!="UNKNOWN"]

dt_MW[,Glyco_GO:={
  ue <- unlist(strsplit(.BY$ENSG,";"))
  any(!is.na(match(ue,exon_pm$ensembl_gene_id)))
},ENSG]
# meltfun <- melt(dt_MW,value.name = "DM_min", measure.vars = c("< -100kD","-100kD to -50kD","-50kD to -20kD","-20kD to 20kD","20kD to 50kD","50kD to 100kd"))
ggplot(dt_MW[type=="fw"&Use==T&score_ref_gbm>0.8&DM_min>0],aes(as.numeric(DM_min),color=Glyco|Glyco_GO))+stat_ecdf()

# 
# qs <- c(min(qfun,na.rm=T),-100000,-50000,-20000,20000,50000,100000,max(qfun,na.rm=T))
# 
# 
# # MW <- as.numeric(Results_CombinedGels$Pairs_Filtered_DM$IsoFrac_MW)
# liFun <- list(Less50kD=MW<50000,More50kD=MW>=50000,Higher100kD=MW>100000,From50To100kD=MW>50000&MW<100000,all=MW>-1000)
# 
# pf <- Results_CombinedGels$Pairs_Filtered
# 
# pf$Glyco <- !is.na(match(pf$hgnc_symbol,Gly$`Gene names`))
# pf[,DM_min:={
#   DM <- unlist(strsplit(.BY$DeltaMass ,";"))
#   DM <- as.numeric(DM)
#   DM[abs(DM)==min(abs(DM))][1]
# },DeltaMass]
# 
# 
# ggplot(pf[type=="fw"&Use==T],aes(as.numeric(DM_min),color=Glyco))+stat_ecdf()

```


## RNA Barplot Counts


```{r}
library(data.table)
library(Peptides)


fa1 <- fread("../../20220114_Isofrac_v2/IsoFrac_20230217_SimilaritiesTesting_4/Figure4_v2/Archive/protein_coding_transcript_info_uniq.txt",sep="\t",header = F)
names(fa1) <- c("Category","ENST","ENSG","Category2","seq")

fa2 <- fread("../../20220114_Isofrac_v2/IsoFrac_20230217_SimilaritiesTesting_4//Figure4_v2/Archive/protein_coding_transcript_ORFinfo_withAScategory.txt",sep="\t",header = T)

## Generate Fasta:
fastaExport <- fa2[,{
  gr <- .BY$transcript_info
  grtemp <- unlist(strsplit(gr,","))
  ENST <- grep("ENST",grtemp,value=T)
  ENSG <- grep("ENSG",grtemp,value=T)
  if(length(ENST)==0){
    ENST <- "noENST"
  }
  if(length(ENSG)==0){
    ENSG <- "noENSG"
  }
  ENST <- paste(ENST,collapse="#")
  ENSG <- paste(ENSG,collapse="#")

  Na <- paste(ENSG,ENST,.GRP,sep="|")
  
  
  

  list(fa=c(paste(">",Na,sep=""),.BY$ORF_seq))
},.(ORF_seq,transcript_info)]

write(fastaExport$fa,"Mengran_Fasta_20240530.fasta")

# length(unique(fa1$seq))
# length(unique(fa2$ORF_seq))


# fa2[,MW:=mw(.BY$ORF_seq),ORF_seq]
# fa1[,MW:=mw(.BY$seq),seq]

# MDC1
# fa1[ENSG=="ENSG00000137337"]



fa1[,.(length(unique(ENSG)),length(unique(seq))),]
fa2[,.(length(unique(ORF_seq))),]

fa1[,Category2_paste:={
paste(sort(unique(Category2)),collapse=";")
},seq]

Fun <- fa1[,.(dim(.SD)[1],length(unique(seq))),Category2_paste]
Fun$V1_rel <- Fun$V1/sum(Fun$V1)
Fun$V2_rel <- Fun$V2/sum(Fun$V2)
length(unique(fa1$seq))

ggplot(Fun, aes(x = "", y = V1, fill = Category2_paste)) +
  geom_col(color ="white") +
  geom_text(aes(label = round(V1_rel*100)),
            position = position_stack(vjust = 0.5)) +
  coord_polar(theta = "y") +
  # scale_fill_brewer() +
  theme_void()+ggtitle(paste(sum(Fun$V1),"coding transcripts"))

ggplot(Fun, aes(x = "", y = V2, fill = Category2_paste)) +
  geom_col(color ="white") +
  geom_text(aes(label = round(V2_rel*100)),
            position = position_stack(vjust = 0.5)) +
  coord_polar(theta = "y") +
  # scale_fill_brewer() +
  theme_void()+ggtitle(paste(sum(Fun$V2),"coding ORFs"))
# Fun$Category2_paste <- "UNKNOWN"
for(i in rev(c('PB_&_GENCODE', 'GENCODE_from_PBcorrected' , 'PB_only' , 'GENCODE_only'))){
  Fun$Category2_paste[grep(i,Fun$Category2_paste)] <- i
  fa1$Category2_paste[grep(i,fa1$Category2_paste)] <- i
  
}
# for(i in c("PB_&_GENCODE;PB_only","GENCODE_from_PBcorrected;PB_&_GENCODE","GENCODE_only;PB_&_GENCODE","GENCODE_only;PB_only","GENCODE_from_PBcorrected;GENCODE_only;PB_only","GENCODE_only;PB_&_GENCODE;PB_only","GENCODE_from_PBcorrected;GENCODE_only;PB_&_GENCODE","GENCODE_from_PBcorrected;GENCODE_only;PB_&_GENCODE;PB_only")){
#   Fun$Category2_paste[Fun$Category2_paste==i] <- "PB_&_GENCODE"
#   
# }
# for(i in c("GENCODE_from_PBcorrected;GENCODE_only")){
#   Fun$Category2_paste[Fun$Category2_paste==i] <- "PB_&_GENCODE"
#   
# }
table(Fun$Category2_paste)
FunFun <- Fun[,{
  list(Transcript=sum(V1),ORF=sum(V2))
},Category2_paste]

FunFunMelt <- melt(FunFun)
FunFunMelt[,value_rel:=value/sum(value),variable]
FunFunMelt[,TotalSum:=sum(value),variable]
FunFunMelt[,GroupLabel:=paste(variable,paste("(n=",TotalSum,")",sep="")),]
FunFunMelt$GroupLabel <- factor(FunFunMelt$GroupLabel,c(grep("ORF",unique(FunFunMelt$GroupLabel),value=T,invert = T),grep("ORF",unique(FunFunMelt$GroupLabel),value=T)))
FunFunMelt$Category2_paste <- gsub("_","\n",FunFunMelt$Category2_paste)
Results_CombinedGels_PacBio$BP_Counts <-ggplot(FunFunMelt, aes(Category2_paste, y = value, fill = Category2_paste,label=paste(round(value_rel*100),"%"))) +
  geom_bar(stat="identity",position="dodge")+facet_wrap("GroupLabel",ncol = 1)+xlab("")+ylab("Count")+
  theme(axis.text.x = element_text(angle=90,hjust=0.5,vjust=0.5),legend.position = "none")+
  # geom_text(hjust=1,angle=90)+
  guides(fill = guide_legend(nrow = 2))
Results_CombinedGels_PacBio$BP_Counts
# scale_fill_brewer() +
  # theme_void()+ggtitle(paste(sum(Fun$V2),"coding ORFs"))




```


# Subfigures

```{r}

# A,B,C,G,D,E,F
f4_p3 <- Results_CombinedGels_PBUP$RNAIdentifiedFraction$Pie

# D & E old figure 7
AUC <- Results_CombinedGels_PBUP$ROC+theme(legend.position = "right")+guides(fill=guide_legend(ncol=1,byrow=TRUE,label.hjust = -0.2))
upperrow <- plot_grid(AUC+xlab("Specificity")+ggtitle("ROC PACBIO Best Matching Entry"),plot_grid(Category_BP+theme(legend.position = "none",axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)),Category_BP_rel,rel_widths = c(0.4,1)),
                      
                      
                      rel_widths = c(1.6,2),labels = c("D","E"),ncol=2)

# E & F

lowerrow <- plot_grid(DeltaMassFig+ggtitle("Isoform Pairs UniProt vs PACBIO\n Uniprot or PACBIO Score > 0.8")+
                        update_geom_defaults(geom="point",new =list(size=0.5) ),DeltaMassFigCounts+ggtitle("Counts")+xlab("")+
                        theme(legend.position = "none",axis.text.x=element_text(angle=45,vjust=1,hjust=1)),
                      ProteinCategory_BP+ theme(legend.position = "bottom",legend.margin = margin(l = -2, unit = "cm"))+guides(fill=guide_legend(nrow=2,byrow=TRUE,label.hjust = -0.2,override.aes = aes(label = "")))
                      ,rel_widths = c(1.3,0.5,1),nrow=1,rel_heights = c(1,0.7),labels = c("F","","G"))


f4_p2 <- plot_grid(f4_p3,upperrow,lowerrow,nrow=3,rel_heights = c(0.7,0.9,1),labels=c("C"))

# A & B 

# placeholder for Mengrans figures:

f <- fread("comprehensiveORF_PacBio_GENCODE_v2.txt",sep = "\t")

ORFCounts <- f[,.(dim(.SD)[1],length(unique(ORF_seq))),GENCODE_gene_ID]
ORFcounts <- ORFCounts[,.("unique ORFs"=length(GENCODE_gene_ID),"unique Transcripts"=sum(V1)),V2]
ORFcounts2 <- ORFcounts[V2>=8,.(V2="8+","unique ORFs"=sum(as.data.frame(.SD)[,2]),"unique Transcripts"=sum(as.data.frame(.SD)[,3]))]
ORFcounts3 <- rbind(ORFcounts[V2<8],ORFcounts2)

ORFcountsAll <- melt(ORFcounts3,id.vars = "V2")

RNAcounts <- ggplot(ORFcountsAll,aes(V2,value,fill=variable,label=value))+geom_bar(stat="identity",position="dodge")+xlab("Variants per Gene")+ylab("Count")+theme(legend.title = element_blank())+geom_text(angle=90,position = position_dodge(width = 0.9), vjust = 0.5,hjust=-0.2)+theme(legend.title = element_blank())

A4 <- fread("Figure4B_data.csv",sep=",")
A4Melt <- melt(A4)
# ggplot(A4Melt,aes(variable,value,label=value,fill=Count_trans_ORFs_per_gene ))+geom_bar(stat="identity",position="dodge")+geom_text(angle=90,position = position_dodge(width = .9),hjust=1)+xlab("# Isoforms per Gene")+theme(legend.position = "bottom",legend.title = element_blank())+ylab("Genes")

RNAcounts <- ggplot(A4Melt,aes(variable,value,fill=Count_trans_ORFs_per_gene,label=value))+geom_bar(stat="identity",position="dodge")+xlab("Variants per Gene")+ylab("Count")+theme(legend.title = element_blank())+geom_text(angle=90,position = position_dodge(width = 0.9), vjust = 0.5,hjust=-0.2)+theme(legend.title = element_blank())+ylim(c(0,4300))

library(pdftools)
# lgg_mengran_placeholder <- lapply(c("./Figure4_v2/4A.pdf","./Figure4_v2/4B.pdf"),function(x){
#   x <<- x
#   xpath <- pdf_convert(x,dpi = 50)
#   Mengran1 <- draw_image(xpath)
#   Mengran1 <- ggdraw()+Mengran1
# })
xpath <- pdf_convert("./Figure4_v2/Mengran_Sketch_mod.pdf",dpi = 300)


Mengran1 <- ggdraw()+draw_image(xpath)
```

# Extracted ID Tables
```{r}
PB <- Results_CombinedGels_PBUP$BestMatch_PB[type=="fw"&Use==T]
PB$DB <- "PacBio"
PB$BestMatch <- PB$ScFiltered_Uniprot
UP <- Results_CombinedGels_PBUP$BestMatch_UP[type=="fw"&Use==T]
UP$DB <- "Uniprot_2022"
names(PB)
SelectedColumns <- c("ENSG","hgnc_symbol","IsoFrac","IsoFrac_MW","Uniprot","Uniprot_MW","Category","DeltaMass","DM_min","nClusters","nGels","Height","gel_MainPeak","Fraction_sd_80","All_Uniprot_Peptides","Success_ISOFRAC_Peptides_UP","all_sequences","score_ref","ScoreType","label","DB")
UP_sel <- UP[,.SD,.SDcols = SelectedColumns]
PB_sel <- PB[,.SD,.SDcols = SelectedColumns]

ProteinIsoformsAll <- rbind(UP_sel,PB_sel)
setnames(ProteinIsoformsAll,
         c("Uniprot","Uniprot_MW","All_Uniprot_Peptides"),
         c("Database_BestMatch","Database_MW","All_Database_Peptides")
         )

fwrite(ProteinIsoformsAll,"SupplementalTable_1_proteinisoforms.txt",sep="\t")
```

## Figure 4 v2 20250313

```{r}
if(!exists("TMPO_Traces")){
 input <- list(GN="ENSG00000120802",inputtable="Exp1_10per_Try")
 dbPath <- "IsoFrac_0.85/TableListNormEnsHdbScanMW_Peaks.sqlite"
dataset <- Extract_IsoFrac_Info(dbPath,input)

dataset$hdbscanTables <- dataset$hdbscanTables[BestCluster!=0]

LM <- as.numeric(dataset$Pairs_Filtered_Compressed$Height)
LM[LM<0.35] <- 3.5
dataset$Pairs_Filtered_Compressed$Height <- LM

TMPO_Traces <- Traces_Plotly(dataset,ShowQuantiles = T,linelwd = 0.8) 
TMPO_Traces$g <- TMPO_Traces$g+ylim(c(-1,5))+theme(legend.title = element_blank())

}



# A,B,C,G,D,E,F
f4_p3 <- plot_grid(AUC,Results_CombinedGels_PBUP$RNAIdentifiedFraction$Pie,rel_widths = c(0.75,1),labels = c("C","D"))

# D & E old figure 7
AUC <- Results_CombinedGels_PBUP$ROC+theme(legend.position = "right")+guides(fill=guide_legend(ncol=1,byrow=TRUE,label.hjust = -0.2))+xlab("Specificity")+ggtitle("ROC PACBIO Best Matching Entry")


xpath <- pdf_convert("./WB_Confirmation/TMPO_PACBIO_example.pdf",dpi = 300)
TMPO_Pacbio <- ggdraw()+draw_image(xpath)


# E & F

lowerrow_RNA_PROTEIN_Integration <- plot_grid(DeltaMassFig+ggtitle("Isoform Pairs UniProt vs PACBIO\n Uniprot or PACBIO Score > 0.8")+
                        update_geom_defaults(geom="point",new =list(size=0.5) ),DeltaMassFigCounts+ggtitle("Counts")+xlab("")+
                        theme(legend.position = "none",axis.text.x=element_text(angle=45,vjust=1,hjust=1)),
                      ProteinCategory_BP+ theme(legend.position = "bottom",legend.margin = margin(l = -2, unit = "cm"))+guides(fill=guide_legend(nrow=2,byrow=TRUE,label.hjust = -0.2,override.aes = aes(label = "")))
                      ,rel_widths = c(1.3,0.5,1),nrow=1,rel_heights = c(1,0.7),labels = c("F","","G"))

#upperrow
#lowerrow

f4_p2 <- plot_grid(f4_p3,TMPO_Pacbio+ggtitle("PACBIO Results for TMPO"),TMPO_Traces$g+ggtitle("ISOFRAC Traces for TMPO"),nrow=3,rel_heights = c(0.7,0.9,1))

# A & B 

# placeholder for Mengrans figures:

f <- fread("comprehensiveORF_PacBio_GENCODE_v2.txt",sep = "\t")

ORFCounts <- f[,.(dim(.SD)[1],length(unique(ORF_seq))),GENCODE_gene_ID]
ORFcounts <- ORFCounts[,.("unique ORFs"=length(GENCODE_gene_ID),"unique Transcripts"=sum(V1)),V2]
ORFcounts2 <- ORFcounts[V2>=8,.(V2="8+","unique ORFs"=sum(as.data.frame(.SD)[,2]),"unique Transcripts"=sum(as.data.frame(.SD)[,3]))]
ORFcounts3 <- rbind(ORFcounts[V2<8],ORFcounts2)

ORFcountsAll <- melt(ORFcounts3,id.vars = "V2")

RNAcounts <- ggplot(ORFcountsAll,aes(V2,value,fill=variable,label=value))+geom_bar(stat="identity",position="dodge")+xlab("Variants per Gene")+ylab("Count")+theme(legend.title = element_blank())+geom_text(angle=90,position = position_dodge(width = 0.9), vjust = 0.5,hjust=-0.2)+theme(legend.title = element_blank())

A4 <- fread("Figure4B_data.csv",sep=",")
A4Melt <- melt(A4)
# ggplot(A4Melt,aes(variable,value,label=value,fill=Count_trans_ORFs_per_gene ))+geom_bar(stat="identity",position="dodge")+geom_text(angle=90,position = position_dodge(width = .9),hjust=1)+xlab("# Isoforms per Gene")+theme(legend.position = "bottom",legend.title = element_blank())+ylab("Genes")

RNAcounts <- ggplot(A4Melt,aes(variable,value,fill=Count_trans_ORFs_per_gene,label=value))+geom_bar(stat="identity",position="dodge")+xlab("Variants per Gene")+ylab("Count")+theme(legend.title = element_blank())+geom_text(angle=90,position = position_dodge(width = 0.9), vjust = 0.5,hjust=-0.2)+theme(legend.title = element_blank())+ylim(c(0,4300))

library(pdftools)
# lgg_mengran_placeholder <- lapply(c("./Figure4_v2/4A.pdf","./Figure4_v2/4B.pdf"),function(x){
#   x <<- x
#   xpath <- pdf_convert(x,dpi = 50)
#   Mengran1 <- draw_image(xpath)
#   Mengran1 <- ggdraw()+Mengran1
# })
xpath <- pdf_convert("./Figure4_v2/Mengran_Sketch_mod.pdf",dpi = 300)
Mengran1 <- ggdraw()+draw_image(xpath)
# G






```


# Validation

```{r}
#### Reanalysing MQ with recent PACBIO entries, together with MQ PSM and Protein FDR set to 1

eviall <- fread("IsoFrac_3Gels_Reanalysis_PACBIO/MQ_2.0.3.1/evidence.txt",sep = "\t")
evi <- eviall[Reverse==""&`Potential contaminant`==""]
head(evi$Proteins)

evi[,MQ:=grepl("sp\\||tr\\||sp_|tr_",.BY$Proteins),Proteins]
evi[,PACBIO:=grepl("ENSG",.BY$Proteins),Proteins]

evi$MappedDB <- "Undefined"
evi[MQ==TRUE&PACBIO==FALSE,MappedDB:="MQ"]
evi[MQ==FALSE&PACBIO==TRUE,MappedDB:="PB"]
evi[MQ==TRUE&PACBIO==TRUE,MappedDB:="MQ_PB"]
table(evi$MappedDB)

PACBIO_SearchResults <- list()
# BP of PEP Results
PACBIO_SearchResults$PEP_BP <- ggplot(evi,aes(MappedDB,-log10(as.numeric(PEP)),fill=MappedDB))+geom_boxplot()#+facet_grid("")##+xlim(c(0,10))

PBonly <- fa1[Category2_paste=="PB_only"]

li <- list(MQ=unique(evi[MappedDB=="MQ"]$Proteins),PB=unique(evi[MappedDB=="PB"]$Proteins),PBMQ=unique(evi[MappedDB=="MQPB"]$Proteins))
gplots::venn(li)

PBspecific <- evi[MappedDB=="PB"]
fa1$ID <- 1:dim(fa1)[1]
# Mapping RNA Categories
MappedCategories <- PBspecific[,{
  gr <- .BY
  gr <- strsplit(Proteins,";")
  humpe <- lapply(gr,function(x){
    x <<- x
    x <- strsplit(x,"|",fixed=T)
    x <- unlist(x)
    resi <- fa1[ENSG==x[1]&ENST==x[2]]
    if(dim(resi)[1]==0){
      resi <- fa1[!is.na(match(ENST,unlist(strsplit(x[2],"#"))))]
      
    }
    if(dim(resi)[1]==0){
      resi <- fa1[!is.na(match(ENSG,unlist(strsplit(x[1],"#"))))]

    }
   

    data.table(Category=paste(unique(resi$Category2),collapse=";"),IDs=paste(unique(resi$ID),collapse=";"))
    
  })

  hu <- rbindlist(humpe)
  if(all(hu$Category=="")){
    stop()
  }
  hu <- apply(hu,2,function(x){paste(unique(x),collapse = "#")})

  # hu$Sequences <- paste(.SD$Sequence,collapse=";")
  hu$Scores <-   paste(.SD$Score,collapse=";")

  as.list(hu)
},.(Proteins,`Leading proteins`,ENSG)]
MappedCategoriesCollapsed <- MappedCategories[,.(Category =paste(sort(unique(unlist(strsplit(Category,";")))),collapse=";"),n=dim(.SD)[1],Scores=paste(Scores,collapse=";")),.(`Leading proteins`,ENSG)]

# Extracting SDS Gele Results from MQ
PBevi <- funni <- lapply(TableListNormEnsHdbScanMW_Peaks,function(x){
  x$InputTable
})
PBevi <- rbindlist(PBevi)

### Plotting
MappedCategoriesCollapsed$Category2_paste <- ""
for(i in rev(c('PB_&_GENCODE', 'GENCODE_from_PBcorrected'  , 'GENCODE_only', 'PB_only'))){
  print(i)
  MappedCategoriesCollapsed$Category2_paste[grep(i,MappedCategoriesCollapsed$Category)] <- i

}


MappedCategoriesCount <- MappedCategoriesCollapsed[,length(`Leading proteins`),.(Category2_paste,SequenceSpecific=n==1,Scores)]
MappedCategoriesCount$ProteinGroupSpecificity <- ""
MappedCategoriesCount[SequenceSpecific==T]$ProteinGroupSpecificity <- "1 Protein"
MappedCategoriesCount[SequenceSpecific==F]$ProteinGroupSpecificity <- "> 1 Proteins"

fa1$n <- 1
FaCount <- fa1[,length(unique(seq)),.(Category2_paste,n==1)]
FaCount$type <- "RNA entries"
# FaCount$ProteinGroupSpecificity
MappedCategoriesCount$type <- "Identified Protein"
# combing both tables
CombinedProteinCounts <- rbind(MappedCategoriesCount,FaCount,fill=T)
CombinedProteinCounts$PACBIOtype="Other"
CombinedProteinCounts[Category2_paste=="PB_only"]$PACBIOtype <- "PB_only"

plotCollecter$Fig5$MQPB_Counts <- ggplot(CombinedProteinCounts[,sum(V1),.(PACBIOtype,ProteinGroupSpecificity,type)],aes(type,V1,fill=ProteinGroupSpecificity,label=V1))+geom_bar(stat="identity",position="stack")+geom_label_repel(color="white",direction = "x",position = position_stack())+ylab("Unique Protein Count")+xlab("")+facet_wrap("PACBIOtype")


# PEP Comparison

MappedCategoriesCountScores<- MappedCategoriesCount[,.(Score=as.numeric(unlist(strsplit(Scores,";")))),.(Category2_paste ,SequenceSpecific,ProteinGroupSpecificity,V1,type)]
plotCollecter$Fig5$MQPB_Scores <- ggplot(MappedCategoriesCountScores,aes((Score),fill=Category2_paste))+geom_density(alpha=0.5)+facet_grid("ProteinGroupSpecificity")

PBonly <- MappedCategoriesCollapsed[Category=="PB_only"&!grepl(";",`Leading proteins`)&n==1]


## PB Specific
PBonly_evi <- eviall[!is.na(match(`Leading proteins`,PBonly$`Leading proteins`))]
```

## WB

```{r}
# Prerun on silver with WB_Analysis.Rmd
library(grid)
library(msa)
library(ggmap)
### Western Blots
gg_layerremove <- function(gg,removeTag="PeakPosition|alpha"){
  
  gg$layers <- lapply(gg$layers,function(x){
    x <- x
    if(any(grepl(removeTag,names(x$data) ))){
      x <- NULL
    }
    
    x
    # names(x)
  })
  gg$layers <- gg$layers[lengths(gg$layers)>0]
  
  
  gg
}
if(!exists("CollectGG_v2")){
  #hu <- load("./WB_Confirmation/WB_Collection.rda")
  hu <- load("./WB_Confirmation/CollectionWB_Supplemental.rda")

  Collection_2_WB_N_backup <- Collection_2_WB_N
  Collection_2_WB_N$gg <- gg_layerremove(Collection_2_WB_N$gg)
  Collection_2_WB_C$gg <- gg_layerremove(Collection_2_WB_C$gg)
  CollectionWB_C$gg <- gg_layerremove(Collection_2_WB_C$gg)
  CollectionWB_C$gg
  Collection_2_WB_C$gg
}
#Subfigure_WB 

WB_BirA_cterm <- ggdraw()+draw_image("WB_Confirmation/BirA_Tag_Design_Cterm_v2.png")
WB_BirA_nterm <- ggdraw()+draw_image("WB_Confirmation/BirA_Tag_Design_nterm_v2.png")
Tead3_WB <- ggdraw()+draw_image("WB_Confirmation/TEAD3_RPE1.jpg")
  # hu <- load("WB_Confirmation/WB_Collection.rda")

Style <- theme(plot.title = element_text(hjust = 0.75),legend.position = "none",axis.text.x = element_blank())

# Tead3
Tead3_nterminal <- Collection_2_WB_N$gg+coord_cartesian(xlim=c(0.09,0.22),ylim=c(0.1,0.9))+ggtitle("Tead3\nnterm")+Style+scale_alpha(range=c(1,1))
Tead3_cterminal <- CollectionWB_C$gg+coord_cartesian(xlim=c(0.015,0.18),ylim=c(0.1,0.9))+ggtitle("Tead3\ncterm")+Style
# Tead3 <- plot_grid(
# ,labels = "")

# NELFA
Collection_2_WB_N$GelPeaks$TEAD3$Ladder
NELFA_WB <- plot_grid(Collection_2_WB_N$gg+coord_cartesian(xlim=c(0.35,0.51),ylim=c(0.1,0.9))+ggtitle("NELFA\nnterm")+Style+scale_alpha(range=c(1,1)),
Collection_2_WB_C$gg+coord_cartesian(xlim=c(0.35,0.51),ylim=c(0.1,0.9))+ggtitle("NELFA\ncterm")+Style+scale_alpha(range=c(1,1)),labels = "")

# CTTNBP2NL
CTTNBP2NL_WB <- plot_grid(Collection_2_WB_N$gg+coord_cartesian(xlim=c(0.5,0.66),ylim=c(0.1,0.9))+ggtitle("CTTNBP2NL\nnterm")+Style+scale_alpha(range=c(1,1)),
Collection_2_WB_C$gg+coord_cartesian(xlim=c(0.5,0.66),ylim=c(0.1,0.9))+ggtitle("CTTNBP2NL\ncterm")+Style,labels = "")+scale_alpha(range=c(1,1))

# RRP1B
RRP1B_WB <- plot_grid(Collection_2_WB_N$gg+coord_cartesian(xlim=c(0.66,0.8),ylim=c(0.1,0.9))+ggtitle("RRP1B\nnterm")+Style+scale_alpha(range=c(1,1)),
Collection_2_WB_C$gg+coord_cartesian(xlim=c(0.66,0.8),ylim=c(0.1,0.9))+ggtitle("RRP1B\ncterm")+Style+scale_alpha(range=c(1,1)),labels = "")


SequenceMappings_selectedGenes <- lapply(GeneList,function(x){
  try(IF_DB_Peptidemapping(x,IsoFrac_Domain_SequenceAnalysis,ProteinSequencesCategories,invertOrder = T,allgels=T,facetPlot = F))

})

SequenceMappings_selectedGenes_v2  <- lapply(GeneList,function(x){
  try(IF_DB_Peptidemapping(x,IsoFrac_Domain_SequenceAnalysis,ProteinSequencesCategories,invertOrder = T,allgels=F,facetPlot = F))

})
SequenceMappings_selectedGenes_v2$TEAD3

```


## Subfigures3

```{r}
# load("plotCollecter.rda")
#
PacBio_Subset <- plot_grid(plotCollecter$Fig5$MQPB_Counts+theme(axis.text.x = element_text(angle=45)),plotCollecter$Fig5$MQPB_Scores,labels = c("A","B"))
# SemiTrypticSubset <- plot_grid(plotCollecter$Fig4$Semitryptic_Counts,plotCollecter$Fig4$Semitryptic_Counts$density+geom_vline(xintercept = 0,lty="dotted"),labels = c("C","D"),rel_widths = c(0.75,1))
SemiTrypticSubset <- plot_grid(plot(plotCollecter$Fig5$Semitryptic_Counts_euler),plotCollecter$Fig5$MetaGenePositions+theme(legend.title = element_blank()),plotCollecter$Fig5$Semitryptic_Counts$density+geom_vline(xintercept = 0,lty="dotted"),labels = c("C","D"),rel_widths = c(0.5,1,1),nrow=1)
# p <- plot(plotCollecter$Fig4$Semitryptic_Counts_euler)
# plotCollecter$Fig5$Semitryptic_Counts_euler
# Sceme <- plot_grid(WB_BirA_nterm,WB_BirA_cterm,scale = 0.8,labels="E")

WB_Subset_subset <- plot_grid(plot_grid(WB_BirA_nterm,Tead3_nterminal,ncol=1,rel_heights = c(0.5,1),scale=c(0.8,1)),
                              plot_grid(WB_BirA_cterm,Tead3_cterminal,ncol=1,rel_heights = c(0.5,1),scale=c(0.8,1)),nrow=1)


Tead3_WB

WB_Subset <- plot_grid(WB_Subset_subset,SequenceMappings_selectedGenes$TEAD3$MSAPlot2,labels = c("F","G"),rel_widths = c(1,1))

```

## MaxQuant Specific Analysis

```{r}
load("../IsoFracCompiler2_noSemitryptic/Experiments/TableListNormEnsHdbScan_v5_wo_semitryptic.rda")
# Results from original session:
UniprotPeptides <- unique(unlist(lapply(TableListNormEnsHdbScan,function(x){
  x$InputTable$Sequence
})))
#
upfa <- seqinr::read.fasta("../ProteinDatabases/HS_UP000005640_gff_compiled_hz20220603.fa",seqtype = "AA",as.string = T)
upfa_dt <- data.table(upfa)
upfa_dt$names <- names(upfa)

upseqs <- unique(unlist(upfa))
upseqs <- unique(c(ProteinSequencesCategories$SEQUENCE,upseqs))

upseqs_ItoL <- gsub("I","L",upseqs,fixed=T)

# grep("TSSPIEVAR",ProteinSequencesCategories$SEQUENCE ,fixed=T)

# Reading New MQ Search
library(data.table)
library(ggplot2)

pepi <- fread("E:/MaxQuant/HZ/20220114_Isofrac_v2/PACBIO_REANALYSIS/IsoFrac_3Gels_Reanalysis_PACBIO/MQ_2.0.3.1/peptides.txt",sep = "\t",nrows = 100)

eviall <- fread("E:/MaxQuant/HZ/20220114_Isofrac_v2/PACBIO_REANALYSIS/IsoFrac_3Gels_Reanalysis_PACBIO/MQ_2.0.3.1/evidence.txt",sep = "\t")
evi <- eviall[PEP<0.01&Reverse==""&`Potential contaminant`==""]
evi$UniProt <- !is.na(match(evi$Sequence,UniprotPeptides)) # Identified with UP in MQ
# evi[,UniProt:=any(.BY$Sequence==UniprotPeptides),Sequence]

# head(evi$Proteins)
# identifying MQ and PACBIO Specific peptides
evi[,MQ:=grepl("sp\\||tr\\||sp_|tr_",.BY$Proteins),Proteins]
evi[,PACBIO:=grepl("ENSG",.BY$Proteins),Proteins]
evi$MappedDB <- "Undefined"
evi[MQ==TRUE&PACBIO==FALSE,MappedDB:="MQ"]
evi[MQ==FALSE&PACBIO==TRUE,MappedDB:="PB"]
evi[MQ==TRUE&PACBIO==TRUE,MappedDB:="MQ_PB"]

PBspecific <- evi[MappedDB=="PB"]

library(parallel)
  #evi[grep("ENSG00000197948",Proteins)]
cl <- makeCluster(100)
clusterExport(cl,"upseqs")

UniprotMappingsList <- parLapply(cl,unique(PBspecific$Sequence),function(x){
  library(stringr)
  library(data.table)
  g1 <- grep(x,upseqs,fixed=T)
  detailsfun <- lapply(g1,function(temps){
    # temps <-  g1
    after <- str_match(upseqs[temps], paste0(x, "(.)"))[2]
    before <- str_match(upseqs[temps], paste0("(.)",x))[2]
    
    data.table(after=after,before=before)
    
  })
  detailsfun <- rbindlist(detailsfun)
  detailsfun$sequence <- x
  detailsfun$positions <- g1
  detailsfun

  
})
UniprotMappings <- rbindlist(UniprotMappingsList,fill=T)
UniprotMappings$Type <-""
UniprotMappings[!grepl("K$|R$",sequence)&!is.na(after)]$Type <- "new c-terminal candidate"# no K or R ending, but within sequence
UniprotMappings[before!="K"&before!="R"&before!="M"]$Type <- "new n-terminal candidate"# no K or R before
UniprotMappings[Type=="new n-terminal candidate"]
UniprotMappings[,Typical:=any(Type==""),sequence]
grep("TSSPIEVAR",upseqs,value=T)
UniprotMappings[sequence=="TSSPIEVAR"]

NotInUniprotFasta <- unique(PBspecific$Sequence)[which(lengths(UniprotMappingsList)<4)]
NotInUniprotFasta2 <-unique(UniprotMappings[Typical==FALSE]$sequence )

# NotInUniprotFasta
# 
# cl <- makeCluster(1)
clusterExport(cl,"upseqs_ItoL")

UniprotMappings2 <- parLapply(cl,unique(NotInUniprotFasta),function(x){
  grep(gsub("I","L",x),upseqs_ItoL,fixed=T)
})
NotInUniprotFasta <- NotInUniprotFasta[lengths(UniprotMappings2)==0]
NotInUniprotFasta <- setdiff(c(NotInUniprotFasta,NotInUniprotFasta2),UniprotPeptides)

# table(lengths(UniprotMappings))


PBspecific$UniProt_Fasta <- is.na(match(PBspecific$Sequence,NotInUniprotFasta))
table(PBspecific$UniProt_Fasta)

gplots::venn(list(NotInUniprotFa=NotInUniprotFasta,OnlyPacBIO=PBspecific[UniProt==F]$Seq,unique(UniprotPeptides)))
# upseqs
# table(PBspecific$UniProt)

MappedCategories <- PBspecific[,{
  cat("\r",.GRP)
  # gr <- .BY
  gr <<- strsplit(Proteins,";|#")
  # stop()
  humpe <- lapply(gr,function(x){
    x <- x
    x <- strsplit(x,"|",fixed=T)
    x <- unlist(x)
    resi <- fa1[ENSG==x[1]&ENST==x[2]]
    if(dim(resi)[1]==0){
      resi <- fa1[ENST==x[2]]
    }
    if(dim(resi)[1]==0){
      resi <- fa1[ENSG==x[1]]
    }
    data.table(Category=paste(unique(resi$Category2),collapse=";"),IDs=paste(unique(resi$ID),collapse=";"))
  })
  hu <- rbindlist(humpe)
  hu <- apply(hu,2,function(x){paste(unique(x),collapse = "#")})
  
  hu <- as.list(hu)
  hu$NotIn_UniProtMQ <- any(!UniProt)
  hu$NotIn_UniProtFa <- any(!UniProt_Fasta)
  
  hu
},.(`Leading proteins`,Proteins)]
table(MappedCategories$NotIn_UniProtMQ)
table(MappedCategories$NotIn_UniProtFa)

MappedCategoriesProtein <- MappedCategories[,{paste(unique(Category),collapse=";")},.(Proteins,`Leading proteins`,NotIn_UniProtMQ,NotIn_UniProtFa)]
MappedCategoriesProtein$CategoryClean <- "Undefined"
# MappedCategoriesProtein
for(i in rev(c('PB_&_GENCODE', 'GENCODE_from_PBcorrected' , 'PB_only' , 'GENCODE_only'))){
  MappedCategoriesProtein$CategoryClean[grep(i,MappedCategoriesProtein$V1)] <- i
}
MappedCategoriesProtein[CategoryClean=="Undefined"]
# MappedCategoriesProtein[Uni]

ENSG00000163539
 
### Protein Counts

MappedCategoriesProtein$V1 <- NULL
# MappedCategoriesProtein[,,Unipr]
temp <- MappedCategoriesProtein[NotIn_UniProtFa==T]
temp
MappedCategoriesProteinTable <- MappedCategoriesProtein[,{
  temp <<- .SD
  temp[,{
    temp2 <<- .SD
    temp2
  },.(CategoryClean)]
  MC <- table(CategoryClean)
  MC <- as.data.table(MC)
  MC <- MC[order(N,decreasing = T)]
},.(NotIn_UniProtFa)]


MappedCategoriesProteinTable$Group <- "Known"
MappedCategoriesProteinTable$Group[MappedCategoriesProteinTable$CategoryClean=="PB_only"] <- "New"

# MC
MC <- MappedCategoriesProteinTable
MC$V1 <- factor(MC$CategoryClean     ,unique(MC$CategoryClean )   )
MC$Group <- "Known"
MC$Group[MC$V1=="PB_only"] <- "New"
MC$NotIn_UniProtFa_name <- "Uniprot & PacBio"
MC$NotIn_UniProtFa_name[MC$NotIn_UniProtFa] <- "PacBio"
PlotCollectionFigure6 <- list()
PlotCollectionFigure6$MQ$Counts <- ggplot(MC,aes(Group,N,label=N,fill=V1,group=N))+geom_bar(stat="identity")+theme(axis.text.x = element_text(angle=0),legend.title = element_blank())+ggtitle("PacBio Database search, no protein FDR")+geom_text(vjust=-0.5,cex=3,position = position_stack(vjust = 0.5))+ylab("Proteins")+xlab("")+facet_grid("NotIn_UniProtFa_name",scale="free")
PlotCollectionFigure6$MQ$Counts_Short <- ggplot(MC,aes(Group,N,label=N,fill=V1,group=N))+geom_bar(stat="identity")+theme(axis.text.x = element_text(angle=0),legend.title = element_blank())+ggtitle("PacBio Database search, no protein FDR")+geom_text(vjust=-0.5,cex=3,position = position_stack(vjust = 0.5))+ylab("Proteins")+xlab("")+facet_grid("NotIn_UniProtFa_name",scale="free")
PlotCollectionFigure6$MQ$Counts

unique(MappedCategoriesProtein[CategoryClean=="PB_only"&NotIn_UniProtFa ==T]$Proteins)
PlotCollectionFigure6$MQ$Counts_v2 <- ggplot(MappedCategoriesProteinTable[NotIn_UniProtFa=="TRUE"],aes(factor(Group,c("Known","New"),c("shared peptides\nUP & PB","PB only")),N,label=N,fill=CategoryClean,group=N))+geom_bar(stat="identity")+theme(axis.text.x = element_text(angle=0),legend.title = element_blank())+ggtitle("PacBio specific proteins")+
  geom_text(vjust=-0.5,cex=3,position = position_stack(vjust = 0.5))+ylab("Protein Groups")+xlab("")#+facet_grid("NotIn_UniProtFa",scales = "free")
PlotCollectionFigure6$MQ$Counts_v2

PB_Only_Candidates <- MappedCategoriesProtein[CategoryClean=="PB_only"&(NotIn_UniProtFa==T)]$Proteins

### Features
evi$Group <- "Known"
evi[!is.na(match(Proteins,PB_Only_Candidates))]$Group <- "All Peptides\nfrom\nNew Proteins"
evi[!is.na(match(Proteins,PB_Only_Candidates))&!is.na(match(Sequence,NotInUniprotFasta))]$Group <- "Specific Peptides\nfrom\nNew Proteins"

Summary <- evi[,.(Median_Score=median(Score,na.rm=T),
                  Score_Max=max(Score,na.rm=T),
                  Score_Min=min(Score,na.rm=T),
                  Peptides=length(unique(Sequence)),
                  Median_Peptide_Length=as.double(median(nchar(Sequence),na.rm=T)))
               ,.(Group,`Leading proteins`,Proteins)]
evi[,PeptideLength:=nchar(.BY$Sequence),Sequence]

evitemp <- evi[,1,.(Group,`Leading proteins`,Proteins,Sequence,PeptideLength,Score)]
SummaryMelt <- melt(Summary)
evitempMelt <- melt(evitemp)
PlotCollectionFigure6$MQ$Features <- ggplot(SummaryMelt,aes(Group,value,color=Group))+ geom_jitter(data = subset(SummaryMelt, Group != "Known"),width = 0.2,alpha=1,pch=20,size=3)+geom_violin(draw_quantiles = c(c(0.25,0.5,0.75)),fill=NA,color="black")+facet_wrap("variable",scales = "free")+theme(legend.position = "none")+xlab("")
PlotCollectionFigure6$MQ$Features

PlotCollectionFigure6$MQ$Features_evi <- ggplot(evitempMelt[variable!="V1"],aes(Group,value,fill=Group))+geom_boxplot(draw_quantiles = c(c(0.25,0.5,0.75)),bw=3)+facet_wrap("variable",scales = "free")+theme(legend.position = "none")+xlab("")
PlotCollectionFigure6$MQ$Features_evi 

x <- "5per"
tmt_10 <- evi[grep(x,`Raw file`)]
dim(tmt_10)
data.table::setDTthreads(50)

#Adding ensg Information

PBspecific[,ENSG:={
  hu <- strsplit(.BY$Proteins,"|",fixed=T)[1][[1]]
  hu <- unlist(strsplit(unlist(hu),";|#"))
  paste(unique(grep("ENSG",hu,value=T)),collapse=";")
},Proteins]
grep(";",PBspecific$ENSG,value=T)

# Reanalyzing Traces for PB_Only Transitions

PBonly_MQ <- lapply(c("5per","8per","10per"),function(x){
  PBspecificTemp <- PBspecific[grep(x,`Raw file`)]
  PBspecificTemp <- PBspecificTemp[,PB_Only:=any(!is.na(match(.BY$Proteins,PB_Only_Candidates))),.(Proteins,Sequence)]
  PBspecificTemp <- PBspecificTemp[,Pick:=any(!is.na(match(Proteins,PB_Only_Candidates))),ENSG]
  PBspecificTemp <- PBspecificTemp[Pick==TRUE]
  # table(PBspecificTemp$PB_Only)
  
  RequiredNames <- names(pepi)[!is.na(match(names(pepi),names(PBspecificTemp)))]
  Reporterstuff <- grep("Reporter intensity corrected",RequiredNames,value=T)
  InfoStuff <- grep("Reporter",RequiredNames,value=T,invert=T)
  ## Add Odd or Even
  PBspecificTemp[,Even:=grepl("Even",.BY$`Raw file`),`Raw file`]
  
  # tmt_10 <- PBspecificTemp[,{
  #   cat("\r",.GRP)
  #   temp <- .SD
  #   qu <- apply(temp[,.SD,.SDcols=Reporterstuff],2,sum)
  #   info <- apply(temp[,.SD,.SDcols=setdiff(InfoStuff,"Sequence")],2,function(x){paste(unique(x),collapse=";")})
  #   as.list(c(info,qu))
  # },.(Sequence, ENSG,Even)]
  
  # hack to get the experimental design right of the fractions and from the tmt channels
  tmt_10 <- PBspecificTemp[,{
    cat("\r",.GRP)
    temp <- .SD
    # if(length(unique(temp$Even))>1){
    #   stop()
    # }
    tempi <- temp[,{
      tempi <- apply(.SD[,.SD,.SDcols=Reporterstuff],2,sum)
      names(tempi) <-Reporterstuff 
      # tempi <- dcast(tempi,Even~V1 )
      as.list(tempi)
    },.(Even)]
    # if()
    tempiEven <- tempi[Even==TRUE]
    names(tempiEven) <- paste("Plex1perEven",names(tempiEven))
    tempiOdd <- tempi[Even==FALSE]
    names(tempiOdd) <- paste("Plex2perOdd",names(tempiOdd))
    qu <- cbind(tempiEven,tempiOdd)
    info <- apply(temp[,.SD,.SDcols=setdiff(InfoStuff,"Sequence")],2,function(x){paste(unique(x),collapse=";")})
    as.list(c(info,qu))
  },.(Sequence, ENSG,PB_Only,UniProt_Fasta)]
  


  # length(unique(PBspecificTemp$Sequence))
  #' Slicing for Reporter intensity
  TMTSubset <- tmt_10[,.SD,.SDcols=grep("Reporter intensity corrected [1234567890]",names(tmt_10),value = T)]
  
  Infotable <- tmt_10[,.SD,.SDcols = setdiff(names(tmt_10),TMTSubset)]
  
  #' Different Experiments:
  if(all(grepl("2021_10_09",PBspecific$`Raw file`))){
    TMTSubset <- TMTSubset[,.SD,.SDcols=grep("Plex[12]$",names(TMTSubset))]
    
    #'Renaming and Assigning of the channels to the corresponding fractions of the experiments
    dim(TMTSubset)
    coN <- colnames(TMTSubset)
    
    # keeping naming of previouse experimental design. Naming has therefor no meaning  
    Odds <- grepl("Plex2",coN)
    Evens <- grepl("Plex1",coN)
    
    VectorEven <- c(5,  11,  13,  8,  12,  15,  2,  9,  6,  14,  10,  3,  7,"P1_Std1",    4,"P1_Std2")
    VectorOdd <- c(19,25,27,22,26,29,16,23,20,28,24,17,21,"P2_Std1",18,"P2_Std2")
    
    
  }else{
    # TMTSubset <- TMTSubset[,.SD,.SDcols=grep("[dn]$|ChymPlex[12]$",names(TMTSubset))]
    
    #'Renaming and Assigning of the channels to the corresponding fractions of the experiments
    dim(TMTSubset)
    coN <- colnames(TMTSubset)
    Odds <- grepl("perOdd",coN)
    Evens <- grepl("perEven",coN)
    VectorEven <- c( "20", "4", "12", "28", "10", "6", "14", "22", "18", "2", "16", "8", "26", "P1_Std1", "24", "P1_Std2")
    VectorOdd <-  c( "15", "29", "7", "13", "25", "9", "3", "17", "21", "27", "23", "11", "19", "P2_Std1", "5", "P2_Std2")
    
    
    
    
  }
  
  
  # Compiling  OrderVecotr
  sapply(1:length(VectorEven),function(x){
    VecString <- paste(" ",x,"$",sep = "")
    String <- grepl(VecString,coN)
    
    coN[String&Evens] <<- VectorEven[x]
    coN[String&Odds] <<- VectorOdd[x]
    
  })
  
  #' Final Matrix preparation
  colnames(TMTSubset) <- coN
  coN_numeric <- coN[order(as.numeric(coN))]
  # coN_numeric <- coN_numeric[!is.na(coN_numeric)]
  setcolorder(TMTSubset,coN_numeric)
  names(TMTSubset)
  TMTSubset_final <- TMTSubset[,.SD,.SDcols=grep("Std",names(TMTSubset),invert = T,value = T)]
  
  #' Normalization
  #' 
  TMTSubset_STD <- TMTSubset[,.SD,.SDcols=grep("Std",names(TMTSubset),invert = F,value = T)]
  list(TMTSubset_STD=TMTSubset_STD,
       TMTSubset_final=TMTSubset_final,
       VectorEven=VectorEven,
       VectorOdd=VectorOdd,
       Info = Infotable,
       xname = x,
       x = x)

})

# Neighbour Mean imputation
stopCluster(cl)
library(parallel)
cl <- makeCluster(1,outfile="")
applying_neighbourZeroImputation <- T
clusterExport(cl,list("Neighbour_Mean_imputation","applying_neighbourZeroImputation"))
PBonly_MQNorm <- parLapply(cl,PBonly_MQ,function(x){
  library(data.table)
  cat("\r Normalisation",x$xname)
  p1 <- x$TMTSubset_STD$P1_Std2
  p2 <- x$TMTSubset_STD$P2_Std2
  TMTSubset_final <- x$TMTSubset_final
  # Sorting Vector
  inc_cols_Even <- sapply(as.numeric(x$VectorEven)[!is.na(as.numeric(x$VectorEven))],function(y){
    which(names(x$TMTSubset_final)==y)
  })
  inc_cols_Odd <- sapply(as.numeric(x$VectorOdd)[!is.na(as.numeric(x$VectorEven))],function(y){
    which(names(x$TMTSubset_final)==y)
  })
  inc_cols_Even <- sort(inc_cols_Even)
  inc_cols_Odd <- sort(inc_cols_Odd)
  # Correction Matrix
  corM <- matrix(p2,ncol=dim(TMTSubset_final)[2],nrow=dim(TMTSubset_final)[1],byrow = F)
  
  for(i in inc_cols_Even){
    corM[,i] <- p1
  }
  su <- apply(pall <- cbind(p2,p1),1,sum,na.rm = T)
  pall <- data.table(pall)
  pall$su <- su
  pallagg<- pall[,{
    list(CorrectionFac=mean(p2,na.rm = T)/mean(p1,na.rm = T),bin = mean(su,na.rm = T))
  },cut(su,50)]
  
  corMave <- matrix(mean(p2,na.rm = T),ncol=dim(TMTSubset_final)[2],nrow=dim(TMTSubset_final)[1],byrow = F)
  corMave[,inc_cols_Even] <- mean(p1,na.rm = T)
  
  sel <- apply(corM,1,function(x){any(is.na(x))|any(x==0)})
  if(any(sel)){
    corM[sel,] <- corMave[sel,]
  }
  
  # converting to matrix
  #TMTSubset_final_singlenorm is the single peptide normalized matrix, peptides with missing values in one of the plexes will be treated with an average correction factor
  TMTSubset_final_singlenorm <- copy(TMTSubset_final)
  TMTSubset_final_singlenorm <- as.matrix(TMTSubset_final_singlenorm)
  
  p1Signal <- apply(TMTSubset_final_singlenorm[,inc_cols_Even],1,sum,na.rm = T)
  p2Signal <- apply(TMTSubset_final_singlenorm[,inc_cols_Odd],1,sum,na.rm = T)
  x$InBothPlexes <- p1Signal!=0&p2Signal!=0
  # barplot(Signal <- table(p1Signal!=0&p2Signal!=0),beside = F,main="Signal in both plexes")
  TMTSubset_final_singlenorm <- TMTSubset_final_singlenorm/corM
  
  # Imputing Mean in single zeros between good measurements
  diffinfo1 <- apply(TMTSubset_final,1,function(x){sum(abs(diff(x)),na.rm = T)/max(x,na.rm = T)})
  diffinfo2 <- apply(TMTSubset_final_singlenorm,1,function(x){sum(abs(diff(x))/max(x,na.rm = T),na.rm = T)})
  
  if(!grepl("2021_10_09",x$xname)&applying_neighbourZeroImputation){
    print("applying Neighbour_Mean_imputation")
    TMTSubset_final_singlenorm <- Neighbour_Mean_imputation(TMTSubset_final_singlenorm)
    TMTSubset_final_singlenorm <- data.table(TMTSubset_final_singlenorm)
    diffinfo3 <- apply(TMTSubset_final_singlenorm,1,function(x){sum(abs(diff(x))/max(x,na.rm = T),na.rm = T)})
    diffinfo <- cbind(diffinfo1,diffinfo2,diffinfo3)
  }else{
    diffinfo <- cbind(diffinfo1,diffinfo2,diffinfo2)
  }
  # 
  x$diffinfo <- diffinfo
  
  TMTSubset_final_avenorm <- copy(TMTSubset_final)
  TMTSubset_final_avenorm <- as.matrix(TMTSubset_final_avenorm)
  TMTSubset_final_avenorm <- TMTSubset_final_avenorm/corMave
  TMTSubset_final_avenorm <- data.table(TMTSubset_final_avenorm)
  
  
  # zscoring on GeneLevel
  TMTSubset_final_singlenorm_Mscale <- copy(TMTSubset_final_singlenorm)
  TMTSubset_final_singlenorm_Mscale <- apply(TMTSubset_final_singlenorm_Mscale,1,function(x){x/mean(x,na.rm = T)})
  TMTSubset_final_singlenorm_Mscale <- t(TMTSubset_final_singlenorm_Mscale)
  
  TMTSubset_final_singlenorm_zscore <- copy(TMTSubset_final_singlenorm)
  TMTSubset_final_singlenorm_zscore <- apply(TMTSubset_final_singlenorm_zscore,1,function(x){
    sdfun <- sd(x,na.rm = T)
    if(is.na(sdfun)){
      sdfun <- 1
    }
    (x-mean(x,na.rm=T))/sdfun
    
  })
  # sel <- which(x$InputTable$Sequence=="LHSISSIDVNGGNRLHSISSIDVNGGNR")
  # plot(TMTSubset_final_singlenorm[sel,],type="l")
  # plot(unlist(TMTSubset_fCinal[sel,]),type="l")
  
  TMTSubset_final_singlenorm_zscore <- t(TMTSubset_final_singlenorm_zscore)
  
  x$TMTSubset_final_singlenorm <- data.table(TMTSubset_final_singlenorm)
  x$TMTSubset_final_avenorm <- data.table(TMTSubset_final_avenorm)
  x$TMTSubset_final_singlenorm_Mscale <- data.table(TMTSubset_final_singlenorm_Mscale)
  x$TMTSubset_final_singlenorm_zscore <- data.table(TMTSubset_final_singlenorm_zscore)
  x
})
stopCluster(cl)



PBonly_MQNorm_dt <- rbindlist(lapply(PBonly_MQNorm,function(x){
  x <- x
  tmt <- x$TMTSubset_final_singlenorm_zscore
  tmt$Seq <- x$Info$Sequence
  tmt$Gel <- x$xname
  ENSG <- paste(sort(unique(unlist(strsplit(x$Info$ENSG,";|#")))),collapse=";")
  tmt$Proteins <- x$Info$Proteins
  tmt$ENSG <- x$Info$ENSG
  tmt$PB_Only <- x$Info$PB_Only  
  tmt$UniProt <- x$Info$UniProt  
  tmt$UniProt <- x$Info$NotIn_UniProt  
  
  tmt
  
  }))
PBonly_MQ_dt_melt <- melt(PBonly_MQNorm_dt,measure.vars = paste("V",as.character(1:28),sep=""))
table(gsub(".*. ","",PBonly_MQ_dt_melt$variable))
ggplot(PBonly_MQ_dt_melt[ENSG==ENSG[2]],aes(as.numeric(gsub("V","",variable))   ,as.numeric(value),group=Seq))+geom_line()+facet_grid("Gel",scales = "free_y")

# Predicting mz
PBonly_MQ_dt_melt$MW <- NA

if(!exists("Models")){
  # load("../IsoFracCompiler/MW_Models_v3.rda")
  print("WARNING: Loading Models Data, make sure, that the path is using the most recent version!")
  load("../IsoFracCompiler2_noSemitryptic/Experiments/MW_Models_v3.rda")
  # list.files("../",pattern="MW_Models_v3.rda",recursive = T)

}
for(i in c("5per|5er","8per","10per")){
  sel <- grepl(i,PBonly_MQ_dt_melt$Gel)
  PBonly_MQ_dt_melt$MW[sel] <- predict(Models[[which(grepl(i,names(Models)))]]$MW_Predictor, as.numeric(PBonly_MQ_dt_melt$variable[sel]))
}

PBonly_MQ_dt_melt$Type <- ""
PBonly_MQ_dt_melt$Type[!is.na(match(PBonly_MQ_dt_melt$Proteins,PB_Only_Candidates))] <- "PB_Only"
PBonly_MQ_dt_melt$Type[!is.na(match(PBonly_MQ_dt_melt$Seq,NotInUniprotFasta))] <- "Pacbio_Fasta"

table(PBonly_MQ_dt_melt$Type)
PBonly_MQ_dt_melt$Gel <- factor(PBonly_MQ_dt_melt$Gel,c("5per","8per","10per"))
PBonly_MQ_dt_melt$PB_Only <- factor(PBonly_MQ_dt_melt$PB_Only,c(T,F))

### functions
# Load necessary libraries
library(ggplot2)
library(stringr)

#### Chatgpt function for plotting the mapped peptide sequences in a given protein sequence
visualize_peptide_mapping <- function(protein_sequence, peptides, peptide_colors, line_break = 50,charactersize=3) {
  # Initialize variables
  protein_length <- nchar(protein_sequence)
  color_vector <- rep("grey", protein_length)
  
  # Ensure peptide_colors has the same length as peptides
  if (length(peptide_colors) != length(peptides)) {
    stop("The length of 'peptide_colors' must match the length of 'peptides'.")
  }
  
  # For each peptide, find its position in the protein sequence and assign the corresponding color
  for (i in seq_along(peptides)) {
    peptide <- peptides[i]
    positions <- unlist(gregexpr(peptide, protein_sequence, fixed = TRUE))
    peptide_color <- peptide_colors[i]
    # If peptide is found, color the positions
    if (any(positions != -1)) {
      for (pos in positions) {
        if (pos != -1) {
          color_vector[pos:(pos + nchar(peptide) - 1)] <- peptide_color
        }
      }
    }
  }
  
  # Split the sequence into lines based on line_break parameter
  split_sequence <- strsplit(protein_sequence, NULL)[[1]]
  num_lines <- ceiling(protein_length / line_break)
  line_positions <- rep(1:num_lines, each = line_break)[1:protein_length]
  position_in_line <- rep(1:line_break, num_lines)[1:protein_length]
  
  # Create a data frame for plotting
  protein_df <- data.frame(
    line = line_positions,
    position_in_line = position_in_line,
    residue = split_sequence,
    color = color_vector
  )
  
  # Create the plot
  g <- ggplot(protein_df, aes(x = position_in_line, y = -line, label = residue, color = color)) +
    geom_text(size = charactersize) +
    scale_color_identity() +  # Use the actual color values provided in the data frame
    theme_void() +
    theme(legend.position = "none") +
    # labs(title = "Peptide Mapping to Protein Sequence") +
    scale_x_continuous(breaks = seq(1, line_break, by = 10))  # Adjust axis breaks as needed
  
  list(gg=g,table=protein_df)
}

####
Traces_Prot_Pure <- function(GNlist,
                             DBpath = "/Users/henno/Documents/Skripte/R_Scripts/DataAnalysis/20210129_AK_ProteinIsoforms/ProteinSequencesCategories.rda",
                             sql = "/Users/henno/Documents/Skripte/R_Scripts/DataAnalysis/20210129_AK_ProteinIsoforms/Experiments/Viewer83_Public/TableListNormEnsHdbScanMW_Peaks.sqlite",
                             setsize=1.5,
                             TotalWidth=40
){
  
  library(ggplot2)
  library(parallel)
  library(msa)
  library(RSQLite)
  library(data.table)
  library(plotly)
  library(Peptides)
  EXAMPLE_ENSG <- GNlist$EXAMPLE_ENSG
  GNname <- GNlist$GN
  gel <- GNlist$gel
  UniprotSelections <<- GNlist$Uniprot
  AlternativeNames <<- GNlist$UniprotName
  AdditionalMarker <- GNlist$AdditionalMarker
  
  SequenceGroups = GNlist$SequenceGroups
  
  if(length(GNlist$SeqWidth)>0){
    TotalWidth <- GNlist$SeqWidth
  }
  # EXAMPLE_ENSG <- c("ENSG00000172216")
  
  # DBpath <-"/Users/henno/Documents/Skripte/R_Scripts/DataAnalysis/20210129_AK_ProteinIsoforms/ProteinSequencesCategories.rda"
  if(!exists("PDB")){
    PDB <- load(DBpath)
    
  }
  ## functions ------------------
  Traces_Plotly <- function(input,ens="ENSG00000120802",mpwcutoff=0,converttoMW=T,cbPalette= c( "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7"),PurePlot =F,showlegend=T){
    
    ens <<- ens
    hdb <- input$hdbscanTables[EnsG==ens]
    tempValues <- input$TMTSubset_final_singlenorm_zscore[hdb$id,]
    
    tempValues$cl <- hdb$cl_2
    tempValues$id <- hdb$id
    if(any(grepl("^V[0-9]",colnames(tempValues)[1]))){
      me <- melt(tempValues,measure.vars = paste("V",1:28,sep = ""),id.vars = c("cl","id"))
      me$variable <- as.numeric(gsub("V","",me$variable))
    }else{
      
      me <- melt(tempValues,measure.vars = as.character(2:29),id.vars = c("cl","id"))
      me$variable <- as.numeric(as.character(me$variable))
      me$variable <- me$variable-min(me$variable)+1
      input <<- input
      me <<- me
      # plot(me$variable2,as.numeric(as.character()))
      # abline(0,1)
      (table(me$variable))
      # stop()
    }
    
    
    if(length(input$Fractions_predicted_MW)>0&converttoMW){
      mds <- input$Fractions_predicted_MW[1:28]
      # mds <- 10^predict(md,2:29)
      # mds <- 10^predict(md,1:28)
      # table(me$variable)
      # (table(me$variable))
      me$variable <- 10^mds[match(me$variable,1:28)]
      # me <- me[order(variable)]
    }
    
    
    
    # med <-  dcast(me,variable~paste(cl,id,sep ="_"),value.var = "value")
    # fig <- plot_ly(med,x=~variable,y=~names(med)[2],type = 'scatter', mode = 'lines')
    # for(i in unique(me$cl)){
    #   add <- grep(paste("^",i,"_",sep = ""),names(med),value=T)
    #   for(u in add){
    #     fig <- fig%>%add_trace(y=add)
    #   }
    # }
    me <- rbind(me,me[,.(value=max(me$value)+1,variable=NA),.(id,cl)])
    
    me$cl <- (factor(me$cl))
    me$seq <- input$dt_gn$seq[me$id]
    me <- me[,.SD[order(variable)],id]
    me <- me[order(id),]
    me[,idtemp:={.GRP},.(cl,id)]
    try({me <- me[order(idtemp),]})
    try({me <- me[order(seq),]})
    try({me <- me[as.numeric(order(cl)),]})
    try({me <- me[order(id),]})
    
    # me <- me[,.SD[order(as.numeric(id))],id]
    
    idCount <- 0
    idRemove <- c()
    for(i in unique(me$cl)){
      id <- me[cl==i]$id
      id <- unique(id)
      id <- (1:length(id))+idCount
      idCount <- max(id)
      idRemove <- c(id[-1])
      
      
    }
    cbp <- colorRampPalette(cbPalette)(max(as.character(me$cl)))
    cbp <- c("#73FDFF",cbp)
    if(PurePlot){
      print("Hu")
      cbp <- "#00000080"
      me$color <- "black"
      me$name <- ""
      me$legendgroup <- ""
      
      
      
      # me$cl <- ""
    }else{
      me$color <- cbp[as.numeric(as.character(me$cl))+1]
      
    }
    # me <- me[color=="#E69F00"]
    me <<- me
    p <- plot_ly(me, 
                 x = ~variable, 
                 y = ~value, 
                 # id = ~id,
                 color=~cl,
                 name=~cl,
                 legendgroup=~cl,
                 type = 'scatter',
                 mode = 'lines',
                 hoverinfo="text",
                 text=me$seq,
                 colors=cbp)%>%
      layout(showlegend = TRUE) %>%
      style(showlegend = FALSE, traces = idRemove)
    # if(PurePlot){
    #   p <- p%>%layout(title = ens, xaxis = list(title = 'MW converted fraction'), 
    #                   yaxis = list(title = "zscore"),showlegend=showlegend)
    # }else{
    # }
    p <- p%>%layout(title = ens, xaxis = list(title = 'MW converted fraction'), 
                    yaxis = list(title = "zscore"), legend = list(title=list(text='hdbscan results')),showlegend=showlegend)
    
    
    list(p=p,data=me,clusterColors=cbp) 
  }
  Traces_Plotly2 <- function(input,ens="ENSG00000120802",mpwcutoff=0,converttoMW=T,cbPalette= c( "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7"),PurePlot =F,showlegend=T,MeanCluster=F,ShowQuantiles=T,colorClusters=NULL,colorSequences=NULL){
    
    ens <<- ens
    hdb <- input$hdbscanTables[EnsG==ens]
    tempValues <- input$TMTSubset_final_singlenorm_zscore[hdb$id,]
    
    tempValues$cl <- hdb$cl_2
    tempValues$id <- hdb$id
    tempValues$sequences <- hdb$Seq
    if(any(grepl("^V[0-9]",colnames(tempValues)[1]))){
      me <- melt(tempValues,measure.vars = paste("V",1:28,sep = ""),id.vars = c("cl","id"))
      me$variable <- as.numeric(gsub("V","",me$variable))
    }else{
      
      me <- melt(tempValues,measure.vars = as.character(2:29),id.vars = c("cl","id"))
      me$variable <- as.numeric(as.character(me$variable))
      me$variable <- me$variable-min(me$variable)+1
      input <<- input
      me <<- me
      # plot(me$variable2,as.numeric(as.character()))
      # abline(0,1)
      # length(table(me$variable))
      # stop()
    }
    
    
    if(length(input$Fractions_predicted_MW)>0&converttoMW){
      mds <- input$Fractions_predicted_MW[1:28]
      # mds <- 10^predict(md,2:29)
      # mds <- 10^predict(md,1:28)
      # table(me$variable)
      # (table(me$variable))
      me$variable <- 10^mds[match(me$variable,1:28)]
      # me <- me[order(variable)]
    }
    
    
    
    # med <-  dcast(me,variable~paste(cl,id,sep ="_"),value.var = "value")
    # fig <- plot_ly(med,x=~variable,y=~names(med)[2],type = 'scatter', mode = 'lines')
    # for(i in unique(me$cl)){
    #   add <- grep(paste("^",i,"_",sep = ""),names(med),value=T)
    #   for(u in add){
    #     fig <- fig%>%add_trace(y=add)
    #   }
    # }
    me <- rbind(me,me[,.(value=max(me$value,na.rm=T)+1,variable=NA),.(id,cl)],fill=T)
    
    me$cl <- (factor(me$cl))
    me$seq <- input$dt_gn$seq[me$id]
    me$match <- 0
    me <- me[,.SD[order(variable)],id]
    me <- me[order(id),]
    me[,idtemp:={.GRP},.(cl,id)]
    try({me <- me[order(idtemp),]})
    try({me <- me[order(seq),]})
    try({me <- me[as.numeric(order(cl)),]})
    try({me <- me[order(id),]})
    
    # me <- me[,.SD[order(as.numeric(id))],id]
    
    idCount <- 0
    idRemove <- c()
    for(i in unique(me$cl)){
      id <- me[cl==i]$id
      id <- unique(id)
      id <- (1:length(id))+idCount
      idCount <- max(id)
      idRemove <- c(id[-1])
      
      
    }
    
    if(MeanCluster){
      print(dim(me))
      me <- me[,{list(value=mean(value,na.rm = T),q90=quantile(value,0.9,na.rm = T),q10=quantile(value,0.1,na.rm = T),n=length(value),seq=paste(unique(seq),collapse="\n"),idtemp=.GRP,id=.GRP)},.(cl,variable)]
      print(dim(me))
    }
    
    cbp <- colorRampPalette(cbPalette)(max(as.character(me$cl)))
    cbp <- c("#73FDFF",cbp)
    alphaset <- NULL
    if(PurePlot){
      cbp <- "#A9A9A9"
      # cbp <- "red"
      me$color <- cbp
      me$name <- ""
      me$legendgroup <- ""
      me$match <- 0
      if(length(colorClusters)>0){
        # me <<- me
        # cbp <- "#00000060"
        sel <- (!is.na(match(me$cl,colorClusters)))
        me$color <- sel
        me$match <- as.numeric(sel)
        
        cbp <- c(cbp,"tomato")
        alphaset <- 0.7
        
      }
      if(length(colorSequences)>0){
        me <<- me
        # cbp <- "#00000040"
        sel <- (!is.na(match(me$seq,colorSequences)))
        me$match <- as.numeric(sel)
        me$color <- sel
        cbp <- c(cbp,"tomato")
        alphaset <- 0.7
        me$alphaset <- c(0.2,1)[as.numeric(sel)+1]
        print(alphaset)
      }
      # me$cl <- ""
    }else{
      me$color <- cbp[as.numeric(as.character(me$cl))+1]
      
    }
    me <- me
    
    if(ShowQuantiles&MeanCluster==T){
      me$q90 <- me$q90-me$value
      me$q10 <- me$value-me$q10
    }else{
      me$q90 <-NA
      me$q10 <-NA
      
    }
    
    # me <<- me
    me <- me[order(match,decreasing = F)] 
    meleftover <- me[match!=0]
    if(any(me$match>0)){
      # print("hu")
      clvec <- c(unique(me[match==0]$cl),unique(me[match>0]$cl))
      colvec <- c(unique(me[match==0]$color),unique(me[match>0]$color))
      
      me$cl <- factor(me$cl,clvec)
      me$color <- factor(me$color,colvec)
      
    }
    
    me <- me[match==0]
    
    p <- plot_ly(me, 
                 x = ~variable, 
                 y = ~value, 
                 # id = ~id,
                 color=~color,
                 name=~cl,
                 legendgroup=~cl,
                 type = 'scatter',
                 mode = 'lines',
                 hoverinfo="text",
                 text=me,
                 colors=cbp,
                 line=list(width=3),
                 alpha=alphaset,
                 error_y=list(
                   type="data",
                   symmetric=F,
                   array=~q90,
                   arrayminus=~q10,
                   width=2           ))%>%
      layout(showlegend = TRUE) 
    
    try(p <- p %>%
          style(showlegend = FALSE, traces = idRemove))
    if(dim(meleftover)[1]>0&1){
      # stop()
      meleftover <<- meleftover
      # p <<- p
      p <- p%>%add_trace(
        x = meleftover$variable,
        y = meleftover$value,
        # id = ~id,
        color=meleftover$color,
        name=meleftover$cl,
        legendgroup=meleftover$cl,
        type = 'scatter',
        mode = 'lines',
        hoverinfo="text",
        text=meleftover$seq,
        colors=cbp,
        line=list(width=3),
        alpha=1,
        error_y=list(
          type="data",
          symmetric=F,
          array=meleftover$q90,
          arrayminus=meleftover$q10,
          width=2
        ), toback = T)
      
      # stop()
    }
    # p
    # 
    # p <- plot_ly(me, 
    #              x = ~variable, 
    #              y = ~value, 
    #              # id = ~id,
    #              color=~cl,
    #              name=~cl,
    #              legendgroup=~cl,
    #              type = 'scatter',
    #              mode = 'lines',
    #              hoverinfo="text",
    #              text=me$seq,
    #              colors=cbp)%>%
    #   layout(showlegend = TRUE) %>%
    #   style(showlegend = FALSE, traces = idRemove)
    # if(PurePlot){
    #   p <- p%>%layout(title = ens, xaxis = list(title = 'MW converted fraction'), 
    #                   yaxis = list(title = "zscore"),showlegend=showlegend)
    # }else{
    # }
    p <- p%>%layout(title = ens, xaxis = list(title = 'MW converted fraction'), 
                    yaxis = list(title = "zscore"), legend = list(title=list(text='hdbscan results')),showlegend=showlegend)
    
    me <<- me
    
    me$Group <- as.character(me$cl)
    # me$cl[1:24] <-
    if(!PurePlot){
      g2 <- ggplot(me[!is.na(me$variable)&cl!=0&cl!=-1],aes(as.numeric(variable)/1000,value,group=id,color = Group))+geom_line()
      
    }else{
      g2 <- ggplot(me[!is.na(me$variable)&cl!=0&cl!=-1],aes(as.numeric(variable)/1000,value,group=id))+geom_line(color="black",alpha=0.5,lwd=1)
      
    }
    # g2 <- g2+ggplot2::scale_color_viridis_b()
    zerocluster <- me[!is.na(me$variable)&(cl==0|cl==-1)]
    if(dim(zerocluster)[1]>0){
      for(i in unique(zerocluster$seq)){
        sel <- zerocluster$seq==i
        # g2 <- g2+annotate("line",zerocluster$variable[sel],zerocluster$value[sel],color="grey40")
        
      }
      
    }
    
    
    list(p=p,data=me,clusterColors=cbp,g2=g2) 
  }
  
  Traces_GGplot <- function(input,ens="ENSG00000120802",mpwcutoff=0,converttoMW=T,PurePlot = T,showlegend = F){
    hdb <- input$hdbscanTables[EnsG==ens]
    tempValues <- input$TMTSubset_final_singlenorm_zscore[hdb$id,]
    
    tempValues$cl <- hdb$cl_2
    tempValues$id <- hdb$id
    if(any(grepl("^V[0-9]",colnames(tempValues)[1]))){
      me <- melt(tempValues,measure.vars = paste("V",1:28,sep = ""),id.vars = c("cl","id"))
      me$variable <- as.numeric(gsub("V","",me$variable))
    }else{
      me <- melt(tempValues,measure.vars = 2:29,id.vars = c("cl","id"))
      
    }
    
    if(length(input$MW_Predictor)>0&converttoMW){
      md <- input$MW_Predictor
      mds <- 10^predict(md,2:29)
      me$variable <- mds[match(me$variable,2:29)]
    }
    if(PurePlot){
      g2 <- ggplot(me,aes(as.numeric(variable),value,group=id,color = as.character(cl)))+geom_line(color="black")
      
    }else{
      g2 <- ggplot(me,aes(as.numeric(variable),value,group=id,color = as.character(cl)))+geom_line()
      
      
    }
    
    list(p=g2+xlab("MW")+ylab("zscore")+ggtitle(ens),data=me,clusterColors=NA) 
  }
  plot.sequence <- function(IFUN,selected=IFUN$ENSEMBL_PROTEIN[1],ClustColor=NULL,ProtSequence=NULL){
    ClustColor <<- ClustColor
    IFUN <<- IFUN
    selected <<- selected
    ProtSequence <<- ProtSequence
    it <- 1
    IFUN$Sequence <- IFUN$Seq
    IFUN$Cluster <- IFUN$cl_2
    # tempi <- sapply(IFUN$Seq,function(x){
    #   grep(x,UniSequences_all$SEQUENCE,fixed = T)
    # })
    IFUN$Sequence <- IFUN$Seq
    IFUN$Cluster <- IFUN$cl_2 
    ClustColorMatch <- ClustColor[match(IFUN$Cluster,ClustColor$cluster2),]
    
    IFUN$Color <- ClustColorMatch$Color
    if(length(IFUN$Color)==0){
      IFUN$Color <- "grey"
      IFUN$color <- "grey"
      
    }
    IFUN$Peaks <- ClustColorMatch$V1
    setdiff(Positions$Position,nchar(ProtSequence))
    Positions <- IFUN[,{
      cat("\r",.GRP)
      sequ <- .BY$Sequence
      gr <- gregexpr(sequ,ProtSequence,fixed = T)
      
      print(gr)
      Posis <- gr[[1]]:(gr[[1]]+attributes(gr[[1]])$match.length-1)
      
      Col <- 0
      if(length(ClustColor)>0){
        Col <- ClustColor$Color[match(Cluster,ClustColor$Cluster)]
      }
      Col <- unique(Col)
      # if(length(Col)>1){
      #   Col <- "#888888"
      # }
      data.table(Position=Posis)
    },.(Sequence,Cluster,Color)]
    
    # Positions$Color <- "grey"
    
    
    Positions$Color[is.na(Positions$Color)] <-   "black"
    
    # Positions <- Positions[Position>0,]
    
    Positions <- Positions[,.(co=paste(unique(Color),collapse = ";"),cl=paste(unique(Cluster),collapse = ";")),Position]
    
    Positions <- Positions[Position>0]
    Positions <- Positions[!is.na(Position)]
    
    LE <- strsplit(ProtSequence,"")
    maxLe <- 40
    LEs <- lengths(LE)
    
    
    yrow <- ceiling(LEs/maxLe)[it]
    vec <<- rep(0,(yrow*maxLe))
    # vec[1:10] <- 1
    Positions <<- Positions
    vec[Positions$Position] <- (Positions$cl)
    # vec[is.na(vec)] <- max(vec,na.rm = T)+1
    try({vec <- as.numeric(factor(as.character(vec),unique(vec),1:length(unique(vec))))})
    vec[is.na(vec)] <- 0
    # vec[is.na(vec)] <- max(vec,na.rm = T)+1
    vec <<- vec
    MA <- matrix(rev(vec),nrow = yrow  ,ncol = maxLe,byrow = T)
    MA <- apply(MA,1,rev)
    co <- unique(Positions$co)
    co <- co[grep(";",co,invert =T)]
    par(mai=c(0.8,0.1,0.5,2))
    image((MA),main="",col=c("#FFFFFF",ClustColor$Color),axes=F)
    mtext(selected,3,line = 1)
    ClustColorLeg <- ClustColor
    ClustColorLeg$cluster2 <- NULL
    ClustColorLeg <- unique(ClustColorLeg)
    legend(par()$usr[2],par()$usr[4],xjust = 0,legend = c(as.character(ClustColorLeg$V1),">x"),fill=c(ClustColorLeg$Color,"#33333340"),xpd=NA,bty="n",cex = 0.8)
    legend(par()$usr[1],par()$usr[3],xjust = 0,legend = c(as.character(ClustColorLeg$V1),">x"),fill=c(ClustColorLeg$Color,"#33333340"),horiz = T,xpd=NA,bty="n")
    
    y <<- 0
    x <<- 0
    sapply(1:length(LE[[it]]),function(ite){
      cat("\r",x)
      
      
      text(x/(maxLe-1),1-y/(yrow-1),LE[[it]][ite],xpd=NA)
      
      x <<- x+1
      if(x%%maxLe==0){
        
        y<<-y+1
        x <<- 0
      }
      
    })
  }
  MappedProteinSequences <- function(P,IDs,TotalWidth =50){
    
    P$data$Sequence <-  P$data$seq
    P$data$Cluster  <-  P$data$cl
    P$data$Color  <-  P$data$cl#as.numeric(factor(P$data$cl))
    
    
    MappedData <- lapply(1:dim(IDs)[1],function(it){
      it <<- it
      ProtSequence <- IDs$SEQUENCE[it]
      
      P$data$SequenceComparisson <- ProtSequence
      
      singlefun <-  strsplit(ProtSequence,"")
      dt <- data.table(Sequence="",Cluster=-1,Color=-1,Position=1:length(unlist(singlefun)),AA=unlist(singlefun),ProtSequence=ProtSequence)
      
      Positions <- P$data[,{
        cat("\r",.GRP)
        sequ <<- .BY$Sequence
        gr <- gregexpr(sequ,ProtSequence,fixed = T)
        
        print(gr)
        Posis <- gr[[1]]:(gr[[1]]+attributes(gr[[1]])$match.length-1)
        
        Col <- 0
        # if(length(ClustColor)>0){
        #   Col <- ClustColor$Color[match(Cluster,ClustColor$Cluster)]
        # }
        Col <- unique(Col)
        # if(length(Col)>1){
        #   Col <- "#888888"
        # }
        data.table(Position=Posis)
      },.(Sequence,Cluster,Color)]
      po <- Positions$Position
      po[po<1] <- NA
      Positions$ProtSequence <- ProtSequence
      Positions$AA <- singlefun[[1]][po]
      dt <- dt
      Positions <- Positions
      PositionsCombi <- rbind(dt,Positions)
      PositionsCombi$ProtName <- IDs$ProteinID[it]
      PositionsCombi
    })
    
    MappedData <- rbindlist(MappedData)
    # stop()
    
    MappedData$Color[is.na(MappedData$Color)] <-   -1
    
    # TotalWidth <- 50
    
    MappedDataAdjusted <- MappedData[,{
      temp <- .SD
      temp <- temp[!is.na(temp$AA)]
      print(dim(temp))
      Row  <- ceiling(temp$Position/TotalWidth)
      Col <- temp$Position-TotalWidth*(Row-1)
      temp$x <- Col
      temp$y <- Row
      # plot(temp$x,
      #      temp$y,type="none")
      # text(temp$x,temp$y,temp$AA)
      # print(dim(temp))
      temp
      
    },.(ProtSequence,ProtName)]
    MappedDataAdjusted
  }
  
  
  ## script
  
  # example plots-----------------
  # sql <- "/Users/henno/Documents/Skripte/R_Scripts/DataAnalysis/20210129_AK_ProteinIsoforms/Experiments/Viewer83_Public/TableListNormEnsHdbScanMW_Peaks.sqlite"
  if(!exists("db")){
    db <- dbConnect(SQLite(),sql)
    dbl <- dbListTables(db)
    gels <- dbReadTable(db,"Gels")
  }
  dbReadTable(db,"CombinedGels_Compressed_Pairs")
  selstring <- grep(gels$gels[gel],dbl,value=T)
  GN <- dbReadTable(db,"GN")
  
  cp <- data.table(dbReadTable(db,"CombinedGels_Pairs_Filtered"))
  
  # dbset <- dbReadTable(db,)
  rm(Candidates)
  Candidates <- as.data.table(cp)[ENSG==EXAMPLE_ENSG,median(as.numeric(Height_lm)),.(IsoFrac_MW,gel,ENSG,DeltaMass)]
  Candidates$MW <-as.numeric(Candidates$IsoFrac_MW )/1000
  Candidates <<- Candidates
  ensVec <- EXAMPLE_ENSG
  
  input <- list(hdbscanTables=dbReadTable(db,grep("hdbscan",selstring,value=T)),
                TMTSubset_final_singlenorm_zscore=dbReadTable(db,grep("zscore",selstring,value=T)),
                Fractions_predicted_MW=dbReadTable(db,grep("Fractions",selstring,value=T)),
                dt_gn=GN[grep(gels$gels[gel],GN$gel),])
  input <- lapply(input,as.data.table)
  # 
  # input$hdbscanTables <- input$hdbscanTables[EnsG==ensVec]
  # sel <- input$dt_gn$ens==ensVec
  # input$TMTSubset_final_singlenorm_zscore <- input$TMTSubset_final_singlenorm_zscore[sel]
  # input$dt_gn <- input$dt_gn[sel]
  # 
  
  # setwd(dirname(rstudioapi::documentPath()))
  
  # hacking clustering to sequence selection any non selected peptides belong to the zero cluster
  input$hdbscanTables$cl_2 <- NULL
  input$hdbscanTables$cl_2 <- 0
  
  
  input$hdbscanTables$cl_2 <- 1#factor(input$hdbscanTables$cl_2,unique(input$hdbscanTables$cl_2),c(0:(length(unique(input$hdbscanTables$cl_2))-1)))
  
  
  P <- Traces_Plotly2(input,EXAMPLE_ENSG,PurePlot = T,converttoMW=T,showlegend = T,cbPalette = c("skyblue","tomato"),MeanCluster=F,ShowQuantiles = T)
  
  
  # P$p<<<<<<
  # P$g2
  # P$p
  # adding Sequence Plot
  # IDs <- ProteinSequencesCategories[ENSG==EXAMPLE_ENSG]
  # # MappedSequences <<- MappedProteinSequences(P, IDs,TotalWidth)
  # # MappedSequences <- MappedSequences[!is.na(match(ProtName,UniprotSelections))]
  # # MappedSequences$mw <- mw(MappedSequences$ProtSequence)
  # # kdinfo <- round(MappedSequences$mw[match(UniprotSelections,MappedSequences$ProtName)]/1000)
  # # if(length(AlternativeNames)==length(UniprotSelections)){
  # #   MappedSequences <- MappedSequences
  # #   UniprotSelections <- UniprotSelections
  # #   AlternativeNames <- AlternativeNames
  # #   MappedSequences$ProtName <- AlternativeNames[match(MappedSequences$ProtName,UniprotSelections)]
  # #   UniprotSelections <- AlternativeNames
  # # }
  # # MappedSequences$ProtName <- factor(MappedSequences$ProtName,UniprotSelections,paste(UniprotSelections,kdinfo,"kD"))
  # # # unique(MappedSequences$ProtName)
  # # 
  # # # g <- ggplot()
  # # # g+annotate("text",x=MappedSequences$x,MappedSequences$y,MappedSequences$AA)
  # # 
  # # MappedSequences$Color[MappedSequences$Cluster==0|MappedSequences$Cluster==-1] <- NA
  # # le <- length(unique(MappedSequences$Cluster)[unique(MappedSequences$Cluster)!="0"&unique(MappedSequences$Cluster)!="-1"])
  # D <- ggplot(MappedSequences[!is.na(AA)&Cluster!=0&Cluster!=-1,],aes(y=Position,fill=factor(Color,0:le,0:le)))+ylab("")+geom_density(alpha=0.5,bw=30)+theme_minimal()+scale_y_reverse()+facet_wrap("ProtName",ncol = 1,labeller = function(x){""})+theme(legend.position = "none",axis.text.x  = element_blank(),axis.text = element_text(size=3), axis.ticks.length.x  = unit(0, "mm"))+xlab("")
  # 
  
  
  ggClean <- function(g){
    g+theme_minimal()+theme(axis.line = element_blank(),
                            plot.background = element_blank(),
                            panel.grid.major = element_blank(),
                            panel.grid.minor = element_blank(),
                            panel.border = element_blank(),
                            axis.text  = element_blank()
                            , axis.ticks.length  = unit(0, "mm"),legend.position = "none")
  }
  
  
  
  
  # # graphics.off()
  # # if(length(AdditionalMarker)>0){
  # #   unique(MappedSequences$ProtSequence)
  # #   MappedSequences <<- MappedSequences
  # #   MappedSequences <- MappedSequences[,{
  # #     temp <- .SD
  # #     se <- temp[,unique(AA),Position]
  # #     posi <- paste(se$V1,collapse="")
  # #     humpi <- lapply(AdditionalMarker,function(x){
  # #       hu <-   unlist(gregexec(x$Pos,posi))
  # #       # xHuha  <<- x
  # #       # stop()
  # #       list(Position=hu,Name= x$name,adj=x$adj,col=x$col)
  # #     })
  # #     humpi <- rbindlist(humpi)
  # #     # print(humpi$col)
  # #     humpi <- humpi[,paste(unique(Name),collapse=";"),.(Position,adj,col)]
  # #     temp$AddedInfo <- humpi[match(temp$Position,humpi$Position)]$V1
  # #     temp$adj <- humpi[match(temp$Position,humpi$Position)]$adj
  # #     temp$LabelCol <- humpi[match(temp$Position,humpi$Position)]$col
  # #     
  # #     temp$AddedInfo[is.na(temp$AddedInfo)] <- NA
  # #     temp
  # #   },ProtSequence]
  # #   # stop()
  # #   
  # # }
  # # if(length(GNlist$SequenceSize)>0){
  # #   setsize <- 
  # #     GNlist$SequenceSize
  # # }
  # # # le <- 3
  # # library(ggrepel)
  # # # plottinhg Sequences---------
  # # MappedSequences <<- MappedSequences
  # # SequencePlot <- ggplot(MappedSequences[!is.na(AA),],aes(x,(y),label=AA,color=factor(Color,0:le,0:le)))+
  # #   geom_text(size=setsize)+scale_y_reverse()+
  # #   facet_wrap("ProtName",ncol = 1)+coord_cartesian(clip="off")
  # # if(length(MappedSequences$AddedInfo)>0){
  # #   SequencePlot <- SequencePlot +     
  # #     geom_label(aes(label=AddedInfo,hjust=adj,bg=LabelCol),color="white",nudge_y=0,alpha=0.9,size=setsize,label.size = 0.1,label.padding = unit(0.15, "lines")
  # #                ,show.legend = F)
  # # }
  # # 
  # # plot(SequencePlot)
  # 
  # 
  # # MappedSequences <<- MappedSequences
  # SequencePlot2 <- AlignMappedSequences(MappedSequences,TotalWidth=TotalWidth)+xlim(-20,TotalWidth)
  # temp <<- SequencePlot2
  # mw_lines <- round(mw(unique((MappedSequences$ProtSequence))))
  # #+theme(legend.position = "none")#+scale_color_viridis_d()
  # 
  library(cowplot)
  # xl <<- c(max(nchar(MappedSequences$ProtSequence)),1)
  # yl <- min(SequencePlot$data$y)
  # yl[1] <- yl[1]-diff(yl)*0.1
  # print(yl)
  # SeqPlotClean <<- ggClean(SequencePlot)+theme(legend.position = "none",plot.margin=unit(c(0,0,0,0), "pt"))+ylab("")+xlab("")
  # Dclean <<- D+theme(plot.margin=unit(c(1,10,1,-15), "pt"),legend.box.spacing = unit(0, "mm"),panel.spacing = unit(0, "mm"))
  # 
  
  
  
  temp <- P$g2+ggtitle(GNlist$GN)+ylab("zscore")+xlab("MW [kD]")
  
  ggplot_data <- ggplot_build(temp)
  x_limits <- ggplot_data$layout$panel_params[[1]]$x.range
  CandidatesSelect <- Candidates[MW>=min(x_limits)&MW<=max(x_limits)]
  CandidatesSelect$DM <- as.numeric(CandidatesSelect$DeltaMass)/max(as.numeric(CandidatesSelect$DeltaMass))*5
  CandidatesSelect$DM[is.na(CandidatesSelect$DM)] <- 0.1
  temp+
    annotate("point",CandidatesSelect$MW,CandidatesSelect$V1,shape=24,color="red",bg="red",size=4,linewidth=4)+
    annotate("point",CandidatesSelect$MW,CandidatesSelect$V1,shape=25,color="red",bg="red",size=4,linewidth=4)+
    annotate("text",CandidatesSelect$MW,CandidatesSelect$V1,shape=25,vjust=2,label=round(as.numeric(CandidatesSelect$DeltaMass)),color="red",bg="red",size=4,linewidth=4)
}



#visualize_peptide_mapping(protein_sequence, peptides,1:length(peptides), line_break = 20)

library(dbscan)
library(cowplot)
PBonly_MQ_dt_melt <- PBonly_MQ_dt_melt[,{
  temp <- .SD
  # t2 <<- temp
  # temp <- t2
  dc <- dcast(temp,Seq~variable,fun.aggregate = median)
  se <- dc$Seq
  dc$Seq <- NULL
  dc_m <- apply(dc,2,function(x){
    x <- as.numeric(x);
    x[is.na(x)] <- -1
    x
  })
  cl1 <- rep(-2,length(se))
  try({
    cl1 <- hdbscan(dc_m,2)$cluster
  },silent = T)
  # cl2 <- hdbscan(dist(dc_m),2)$cluster
  temp$cl <- as.double(cl1[match(temp$Seq,se)])
  temp
},ENSG]


pdf("PB_Only_Traces.pdf",width=14)
ggcandidates <- unique(PBonly_MQ_dt_melt$ENSG)
g <- ggcandidates[[1]]
g <- "ENSG00000067208"
for(g in grep(";",ggcandidates,invert=T,value=T)){
  print(g)
  pseqs <- fa1[ENSG==g]
  
  tempi <- PBonly_MQ_dt_melt[ENSG==g]
  tempiseqs <- tempi[,1,.(Type,Seq)]
  
  fun <- lapply(1:length(pseqs$seq),function(it){
    x <- pseqs$seq[it]
    v <- visualize_peptide_mapping(x,tempiseqs$Seq,as.numeric(factor(tempiseqs$Type,c("","PB_Only","Pacbio_Fasta"),c("grey","orange","red")))+1,line_break = 25,charactersize = 2)#+theme(point)
    if(all(v$table$color=="black")){
      v <- NULL
    }else{
      v$gg <- v$gg+ggtitle(paste(pseqs$ENSG[it],pseqs$ENST[it],sep="\n"))+theme(legend.title = element_text(size=1))
    }

    v$gg
  })
  
  ## Extracting Traces from original Isofrac Analysis
  gel_selection <- unique(tempi$Gel)
  
  ggtraces_list <- lapply(sort(as.numeric(factor(gel_selection,c("5per","8per","10per"),1:3)),decreasing = T),function(x){
    g_id_list   <-  list(EXAMPLE_ENSG=g,
                      GN="",
                      gel=x
                      
    )
    ggtraces <- ggplot()
    try({
      ggtraces <-  Traces_Prot_Pure(g_id_list,sql = "./IsoFrac_0.84/TableListNormEnsHdbScanMW_Peaks.sqlite",DBpath = "ProteinSequencesCategories2.rda")
      
    })
    ggtraces+theme(legend.position = "none")
  })
  
  
  ggtraces_list <- append(ggtraces_list,list(ncol=1))

  gg <- ggplot(tempi,aes(round(10^as.numeric(MW)/1000,1)   ,as.numeric(value),group=Seq,color=as.character(cl)))+
    geom_line()+facet_wrap(Gel~Type,scales = "free",nrow = length(unique(tempi$Gel)))+ggtitle(g)+theme(legend.position = "none")+ylab("peptide traces [zscore]")+xlab("MW [kD]")+annotate("point",Candidates$MW,Candidates$V1,pch=24,bg="black")+annotate("point",Candidates$MW,Candidates$V1,pch=25,bg="black")
  ggnew <- plot_grid(plot_grid(gg,  do.call(plot_grid,ggtraces_list),rel_widths =  c(1,0.4),nrow=1),do.call(plot_grid,fun),rel_widths = c(1,0.5))
  
  # dbDisconnect(db)
  # rm(db)
  plot(ggnew)
}

dev.off()
system("open PB_Only_Traces.pdf")

# save(PBonly_MQ,file="PBonly_MQ.rda")

# load("TableListNormEnsHdbScanMW_Peaks.sqlite")


## compare Databaseposition of selected proteins



PBspecific[ENSG=="ENSG00000157827"&!UniProt_Fasta,1,.(Sequence,PACBIO)]

UP <- ProteinSequencesCategories[grep("ENSG00000159692",ENSG)]
PB <- fa1[ENSG=="ENSG00000157827"]$Sequence
fa1[grep("VHDFPSGNGTGGR",seq)]

sequenceCheck <- lapply(ggcandidates,function(x){
  list(UP=ProteinSequencesCategories[grep(x,ENSG,fixed=T)],
       PB=fa1[grep(x,ENSG,fixed=T)],
       SpecificPeptides=PBspecific[ENSG==x&!UniProt_Fasta,1,.(Sequence,PACBIO)],
       ENSG=x)
})
library(msa)
names(sequenceCheck) <- ggcandidates
save(sequenceCheck,file="PacBIO_sequenceCheck.rda")
## Alignment of Sequences:
# msa::msa(PB,UP)
for(i in sequenceCheck){
  UP <- i$UP
  PB <- i$PB
  sp <- i$SpecificPeptides$Sequence
  ve <- c(UP$SEQUENCE,PB$seq,sp)
  names(ve) <- c(UP$ID,PB$ENST,i$SpecificPeptides$Sequence)
  dfAlign <- msa(se <- ve,type = "protein",verbose=F,method = "ClustalOmega")
  msaPrettyPrint(dfAlign, output="pdf", file=paste("Alignment_",i$ENSG,".pdf",sep=""), 
                 showNames="left", showLogo="top", logoColors="chemical",
                 shadingMode="similar", showLegend=TRUE)
  
}


#####
  
Mapping_Wrapper <- function(Proteins,Name,Sequences,Categories=NULL){
  if(length(Categories)==0){
    Categories <- rep(2,length(Sequences))
  }else{
    # Categories <- as.factor(Categories)
  }
  fun <<- lapply(1:length(Proteins),function(it){
    x <- Proteins[it]
    v <- visualize_peptide_mapping(x,Sequences,Categories,line_break = 25,charactersize = 2)#+theme(point)
    # if(all(v$table$color=="black")){
    #   v <- NULL
    # }else{
    #   v$gg <- v$gg+ggtitle(paste(Name[it]))+theme(legend.title = element_text(size=1))
    # }
    
    v$gg+ggtitle(Name[it])+theme(title = element_text(size=4))
  })
  fun <- fun[which(lengths(fun)>0)]
  list(gg=do.call(plot_grid,fun),Object=fun)
  
}


#ENSG00000153827


pdf("CheckPeptides_PB.pdf",width=15)

for(i in ggcandidates){
  UP <- ProteinSequencesCategories[ENSG==i]
  UP <- UP[order(nchar(SEQUENCE))]
  
  PB_db <- fa1[ENSG==i]
  commonsequences <- intersect(UP$SEQUENCE,PB_db$seq)
  
  UP <- UP[is.na(match(UP$SEQUENCE,commonsequences)),]
  PB_db <- PB_db[is.na(match(PB_db$seq,commonsequences)),]
  
  PBspec <- PBonly_MQ_dt_melt[ENSG==i&PB_Only==TRUE,1,.(Seq,Type)]
  unclear <- aregexec("LGAGALNAGSYASLGR",UP$SEQUENCE)
  UP$SEQUENCE[sapply(unclear,"[[",1)!="-1"]
  
  grep("LGAGALNAGSYASLGR",PB_db$seq,value=T)
  PB_db$seq[4
            ]
  
  
  try({
    if(dim(UP)[1]>0){
      UPGG <- Mapping_Wrapper(UP$SEQUENCE,UP$ProteinID,PBspec$Seq,rep("red",length(PBspec$Seq)))$gg
    }else{
      UPGG <- ggplot()
    }
  })
  try({
    if(dim(PBspec)[1]>0){
      PBGG <- Mapping_Wrapper(PB_db$seq,PB_db$ENST,PBspec$Seq,rep("red",length(PBspec$Seq)))$gg  
    }else{
      PBGG <- ggplot()
    }
    
  })
 
  if(length(UPGG)==0){
    UPGG <- ggplot()
  }
  if(length(PBGG)==0){
    PBGG <- ggplot()
  }
  
  p<- plot_grid(UPGG,PBGG,labels = i, label_y = 1.2, rel_heights = c(.65,.35)

  )
  plot(p)
}




dev.off()
system("open ./CheckPeptides_PB.pdf")



save(Results_CombinedGels_PacBio,file="Results_CombinedGels_PacBio.rda")


```




# Figure 4 

## Tead3 Traces
```{r}
source("./IsoFrac_0.85/Isofrac_Functions.R")

library(RSQLite)
library(data.table)
Extract_IsoFrac_Info <- function(dbPath,input){
  
  print("Connecting")
  db <- dbConnect(SQLite(),dbPath)
  print("Checking")
  gnhu <<- input$GN
  if(gnhu==""){
    return(NULL)
  }
  
  dt_gn_all <<- data.table(dbReadTable(db,"GN"))
  dt_gn_sel <- dt_gn_all[gel==input$inputtable]
  
  gnpos <- (which(dt_gn_sel$ens==input$GN))
  dataset_init <- list()
  tabname <<- paste("TMTSubset_final_singlenorm_zscore",input$inputtable,sep = "_#_")
  temp <- dbReadTable(db,tabname)
  TMTValues <- data.table(temp[gnpos,])
  
  dataset_init$inputtable <-TMTValues
  
  # etracting MW info
  tabname <- paste("Protein_MW",input$inputtable,sep = "_#_")
  dataset_init$Protein_MW <- dbReadTable(db,tabname)
  
  tabname <- paste("Fractions_MW",input$inputtable,sep = "_#_")
  dataset_init$Fractions_MW <- dbReadTable(db,tabname)
  
  tabname <- paste("hdbscanTables",input$inputtable,sep = "_#_")
  dataset_init$hdbscanTables <- data.table(dbGetQuery(db,paste("SELECT * FROM",paste("'",tabname,"'",sep = "") ,"WHERE EnsG = ",paste("'",input$GN,"'",sep=""))))
  
  # making clusters compatible
  d <- dataset_init$hdbscanTables
  nonzeros <- d$BestCluster[d$BestCluster!="0"]
  nonzeros <- c(0,unique(d$BestCluster))
  nonzeros[is.na(nonzeros)] <- 0
  d$BestCluster <- factor(d$BestCluster,unique(nonzeros))
  dataset_init$hdbscanTables <- d
  
  dataset_init$ID <- gnpos
  dataset_init$dt_gn <- dt_gn_sel[gnpos]
  # dbReadTable(db,"CombinedGels_Compressed_Pairs")
  ## For Network
  # dbListTables(db)
  dataset_init$IdentifiedPeaks_Table <- data.table(dbGetQuery(db,paste("SELECT * FROM IdentifiedPeaks_Table WHERE Ens=",paste("'",input$GN,"'",sep="")
                                                                       ,"AND gel=",paste("'",input$inputtable,"'",sep="")))
                                                   ,integer64="double")
  dataset_init$CombinedGels_Pairs_Filtered <- data.table(dbGetQuery(db,paste("SELECT * FROM CombinedGels_Pairs_Filtered WHERE ENSG =",paste("'",input$GN,"'",sep=""))),integer64="double")
  dataset_init$CombinedGels_Pairs_Filtered_PB <- data.table(dbGetQuery(db,paste("SELECT * FROM CombinedGels_Pairs_Filtered_PACBIO WHERE ENSG =",paste("'",input$GN,"'",sep=""))),integer64="double")
  
  dataset_init$Pairs_Filtered_Compressed <- data.table(dbGetQuery(db,paste("SELECT * FROM CombinedGels_Compressed_Pairs WHERE ENS =",paste("'",input$GN,"'",sep=""))),integer64="double")
  
  
  ###
  dataset_init$IsoFrac_PeakResultsTable <- DF_numericClass(data.table(dbGetQuery(db,paste("SELECT * FROM IsoFrac_PeakResultsTable WHERE ENS =",paste("'",input$GN,"'",sep=""),"AND gel =",paste("'",input$inputtable,"'",sep="")))))
  ####
  tempPeaks <- dataset_init$Pairs_Filtered_Compressed[,{
    hu <-  lapply(.BY,strsplit,split=";")
    # .BY <- NULL
    names(hu) <- paste(names(hu),"split",sep="_")
    hu$Compressed_MW_Variants <- .GRP
    lapply(hu,unlist)
    
  },.(gels,clusters,heights,Peaks_collapse,MW_lm)]
  tempPeaks <- tempPeaks[gels_split==input$inputtable]
  tempPeaks$cluster2 <- tempPeaks$clusters_split
  tempPeaks$MW_lm_Compressed_ <- tempPeaks$MW_lm
  # dataset_init$IdentifiedPeaks_Table
  dataset_init$IdentifiedPeaks_Table <- dataset_init$IdentifiedPeaks_Table[!is.na(match(Peak,tempPeaks$Peaks_collapse_split))&!is.na(match(cluster2,tempPeaks$clusters_split))]
  # dataset_init$dt_gn <- dt_gn_all[ens==input$GN]
  
  dbDisconnect(db)
  
  dataset_init
  
}
Traces_Plotly <- function(input_data,mpwcutoff=0,converttoMW=T,cbPalette= c( "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7"),PurePlot =F,showlegend=T,MeanCluster=F,ShowQuantiles=T,addPolygon=T,linelwd=2,  vjust_value =-2,StarSize=5){
  
  #mpwcutoff=0,converttoMW=T,cbPalette= c( "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7"),PurePlot =F,showlegend=T,MeanCluster=F,ShowQuantiles=T,addPolygon=T
  
  input_data <<- input_data
  ens <- paste(unique(input_data$dt_gn$ens, input_data$dt_gn$hgnc_symbol),collapse="")
  hdb <- input_data$hdbscanTables
  tempValues <- input_data$inputtable
  hdb <- hdb[match(input_data$ID,hdb$id),]
  
  tempValues$cl <- as.numeric(as.character(hdb$cl_2))
  vec <- 0:max(tempValues$cl,na.rm=T)
  tempValues$cl <- factor(tempValues$cl,vec,c("noise traces",vec[-1]))
  tempValues$id <- hdb$id
  
  
  
  if(any(grepl("^V[0-9]",colnames(tempValues)[1]))){
    me <- melt(tempValues,measure.vars = paste("V",1:28,sep = ""),id.vars = c("cl","id"))
    me$variable <- as.numeric(gsub("V","",me$variable))
  }else{
    colnames(tempValues) <- gsub("X","",colnames(tempValues))
    me <- melt(tempValues,measure.vars = as.character(2:29),id.vars = c("cl","id"))
    me$variable <- as.numeric(as.character(me$variable))
    me$variable <- me$variable-min(me$variable)
    # me <<- me
    # plot(me$variable2,as.numeric(as.character()))
    # abline(0,1)
    # length(table(me$variable))
    # stop()
  }
  # me <<- me
  # stop()
  
  me <- me[!is.na(variable)]
  
  if(length(input_data$Fractions_MW$x.Fractions_predicted_MW)>0&converttoMW){
    mds <- input_data$Fractions_MW$x.Fractions_predicted_MW[1:28]
    # mds <- 10^predict(md,2:29)
    # mds <- 10^predict(md,1:28)
    # table(me$variable)
    # (table(me$variable))
    me$variable <- 10^mds[match(me$variable,1:28)]
    # me <- me[order(variable)]
  }
  
  
  # med <-  dcast(me,variable~paste(cl,id,sep ="_"),value.var = "value")
  # fig <- plot_ly(med,x=~variable,y=~names(med)[2],type = 'scatter', mode = 'lines')
  # for(i in unique(me$cl)){
  #   add <- grep(paste("^",i,"_",sep = ""),names(med),value=T)
  #   for(u in add){
  #     fig <- fig%>%add_trace(y=add)
  #   }
  # }
  me <- rbind(me,me[,.(value=max(me$value)+1,variable=NA),.(id,cl)])
  
  me$cl <- (factor(me$cl))
  me$seq <- input_data$dt_gn$seq[match(me$id,input_data$ID)]
  
  me <- me[,.SD[order(variable)],id]
  me <- me[order(id),]
  me[,idtemp:={.GRP},.(cl,id)]
  try({me <- me[order(as.numeric(idtemp)),]})
  try({me <- me[order(as.character(seq)),]})
  try({me <- me[as.numeric(order(as.numeric(cl))),]})
  try({me <- me[order(as.numeric(id)),]})
  
  # me <- me[,.SD[order(as.numeric(id))],id]
  
  idCount <- 0
  idRemove <- c()
  for(i in unique(me$cl)){
    id <- me[cl==i]$id
    id <- unique(id)
    id <- (1:length(id))+idCount
    idCount <- max(id)
    idRemove <- c(id[-1])
    
    
  }
  
  if(MeanCluster){
    me <- me[,{list(value=mean(value,na.rm = T),q90=quantile(value,0.9,na.rm = T),q10=quantile(value,0.1,na.rm = T),n=length(value),seq=paste(unique(seq),collapse="\n"),idtemp=.GRP,id=.GRP)},.(cl,variable)]
  }
  
  
  cbp <- colorRampPalette(cbPalette)(max(as.numeric(me$cl),na.rm=T))
  cbp <- c("#73FDFF",cbp)
  
  if(PurePlot){
    cbp <- "#00000080"
    me$color <- "black"
    me$name <- ""
    me$legendgroup <- ""
    # me$cl <- ""
  }else{
    me$color <- cbp[as.numeric(as.character(me$cl))+1]
    
  }
  
  me <- me
  me$q90a <- me$q90
  me$q10a <- me$q10
  if(ShowQuantiles&MeanCluster==T){
    me$q90 <- abs(me$q90-me$value)
    me$q10 <- abs(me$value-me$q10)
  }else{
    me$q90 <-NA
    me$q10 <-NA
    
  }
  
  
  me <<- me[!is.na(me$cl),]
  # me <- me[]
  # me <- me8
  
  
  g2 <- ggplot(me,aes(as.numeric(variable)/1000,value,group=id,color = as.character(cl)))+geom_line(lwd=linelwd)
  # Adding Peaks
  Peaks <- input_data$Pairs_Filtered_Compressed
  
  
  g2 <- g2+ annotate("point", x = 10^as.numeric(Peaks$MW_lm_Compressed_)/1000, y = as.numeric(Peaks$Height), color = "red",bg="red", shape = 24, size = StarSize,lwd=20)+ 
    annotate("point", x = 10^as.numeric(Peaks$MW_lm_Compressed_)/1000, y = as.numeric(Peaks$Height), color = "red",bg="red", shape = 25, size = StarSize,lwd=20)+
    annotate("text", x = 10^as.numeric(Peaks$MW_lm_Compressed_)/1000, y = as.numeric(Peaks$Height),label=round(10^as.numeric(Peaks$MW_lm_Compressed_)/1000,1),vjust=vjust_value,hjust=0.5,angle=0, color = "red",bg="red", shape = 25, size = 5,lwd=20)+ylab("peptide intensity trace [zscore]")+xlab("MW [kD]")+
    ggplot2::scale_color_viridis_d()
  
  p <- plot_ly(me, 
               x = ~variable, 
               y = ~value, 
               # id = ~id,
               color=~cl,
               name=~cl,
               legendgroup=~cl,
               type = 'scatter',
               mode = 'lines',
               hoverinfo="text",
               text=me$seq,
               colors=cbp,
               line=list(width=3))%>%
    layout(showlegend = TRUE) 
  
  p <- try(p%>%
             style(showlegend = FALSE, traces = idRemove))
  # print("hu")
  if(addPolygon&MeanCluster){
    for(i in unique(me$cl)){
      print(i)
      mei <- me[cl==i]
      
      
      mei <- mei[order(variable,decreasing = T)]
      mei <- mei[!is.na(mei$variable)]
      mei$color[is.na(mei$color)] <- "#73FDFF"
      
      
      p <- p %>% add_polygons(
        x=c(mei$variable,rev(mei$variable),mei$variable[1]),
        y=c(mei$q90a,rev(mei$q10a),mei$q90a[1]),
        line=list(width=2,color="transparent"),
        fillcolor=paste(mei$color[1],20,sep=""), inherit = FALSE,legendgroup=mei$cl[1],name="0.1 to 0.9 Q")
    }
    
  }
  # print("hu")
  
  p <- p%>%layout(title = ens, xaxis = list(title = 'MW converted fraction'), 
                  yaxis = list(title = "zscore"), legend = list(title=list(text='Peptide Clusters')),showlegend=showlegend)
  
  
  list(p=p,g=g2,data=me,clusterColors=cbp) 
}

dbPath <- "IsoFrac_0.84/TableListNormEnsHdbScanMW_Peaks.sqlite"

input <- list(GN="ENSG00000007866",inputtable="Exp1_10per_Try")
dataset <- Extract_IsoFrac_Info(dbPath,input)
Tead3_Traces <- Traces_Plotly(dataset,linelwd = 0.75,StarSize = 2.5)
```

### TMPO Traces

```{r}
if(!exists("TMPO_Traces")){
 input <- list(GN="ENSG00000120802",inputtable="Exp1_10per_Try")
  dbPath <- "IsoFrac_0.85/TableListNormEnsHdbScanMW_Peaks.sqlite"

dataset <- Extract_IsoFrac_Info(dbPath,input)
dataset$hdbscanTables <- dataset$hdbscanTables[BestCluster!=0]

LM <- as.numeric(dataset$Pairs_Filtered_Compressed$Height)
LM[LM<0.35] <- 3.5
dataset$Pairs_Filtered_Compressed$Height <- LM

TMPO_Traces <- Traces_Plotly(dataset,ShowQuantiles = T,linelwd = 0.5,StarSize = 2.5) 
TMPO_Traces$g <- TMPO_Traces$g+ylim(c(-1,5))+theme(legend.title = element_blank())


dataset$CombinedGels_Pairs_Filtered
}



# A,B,C,G,D,E,F
f4_p3 <- plot_grid(AUC,Results_CombinedGels_PBUP$RNAIdentifiedFraction$Pie,rel_widths = c(0.75,1),labels = c("D","E"))

# D & E old figure 7
AUC <- Results_CombinedGels_PBUP$ROC+theme(legend.position = "right")+guides(fill=guide_legend(ncol=1,byrow=TRUE,label.hjust = -0.2))+xlab("Specificity")+ggtitle("ROC PACBIO Best Matching Entry")


xpath <- pdf_convert("./WB_Confirmation/TMPO_PACBIO_example.pdf",dpi = 300)
TMPO_Pacbio <- ggdraw()+draw_image(xpath)


# E & F

lowerrow_RNA_PROTEIN_Integration <- plot_grid(DeltaMassFig+ggtitle("Isoform Pairs UniProt vs PACBIO\n Uniprot or PACBIO Score > 0.8")+
                        update_geom_defaults(geom="point",new =list(size=0.5) ),DeltaMassFigCounts+ggtitle("Counts")+xlab("")+
                        theme(legend.position = "none",axis.text.x=element_text(angle=45,vjust=1,hjust=1)),
                      ProteinCategory_BP+ theme(legend.position = "bottom",legend.margin = margin(l = -2, unit = "cm"))+guides(fill=guide_legend(nrow=2,byrow=TRUE,label.hjust = -0.2,override.aes = aes(label = "")))
                      ,rel_widths = c(1.3,0.5,1),nrow=1,rel_heights = c(1,0.7),labels = c("F","","G"))



#upperrow
#lowerrow

f4_p2 <- plot_grid(TMPO_Pacbio+ggtitle("PACBIO Results for TMPO"),
                   plot_grid(TMPO_Traces$g+ggtitle("ISOFRAC Traces for TMPO")+ylim(c(-1,6)),AUC,nrow=1,labels=c("","E"),rel_widths = c(1,0.85)),
                   Results_CombinedGels_PBUP$RNAIdentifiedFraction$Pie,labels = c("C","D","F"),nrow=3,scale = c(0.9,0.9,1),rel_widths = c(1,0.75))


# A & B 

# placeholder for Mengrans figures:

f <- fread("comprehensiveORF_PacBio_GENCODE_v2.txt",sep = "\t")

ORFCounts <- f[,.(dim(.SD)[1],length(unique(ORF_seq))),GENCODE_gene_ID]
ORFcounts <- ORFCounts[,.("unique ORFs"=length(GENCODE_gene_ID),"unique Transcripts"=sum(V1)),V2]
ORFcounts2 <- ORFcounts[V2>=8,.(V2="8+","unique ORFs"=sum(as.data.frame(.SD)[,2]),"unique Transcripts"=sum(as.data.frame(.SD)[,3]))]
ORFcounts3 <- rbind(ORFcounts[V2<8],ORFcounts2)

ORFcountsAll <- melt(ORFcounts3,id.vars = "V2")

RNAcounts <- ggplot(ORFcountsAll,aes(V2,value,fill=variable,label=value))+geom_bar(stat="identity",position="dodge")+xlab("Variants per Gene")+ylab("Count")+theme(legend.title = element_blank())+geom_text(angle=90,position = position_dodge(width = 0.9), vjust = 0.5,hjust=-0.2)+theme(legend.title = element_blank())

A4 <- fread("Figure4B_data.csv",sep=",")
A4Melt <- melt(A4)
# ggplot(A4Melt,aes(variable,value,label=value,fill=Count_trans_ORFs_per_gene ))+geom_bar(stat="identity",position="dodge")+geom_text(angle=90,position = position_dodge(width = .9),hjust=1)+xlab("# Isoforms per Gene")+theme(legend.position = "bottom",legend.title = element_blank())+ylab("Genes")

RNAcounts <- ggplot(A4Melt,aes(variable,value,fill=Count_trans_ORFs_per_gene,label=value))+geom_bar(stat="identity",position="dodge")+xlab("Variants per Gene")+ylab("Count")+theme(legend.title = element_blank())+geom_text(angle=90,position = position_dodge(width = 0.9), vjust = 0.5,hjust=-0.2)+theme(legend.title = element_blank())+ylim(c(0,4300))

library(pdftools)
# lgg_mengran_placeholder <- lapply(c("./Figure4_v2/4A.pdf","./Figure4_v2/4B.pdf"),function(x){
#   x <<- x
#   xpath <- pdf_convert(x,dpi = 50)
#   Mengran1 <- draw_image(xpath)
#   Mengran1 <- ggdraw()+Mengran1
# })
xpath <- pdf_convert("./Figure4_v2/Mengran_Sketch_mod.pdf",dpi = 300)
Mengran1 <- ggdraw()+draw_image(xpath)
# G

f4_p1 <- plot_grid(Mengran1,Results_CombinedGels_PacBio$BP_Counts,RNAcounts+theme(legend.position = "bottom"),
                   rel_widths = c(0.9,0.7,0.9),ncol=3,labels=c("A","B",""),scale = c(0.8,1,1))#RNAcounts+theme(legend.position = "bottom")
f4 <- plot_grid(f4_p1,f4_p2,ncol=1,rel_heights = c(0.4,1),labels = c("","","G"))
pdf("Figure_4.pdf",height=12,width=10)
plot(f4)
graphics.off()
ggsave("Figure_4.jpeg",f4,width = 10, height = 12)



```

# Figure 5 

```{r}
FMNL2_Example_PACBIO <- ggdraw()+draw_image("WB_Confirmation/20241212_IdentifyNewIsofroms_v2.png")
xpath <- pdftools::pdf_convert("./WB_Confirmation/FMNL2_NewVariant.pdf",dpi = 300)
FMNL2_Example_PACBIO <- ggdraw()+draw_image(xpath)

if(!exists("evitempMeltMax")){
  evitempMeltMax <- evitempMelt[variable!="V1",{.SD[value==max(value)][1,]},.(Sequence,variable)]
  evitempMeltMax[Group!="Known"]
  ShowPoints <- evitempMeltMax[variable!="V1"&Group=="Specific Peptides\nfrom\nNew Proteins"]

}
pvalue <- evitempMeltMax[,{
  wc <<- wilcox.test(value[Group=="Known"],value[Group!="Known"])
  wc$p.value
  
},variable]
pvalue$Group <- "Specific Peptides\nfrom\nNew Proteins"

unique(evitempMeltMax[Group!="Known"]$Sequence)
PlotCollectionFigure6$MQ$Features_evi <- ggplot(evitempMeltMax[variable!="V1"],aes(Group,value,fill=Group))+geom_boxplot()+facet_wrap("variable",scales = "free")+theme(legend.position = "none")+xlab("")+
  geom_jitter(data = ShowPoints, aes(x = Group, y = value),width = 0.2, height = 0)+geom_text(data = pvalue, aes(x = Group, y = c(40,450),label=paste("p =",round(V1,2))),size=4)
PlotCollectionFigure6$MQ$Features_evi 

if(!exists("IsoFrac_Domain_SequenceAnalysis")){
 load("IsoFrac_Domain_SequenceAnalysis_woSemi.rda") 
  
}  
if(!exists("SemiTrypticInfo")){
  load("SemiTrypticInfo2.rda")

}



library(Peptides)
library(msa)
library(cowplot)

# try(Tead3 <- IF_DB_Peptidemapping_modified(GeneList$TEAD3,IsoFrac_Domain_SequenceAnalysis,PrFeatures_evioteinSequencesCategories,invertOrder = T,allgels=T,facetPlot = F))
try(Tead3 <- IF_DB_Peptidemapping_modified(GeneList$TEAD3,IsoFrac_Domain_SequenceAnalysis,ProteinSequencesCategories,invertOrder = T,allgels=T,facetPlot = F))
try(NELFA <- IF_DB_Peptidemapping_modified(GeneList$NELFA,IsoFrac_Domain_SequenceAnalysis,ProteinSequencesCategories,invertOrder = T,allgels=T,facetPlot = F))
try(CTTNBP2N <- IF_DB_Peptidemapping_modified(GeneList$CTTNBP2N,IsoFrac_Domain_SequenceAnalysis,ProteinSequencesCategories,invertOrder = T,allgels=T,facetPlot = F))
try(RRP1B <- IF_DB_Peptidemapping_modified(GeneList$RRP1B,IsoFrac_Domain_SequenceAnalysis,ProteinSequencesCategories,invertOrder = T,allgels=T,facetPlot = F))
pdf("Selected_Trace_Plots.pdf",width=10,height=12)
plot_grid(
  NELFA$gg+theme(legend.position = "none"),NELFA$MSAPlot+theme(legend.position = "none"),
  CTTNBP2N$gg+theme(legend.position = "none"),CTTNBP2N$MSAPlot+theme(legend.position = "none"),
  RRP1B$gg+theme(legend.position = "none"),RRP1B$MSAPlot+theme(legend.position = "none")

)
dev.off()
#system("open Selected_Trace_Plots.pdf")

# Tead3$gg
# HUmpi$MSAPlot3
# HUmpi$MSAPlot2

PlotCollectionFigure6$MQ$Counts
scales_add_defaults <<- scales_add_defaults
PlotCollectionFigure6$MQ$Counts

PacBioSection_v0 <- plot_grid(
  PlotCollectionFigure6$MQ$Counts_v2+theme(legend.position = "bottom")+guides(fill=guide_legend(nrow=2,byrow=TRUE)),FMNL2_Example_PACBIO,labels = c("A","B"),rel_widths = c(0.3,1),scale=c(0.9,0.9))
PacBioSection <- plot_grid(FMNL2_Example_PACBIO,labels = c("?"),scale=c(0.9,0.9))

PacBioExamples <- plot_grid(ggplot(),ggplot(),labels = c("C","D"))

FirstRow <- plot_grid(PacBioSection,ncol=1,labels="")
# Tead3_Traces <-list(g=ggplot())
#
WB_Subset_subset <- plot_grid(plot_grid(WB_BirA_nterm,Tead3_nterminal,ncol=1,rel_heights = c(0.5,1),scale=c(0.6,1)),
                              plot_grid(WB_BirA_cterm,Tead3_cterminal,ncol=1,rel_heights = c(0.5,1),scale=c(0.8,1)),nrow=1)
WB_Subset_1 <- plot_grid(WB_BirA_nterm,WB_BirA_cterm,Tead3_WB,ncol=1,rel_heights = c(0.3,0.3,1),labels=c("G",""),scale=c(0.9,0.9,1))
WB_Subset <- plot_grid(WB_Subset_1,plot_grid(Tead3_Traces$g+xlim(c(15,60))+theme(legend.title = element_blank()),Tead3$MSAPlot3,ncol=1,scale = 0.9,labels=c("F","G")),labels = c("",""),rel_widths = c(0.5,1))

WB_Subset <- plot_grid(
  plot_grid(Tead3_Traces$g+xlim(c(15,60))+theme(legend.title = element_blank()),Tead3$MSAPlot3,ncol=1,scale = 0.9,labels=c("E","F")),
  WB_Subset_1,labels = c("",""),rel_widths = c(1,0.75))

library(cowplot)
FMNL2_Example_PACBIO <- ggdraw()+draw_image("WB_Confirmation/20241212_IdentifyNewIsofroms_v2.png")
xpath <- pdftools::pdf_convert("./WB_Confirmation/FMNL2_NewVariant.pdf",dpi = 300)
FMNL2_Example_PACBIO <- ggdraw()+draw_image(xpath)

upperrow_RNA_PROTEIN_Integration <- plot_grid(Category_BP+theme(legend.position = "none",axis.text.x = element_text(angle=90,hjust=1,vjust=0.5)),
                                              Category_BP_rel+theme(legend.position = "bottom",legend.text = element_text(size=4))+guides(color = guide_legend(override.aes = list(label = NA_character_)))+guides(fill=guide_legend(nrow=2,byrow=TRUE)),rel_heights = c(0.75,1),ncol=1)

lowerrow_RNA_PROTEIN_Integration <- plot_grid(
  DeltaMassFig+ggtitle("Isoform Pairs UniProt vs PACBIO\n Uniprot or PACBIO Score > 0.8")+update_geom_defaults(geom="point",new =list(size=0.5) )+theme(legend.position = "none",legend.text = element_text(size=5)),
  DeltaMassFigCounts+ggtitle("Counts")+xlab("")+
                        theme(legend.position = "none",axis.text.x=element_text(angle=45,vjust=1,hjust=1)),
                      ProteinCategory_BP+ theme(legend.position = "bottom",legend.margin = margin(l = -2, unit = "cm"))+guides(fill=guide_legend(nrow=2,byrow=TRUE,label.hjust = -0.2,override.aes = aes(label = "")))
                      ,rel_widths = c(1.3,0.7,1),nrow=1,rel_heights = c(1,0.7),labels = c("B","","C"))

pdf("Figure_5_v3.pdf",height=12,width=10)
fig5 <- plot_grid(
  plot_grid(
    upperrow_RNA_PROTEIN_Integration,
    plot_grid(lowerrow_RNA_PROTEIN_Integration,FMNL2_Example_PACBIO,ncol=1,labels=c("","D")),ncol=2,rel_widths = c(0.3,1),labels=c("A","B")),
                  
          # Sceme,
          WB_Subset,ncol = 1,rel_heights = c(1,1,1.5),scale=c(1,1,1),labels=c("",""))
fig5

dev.off()

ggsave("Figure_5_v3.jpeg",fig5,width = 10, height = 12)
# read_parquet("E:/MaxQuant/TIA/20240621_SEC_SDS_Cytosol_v4_60ug/lib.predicted.speclib")

```

